{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import diverse_yeast_tools as dyt\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Use yeast_esr_expression functions? (Y/n) Y\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added C:\\Users\\heineib\\Documents\\GitHub\\yeast_esr_expression_analysis\\ to path\n",
      "Importing yeast_esr_exp.  If autoreload, may need to reset base_dir and data_processing dir \n",
      "  yeast_esr_exp.base_dir=base_dir \n",
      " yeast_esr_exp.data_processing_dir = data_processing_dir\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Use y1000plus_tools functions? (Y/n) Y\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added C:\\Users\\heineib\\Documents\\GitHub\\y1000plus_tools\\ to path\n",
      "y1000plus_tools dir is unset\n",
      "y1000plus data dir is unset\n",
      "y1000plus_tools dir is C:\\Users\\heineib\\Documents\\GitHub\\y1000plus_tools\\\n",
      "y1000plus data dir is C:\\Users\\heineib\\Documents\\GitHub\\y1000plus_tools\\data\\\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "import ast\n",
    "\n",
    "import json\n",
    "import pickle\n",
    "\n",
    "import warnings\n",
    "\n",
    "from Bio import SeqIO, pairwise2, PDB\n",
    "\n",
    "\n",
    "\n",
    "import sys\n",
    "# if not(base_dir in sys.path): \n",
    "#     sys.path.append(base_dir)\n",
    "#     print(\"Added \" + base_dir + \" to path\" )\n",
    "\n",
    "\n",
    "yeast_esr_flag = input(\"Use yeast_esr_expression functions? (Y/n)\")\n",
    "\n",
    "if yeast_esr_flag =='Y':\n",
    "    yeast_esr_exp_path = os.path.normpath('C:/Users/heineib/Documents/GitHub/yeast_esr_expression_analysis') + os.sep\n",
    "    #io_library_path_core = io_library_path + 'core' + os.sep\n",
    "    if not(yeast_esr_exp_path in sys.path):\n",
    "        sys.path.append(yeast_esr_exp_path)\n",
    "        print(\"Added \" + yeast_esr_exp_path + \" to path\" )\n",
    "\n",
    "    import yeast_esr_exp\n",
    "    yeast_esr_exp.base_dir = yeast_esr_exp_path\n",
    "    yeast_esr_exp.data_processing_dir = yeast_esr_exp_path + os.path.normpath('expression_data') + os.sep\n",
    "\n",
    "\n",
    "y1000plus_tools_flag = input(\"Use y1000plus_tools functions? (Y/n)\")\n",
    "\n",
    "if y1000plus_tools_flag =='Y':\n",
    "    y1000plus_tools_path = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools') + os.sep\n",
    "    #io_library_path_core = io_library_path + 'core' + os.sep\n",
    "    if not(y1000plus_tools_path in sys.path):\n",
    "        sys.path.append(y1000plus_tools_path)\n",
    "        print(\"Added \" + y1000plus_tools_path + \" to path\" )\n",
    "\n",
    "    import y1000plus_tools\n",
    "    y1000plus_tools.base_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools') + os.sep\n",
    "    print(\"y1000plus_tools dir is \" + y1000plus_tools.base_dir )                      \n",
    "    y1000plus_tools.y1000plus_dir = y1000plus_tools.base_dir + 'data' + os.sep \n",
    "    print(\"y1000plus data dir is \" + y1000plus_tools.y1000plus_dir)\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "#from venn import venn\n",
    "\n",
    "\n",
    "genomes_base = os.path.normpath('G:/My Drive/Crick_LMS/external_data/genomes')\n",
    "y1000plus_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools/data') + os.sep\n",
    "base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#Load y1000 species table\n",
    "y1000_species = pd.read_csv(y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/y1000_species_table.csv\"), index_col=0)\n",
    "\n",
    "#Load S.Cer lookup table: \n",
    "scer_lookup_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/saccharomyces_cerevisiae.csv')\n",
    "\n",
    "scer_lookup = pd.read_csv(scer_lookup_fname, index_col=0)\n",
    "scer_y1000_id_gene_id_lookup = dict(zip(scer_lookup['y1000_id'], scer_lookup.index))\n",
    "\n",
    "spec_orig_genome_lookup = dict(zip(y1000_species['spec_og_id'], y1000_species['original_genome_id']))\n",
    "\n",
    "protein_dir = os.path.normpath(y1000plus_dir + 'shen_2018_data/0_332yeast_genomes/332_genome_annotations/pep') + os.sep \n",
    "\n",
    "#Percentage of species that need to be present in the orthogroup in order to use the orthogroup. \n",
    "pct_present_thresh = 0.4\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_annotations_df = pd.read_csv(base_dir + os.sep + 'eggnog_mapper' + os.sep +  'annotation' + os.sep + 'yeastmine_annotation_summary_20211007.csv', index_col = 0)\n",
    "\n",
    "scer_orf_table = pd.read_table(base_dir + os.sep + 'eggnog_mapper' + os.sep +  'annotation' + os.sep + 'yeastmine_orf_list_20211025.tsv', index_col=1, names= ['SGD_id', 'species', 'Name', 'full_name', 'qualifier', 'status'])\n",
    "\n",
    "scer_orf_table_not_dub = scer_orf_table[scer_orf_table['qualifier'] != 'Dubious']\n",
    "\n",
    "sc_annotations_orf = sc_annotations_df.loc[list(set(sc_annotations_df.index) & set(scer_orf_table_not_dub.index ))]\n",
    "\n",
    "ec_counts = []\n",
    "for eclist in sc_annotations_orf['EC number']: \n",
    "    if type(eclist)==float:\n",
    "        assert np.isnan(eclist), 'Float but not NaN ' + eclist\n",
    "        ec_counts.append(0)\n",
    "    else: \n",
    "        eclist_list = ast.literal_eval(eclist)\n",
    "        ec_counts.append(len(eclist_list))\n",
    "        \n",
    "\n",
    "print('{} s. cerevisiae proteins with an EC number'.format(sum(~(sc_annotations_orf['EC number'].isna()))))\n",
    "\n",
    "\n",
    "## How many have a PDB annotation\n",
    "\n",
    "print('{} s. cerevisiae proteins with a PDB id'.format(sum(~(sc_annotations_orf['PDB ID'].isna()))))\n",
    "\n",
    "sc_annotations_ec_subset = sc_annotations_orf[~(sc_annotations_orf['EC number'].isna())]\n",
    "\n",
    "print('{} s. cerevisiae proteins with and EC number and a PDB id'.format(sum(~(sc_annotations_ec_subset['PDB ID'].isna()))))\n",
    "\n",
    "## How many have a pathway identified\n",
    "\n",
    "sc_annotations_orf['N_pathways'] = [len(ast.literal_eval(pway_list)) for pway_list in sc_annotations_orf['Gene.pathways.name']]\n",
    "\n",
    "sc_annotations_orf[sc_annotations_orf['N_pathways']>0]\n",
    "\n",
    "# N_pathways_EC = pd.Series([len(ast.literal_eval(pway_list)) for pway_list in sc_annotations_ec_subset['Gene.pathways.name']])\n",
    "\n",
    "print('{} s. cerevisiae proteins with a Pathway'.format(sum(sc_annotations_orf['N_pathways']>0)))\n",
    "\n",
    "print('{} s. cerevisiae proteins with and EC number and a Pathway'.format(sum(((sc_annotations_orf['N_pathways']>0) & (~(sc_annotations_orf['EC number'].isna()))))))\n",
    "\n",
    "\n",
    "sc_annotations_pway_subset = sc_annotations_orf[sc_annotations_orf['N_pathways']>0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Proteins missing yeast EC number but which have a curated pathway\n",
    "sc_annotations_orf[(sc_annotations_orf['N_pathways']>0) & sc_annotations_orf['EC number'].isna()]\n",
    "\n",
    "#This list makes me think that we want the pathways and not to rely exclusively on EC numbers (since there are some potentially very interesting enzymes in here)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summarize size distribution of ortholog mappings: \n",
    "orthogroup_fname = y1000plus_dir + os.path.normpath(\"shen_2018_data/orthomcl_output/orthomcl_clusters.txt\")\n",
    "\n",
    "og_sizes = []\n",
    "\n",
    "with open(orthogroup_fname, 'r') as f: \n",
    "    for line in f: \n",
    "        og_genes = line.split()\n",
    "        og = og_genes[0].strip(':')\n",
    "        genes = og_genes[1:]\n",
    "        og_sizes.append(len(genes))\n",
    "        \n",
    "fig, ax = plt.subplots(1,1)\n",
    "\n",
    "ax.hist(og_sizes, log=True, bins = 50)\n",
    "ax.set_title('Orthogroup Sizes from Shen et al 2018')\n",
    "ax.set_xlabel('Size of orthogroup')\n",
    "ax.set_ylabel('Frequency (Log scale)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For a given S. cerevisiae protein, provide its og number\n",
    "# Make mapping of S. cerevisiae proteins to orthogroups\n",
    "\n",
    "# Species\n",
    "spec_name = 'Saccharomyces cerevisiae'\n",
    "\n",
    "#y1000_species = pd.read_csv(y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/y1000_species_table.csv\"), index_col=0)\n",
    "\n",
    "spec_id = y1000_species[y1000_species['Species name']==spec_name]['spec_og_id'].values[0]\n",
    "\n",
    "spec_og_mapping = {}\n",
    "with open(orthogroup_fname, 'r') as f: \n",
    "    for line in f: \n",
    "        og_genes = line.split()\n",
    "        og = og_genes[0].strip(':')\n",
    "        genes = og_genes[1:]\n",
    "        \n",
    "        gene_ids = [gene_id for gene_id in genes if int(gene_id.split('_')[0])==spec_id]\n",
    "        \n",
    "        if len(gene_ids) >= 1:\n",
    "            for gene_id in gene_ids: \n",
    "                spec_og_mapping[gene_id] = og\n",
    "\n",
    "                \n",
    "og_set = []\n",
    "\n",
    "for sc_genename in sc_annotations_pway_subset.index:   #sc_annotations_ec_subset.index: \n",
    "    #sc_genename = 'YER144C'\n",
    "    y1000_id = scer_lookup.loc[sc_genename, 'y1000_id']\n",
    "    og_in = spec_og_mapping[y1000_id]\n",
    "    og_set.append(og_in)\n",
    "    \n",
    "og_set = list(set(og_set))\n",
    "\n",
    "print('{} orthogroups contain S. cerevisiae genes with SGD curated pathway'.format(len(og_set)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#species_set = pd.read_table(os.path.normpath(base_dir + '/alphafold/species_selection.txt'), index_col = 0)\n",
    "#Species name\tTime_tree_name\toriginal_genome_id\tAbbreviation\tspec_og_id\tAlphafold_structures\tLong_read_seqs\tExperimental_data\tMajor clade\tLoad\tUniprot Proteome\tCUG usage\tNotes\tSequence source for Alphafold\tSequence Source for Trees\tGenome Sequence for Uniprot\n",
    "\n",
    "species_set = pd.read_csv(os.path.normpath(base_dir + '/alphafold/species_selection.csv'), index_col = 0)\n",
    "species_set_filt = species_set.loc[species_set['Load']=='Y']\n",
    "\n",
    "spec_id_map = {}\n",
    "time_tree_names = []\n",
    "for spec_fullname, (time_tree_name, spec_orig, spec_abbrev, spec_og_id, alphafold_structures,long_read_flag, experimental_data, clade, load_flag, uniprot, cug_usage, notes, seq_source_af, seq_source_trees, uniprot_genome) in species_set[species_set['Load']=='Y'].iterrows():\n",
    "    y1000_spec_lookup = y1000_species[y1000_species['original_genome_id']==spec_orig]\n",
    "    clade_true = y1000_spec_lookup['Major clade'].values[0]\n",
    "    assert clade_true==clade, 'Clade mismatch: {}, true spelling: {}, bad spelling: {}'.format(spec_fullname, clade_true, clade)\n",
    "    spec_id = y1000_spec_lookup['spec_og_id'].values[0]\n",
    "    assert spec_og_id == spec_id, 'OG id mismatch: {}, from species selection table: {}, from original database: {}'.format(spec_fullname, spec_og_id, spec_id)\n",
    "    spec_id_map[int(spec_og_id)] = spec_orig\n",
    "    time_tree_names.append(time_tree_name)\n",
    "    \n",
    "# spec_ids = set([str(spec_id_int) for spec_id_int in spec_id_map.keys()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_fname = y1000plus_dir + 'shen_2018_data' + os.sep + '332_2408OGs_time-calibrated_phylogeny_species-names_updated.newick'\n",
    "t = y1000plus_tools.Tree(tree_fname)\n",
    "\n",
    "ts = y1000plus_tools.TreeStyle()\n",
    "ts.show_leaf_name = False\n",
    "#ts.show_branch_length = True\n",
    "\n",
    "# for node in t.traverse():\n",
    "#     if not(node.is_leaf()):\n",
    "#         node_label = TextFace(node.name)\n",
    "#         node.add_face(node_label, column=1, position = \"branch-bottom\")\n",
    "\n",
    "# post_WGH_colors = {'low':  '#8cc3f6', # '#d3d3fe', #'#3192ff',#'#7eeaf7', ##2DD7ED',      #'#e6fcff', \n",
    "#                    'high': '#fcbba1'} #'#59E3EB'}  #'#ffebe6'}\n",
    "\n",
    "\n",
    "# node_color_dict = {'KLE':\"#deb9f6\",\n",
    "#                    'ZT': \"YellowGreen\",\n",
    "#                    'Post_WGH': \"LightYellow\",   #default color for post WGH\n",
    "#                    'low': '#8cc3f6',    #syntenic orthologs of low LFC ohnolog\n",
    "#                    'high': '#fcbba1',    #syntenic orthologs of high LFC ohnolog\n",
    "#                    'outgroup': 'LightGrey'\n",
    "#                   }\n",
    "\n",
    "#K.lactis, K. marxianus, K. servazzii, W.anomalus, P.pastoris, C.albicans, C.tropicalis, D.hansenii, G.candidus, R.mucilaginosa, S.pombe, S.cerevisiae\n",
    "\n",
    "\n",
    "# leaves_to_view = {'Geotrichum_candidum':('outgroup', 'Ylip'),\n",
    "#                   'Komagataella_pastoris': ('outgroup','Kpas'),\n",
    "#                   'Debaryomyces_hansenii': ('outgroup','Dhan'),\n",
    "#                   'Candida_albicans': ('outgroup','Calb'),\n",
    "#                   'Candida_tropicalis': ('outgroup','Ctro'),\n",
    "#                   'Wickerhamomyces_anomalus': ('outgroup', 'Wano'), \n",
    "#                  'Lachancea_thermotolerans':  ('KLE','Lthe'),\n",
    "#                   'Kluyveromyces_lactis': ('KLE','Klac'),\n",
    "#                   'Kluyveromyces_marxianus': ('KLE','Kmar'),\n",
    "#                   'Zygosaccharomyces_rouxii': ('ZT', 'Zrou'),\n",
    "#                   'Kazachstania_naganishii': ('Post_WGH','Knag'),\n",
    "#                   'Saccharomyces_cerevisiae': ('Post_WGH','Scer')\n",
    "\n",
    "#                  }\n",
    "\n",
    "\n",
    "#t.prune(leaves_to_view.keys())\n",
    "\n",
    "t.prune(time_tree_names)\n",
    "\n",
    "t.ladderize()\n",
    "\n",
    "t.write(outfile=base_dir + os.sep + 'alphafold' + os.sep + 'species_selection.nw')\n",
    "\n",
    "for node in t.get_leaves():\n",
    "    name = node.name\n",
    "    #group, abbrev = leaves_to_view[name]\n",
    "    \n",
    "#     nstyle = NodeStyle()\n",
    "#     node_color = node_color_dict[group]\n",
    "#     nstyle['bgcolor']=node_color\n",
    "#     node.set_style(nstyle)\n",
    "    \n",
    "    node_label_face = y1000plus_tools.TextFace(name) # + \" ({})\".format(abbrev)) #, fsize=16, fgcolor=\"black\")\n",
    "    node.add_face(node_label_face, column=1, position = \"aligned\")\n",
    "\n",
    "fname = base_dir + os.sep + 'alphafold' + os.sep + 'species_selection.svg'\n",
    "t.render(fname, tree_style=ts)\n",
    "#t.render('%%inline', tree_style=ts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spec_id_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# For a given set of species, and a given orthogroup, extract orthogroup genes\n",
    "\n",
    "og_out_data = {}  # (N_genes, pct_present, og_genes_out)\n",
    "\n",
    "for og_in in og_set: \n",
    "\n",
    "    with open(orthogroup_fname, 'r') as f: \n",
    "        for line in f: \n",
    "            og_genes = line.split()\n",
    "            og = og_genes[0].strip(':')\n",
    "\n",
    "            if og==og_in: \n",
    "                genes = og_genes[1:]\n",
    "                og_genes_out = []\n",
    "                for gene in genes: \n",
    "                    spec_id = gene.split('_')[0]\n",
    "                    if spec_id in spec_ids: \n",
    "                        og_genes_out.append(gene)\n",
    "\n",
    "    \n",
    "    #Calculate the percentage of species contained in the orthogroup.\n",
    "    og_genes_out_specs = [og_gene.split('_')[0] for og_gene in og_genes_out]\n",
    "    pct_present = len(set(og_genes_out_specs))/len(spec_ids)\n",
    "\n",
    "    #Extract Fastas for the orthogroups.  \n",
    "    \n",
    "    og_out_data[og_in] = (len(og_genes_out), pct_present, og_genes_out)\n",
    "\n",
    "\n",
    "# Save og_out_data\n",
    "og_unfiltered_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_genes_unfiltered.json'\n",
    "\n",
    "with open(og_unfiltered_fname, 'w') as f:\n",
    "    json.dump(og_out_data, f, sort_keys=True, indent=4 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load og_out_data\n",
    "og_unfiltered_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_genes_unfiltered.json'\n",
    "with open(og_unfiltered_fname, 'r') as f:\n",
    "    og_out_data = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pct_present_list = []\n",
    "#pct_present_thresh = 0.4\n",
    "og_out_data_filt = {}\n",
    "total_genes = 0\n",
    "for og, (N_genes, pct_present, og_genes_out) in og_out_data.items():\n",
    "    pct_present_list.append(pct_present)\n",
    "    if pct_present > pct_present_thresh: \n",
    "        og_out_data_filt[og] = (N_genes, pct_present, og_genes_out)\n",
    "        total_genes = total_genes + N_genes\n",
    "        \n",
    "print('{} total genes in {} orthogroups'.format(total_genes, len(og_out_data_filt)))\n",
    "\n",
    "# Save filtered og_out_data\n",
    "\n",
    "og_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_genes.json'\n",
    "\n",
    "with open(og_fname, 'w') as f:\n",
    "    json.dump(og_out_data_filt, f, sort_keys=True, indent=4 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load og_out_data_filt\n",
    "og_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_genes.json'\n",
    "with open(og_fname, 'r') as f:\n",
    "    og_out_data_filt = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make fastas for all proteins without filtering out S. cer and C. albicans using shen genomes.  \n",
    "fasta_out_fname = base_dir + os.sep + 'alphafold' + os.sep + 'all_proteins.fasta'\n",
    "\n",
    "source = 'shen_2018'\n",
    "\n",
    "#lists of gene_full names for each species\n",
    "gene_full_sets = {}\n",
    "\n",
    "with open(fasta_out_fname, 'w') as f_out: \n",
    "\n",
    "    #for each species: \n",
    "    for spec_id, spec_name in spec_id_map.items(): \n",
    "\n",
    "        spec = spec_id_map[spec_id]\n",
    "\n",
    "        spec_orig_genome = spec_orig_genome_lookup[spec_id]\n",
    "        \n",
    "        if spec_orig_genome == 'saccharomyces_cerevisiae':\n",
    "            protein_fname = y1000plus_dir + os.path.normpath('shen_2018_data/0_332yeast_genomes/332_genome_annotations/Saccharomyces_cerevisiae_S288C_protein.fasta')\n",
    "        elif spec_orig_genome == 'candida_albicans': \n",
    "            protein_fname = y1000plus_dir + os.path.normpath('shen_2018_data/0_332yeast_genomes/332_genome_annotations/Candida_albicans_SC5314_A22_current_default_protein.fasta')\n",
    "        else: \n",
    "            protein_fname = protein_dir + spec_orig_genome + '.max.pep'\n",
    "        seq_records = SeqIO.parse(protein_fname, \"fasta\")\n",
    "        \n",
    "        spec_lookup_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/' + spec_orig_genome + '.csv')\n",
    "\n",
    "        spec_lookup = pd.read_csv(spec_lookup_fname, index_col=0)\n",
    "        \n",
    "        if spec_orig_genome == 'saccharomyces_cerevisiae':\n",
    "            #for S. cerevisiae the orf name is the 'gene_full'\n",
    "            gene_full_y1000_id_lookup = dict(zip(spec_lookup['y1000_id'], spec_lookup.index))\n",
    "        else: \n",
    "            gene_full_y1000_id_lookup = dict(zip(spec_lookup['y1000_id'], spec_lookup['gene_full']))\n",
    "\n",
    "        \n",
    "        \n",
    "        #for each orthogroup:\n",
    "        #extract protein seq if it comes from the right species.\n",
    "        \n",
    "        #gene_full: (y1000_id, og)\n",
    "        seqs_to_get = {}\n",
    "        gene_full_set = []\n",
    "        for og, (N_genes, pct_present, og_genes_out) in og_out_data.items():\n",
    "            for y1000_id in og_genes_out: \n",
    "                spec_id_check = int(y1000_id.split('_')[0])\n",
    "                if spec_id_check==spec_id:\n",
    "                    gene_full = gene_full_y1000_id_lookup[y1000_id]\n",
    "                    seqs_to_get[gene_full] = (y1000_id, og)\n",
    "                    gene_full_set.append(gene_full)\n",
    "        \n",
    "        gene_full_sets[spec_orig_genome] = gene_full_set\n",
    "        \n",
    "        for seq_record in seq_records:\n",
    "            #gene_full = 'augustus_masked-Deha2C-processed-gene-4.36'\n",
    "            if spec_orig_genome == 'saccharomyces_cerevisiae':\n",
    "                gene_full_check = seq_record.description.split()[0] #SC specific\n",
    "            elif spec_orig_genome == 'candida_albicans': \n",
    "                gene_full_check = seq_record.description\n",
    "            else: \n",
    "                gene_full_check = seq_record.description.split()[1].split('=')[1]\n",
    "            #print(gene_full)\n",
    "            if (gene_full_check in gene_full_set):\n",
    "                ##find which y1000_id was matched\n",
    "                #y1000_rlookup = genes_lookup['gene_full'] == gene_full\n",
    "        #         for gene, tf in y1000_rlookup.items(): \n",
    "        #             if tf:\n",
    "        #                 y1000_id=gene\n",
    "        #         gene_id = genes_lookup.loc[y1000_id, 'gene_id']\n",
    "                \n",
    "                (y1000_id, og) = seqs_to_get[gene_full_check]\n",
    "                protein_seq = seq_record.seq\n",
    "                f_out.write('>' + spec_orig_genome + '__' + og + '__' + y1000_id + ' gene_full=' + gene_full_check +\n",
    "                        ' source=' + source + '\\n')\n",
    "                f_out.write(str(protein_seq) + '\\n')  #I wonder why some of the bases were in lower case\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Find matches comparing shen and uniprot proteome. \n",
    "\n",
    "# #Only need to redo this if you are re-calculating pairwise_scores\n",
    "\n",
    "# shen_proteome_dir = os.path.normpath(y1000plus_dir + 'shen_2018_data/0_332yeast_genomes/332_genome_annotations/pep') + os.sep \n",
    "# uniprot_proteome_dir = os.path.normpath('C:/Users/heineib/Google Drive/Crick_LMS/external_data/genomes/diverse_yeast/proteomes') + os.sep \n",
    "\n",
    "# #(orig_genome ame: (shen_fname, uniprot_fname)): \n",
    "# proteome_fasta_dict = {'kluyveromyces_lactis': {'shen': shen_proteome_dir + 'kluyveromyces_lactis.max.pep', \n",
    "#                                                 'uniprot': uniprot_proteome_dir + 'Klac_uniprot-proteome_UP000000598.fasta'\n",
    "#                                                }, \n",
    "#                        'wickerhamomyces_anomalus': {'shen': shen_proteome_dir + 'wickerhamomyces_anomalus.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Wano_uniprot-proteome_UP000094112.fasta'\n",
    "#                            }, \n",
    "#                        'lachancea_thermotolerans': {'shen': shen_proteome_dir + 'lachancea_thermotolerans.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Lthe_uniprot-proteome_UP000002036.fasta'\n",
    "#                            }, \n",
    "#                        'komagataella_pastoris': {'shen': shen_proteome_dir + 'komagataella_pastoris.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Kpas_uniprot-proteome_UP000000314.fasta'\n",
    "#                            }, \n",
    "#                         'candida_tropicalis': {'shen': shen_proteome_dir + 'candida_tropicalis.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Ctro_uniprot-proteome_UP000002037.fasta'\n",
    "#                            },\n",
    "#                         'debaryomyces_hansenii': {'shen': shen_proteome_dir + 'debaryomyces_hansenii.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Dhan_uniprot-proteome_UP000000599.fasta'\n",
    "#                            }, \n",
    "#                         'yHMPu5000034957_hanseniaspora_osmophila_160519': {'shen': shen_proteome_dir + 'yHMPu5000034957_hanseniaspora_osmophila_160519.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Hosm_uniprot-proteome_UP000095728.fasta'\n",
    "#                            }, \n",
    "#                         'kluyveromyces_marxianus': {'shen': shen_proteome_dir + 'kluyveromyces_marxianus.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Kmar_uniprot-proteome_UP000065495.fasta'\n",
    "#                            }, \n",
    "#                         'zygosaccharomyces_rouxii': {'shen': shen_proteome_dir + 'zygosaccharomyces_rouxii.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Zrou_uniprot-proteome_UP000008536.fasta'\n",
    "#                            }\n",
    "#                       }\n",
    "\n",
    "# study1 = 'shen'\n",
    "# study2 = 'uniprot'\n",
    "\n",
    "# match_points = 1\n",
    "# mismatch_points = -1\n",
    "# gap_open = -.5\n",
    "# gap_extension = -0.1\n",
    "# NN_top_scores = 5\n",
    "# penalize_end_gaps = True \n",
    "\n",
    "\n",
    "# pairwise_scores = {}\n",
    "# spec_orig_genomes = list[proteome_fasta_dict.keys()]#[ 'yHMPu5000034957_hanseniaspora_osmophila_160519']  #'zygosaccharomyces_rouxii','kluyveromyces_marxianus','candida_tropicalis', 'debaryomyces_hansenii',\n",
    "\n",
    "# for spec_orig_genome in spec_orig_genomes:\n",
    "#     print(spec_orig_genome)\n",
    "#     #spec_orig_genome = 'komagataella_pastoris'# 'lachancea_thermotolerans' #'wickerhamomyces_anomalus' # 'kluyveromyces_lactis'\n",
    "\n",
    "#     use_subset = True\n",
    "#     subset = set(gene_full_sets[spec_orig_genome])\n",
    "#     pairwise_score_loaded = False #if you have already saved some and want to do the rest\n",
    "\n",
    "#     # #Check timing with time module\n",
    "#     # #python parallel programming library\n",
    "#     # #ipy_parallel dask\n",
    "#     # #https://ipyparallel.readthedocs.io/en/stable/development/messages.html#parallel-messages\n",
    "\n",
    "#     # #All by all pairwise sequence alignment, output top NN results\n",
    "\n",
    "#     genenames_study1 = []\n",
    "#     AA_seqs_study1 = SeqIO.parse(proteome_fasta_dict[spec_orig_genome][study1], \"fasta\")\n",
    "#     for seq_study1 in AA_seqs_study1: \n",
    "#         genename_study1 = dyt.gene_id_retrieve(study1, seq_study1)\n",
    "#     #     if spec=='Smik': \n",
    "#     #         genename_study1_dash = genename_study1\n",
    "#     #         genename_study1 = 'smik' + genename_study1_dash.split('smik_')[1]\n",
    "#         genenames_study1.append(genename_study1)\n",
    "\n",
    "#     if use_subset: \n",
    "#         print('Using Subset genes')\n",
    "#     else: \n",
    "#         subset = set(genenames_study1)\n",
    "\n",
    "#     if pairwise_score_loaded: \n",
    "#         genes_to_search = ((subset & set(genenames_study1)) - set(pairwise_score.keys()))\n",
    "#     else: \n",
    "#         pairwise_score = {}\n",
    "#         genes_to_search = (subset & set(genenames_study1))\n",
    "\n",
    "\n",
    "#     AA_seqs_study1 = SeqIO.parse(proteome_fasta_dict[spec_orig_genome][study1], \"fasta\")\n",
    "\n",
    "\n",
    "#     N = len(genes_to_search)\n",
    "#     for seq_study1 in AA_seqs_study1: \n",
    "#         check_gene = False\n",
    "#         genename_study1 = dyt.gene_id_retrieve(study1, seq_study1)\n",
    "#     #     if spec=='Smik': \n",
    "#     #         genename_study1_dash = genename_study1\n",
    "#     #         genename_study1 = 'smik' + genename_study1_dash.split('smik_')[1]\n",
    "#         if genename_study1 in genes_to_search:\n",
    "#             check_gene = True\n",
    "\n",
    "#         if check_gene: \n",
    "#             print(str(N) + ' ' + genename_study1)\n",
    "#             AA_seqs_study2 = SeqIO.parse(proteome_fasta_dict[spec_orig_genome][study2], \"fasta\")\n",
    "#             pairwise_score_comps = {}\n",
    "#             for seq_study2 in AA_seqs_study2: \n",
    "#                 genename_study2 = seq_study2.id\n",
    "#                 score = pairwise2.align.globalms(seq_study2,seq_study1, match_points, mismatch_points, gap_open, gap_extension, penalize_end_gaps = penalize_end_gaps, score_only = True )\n",
    "#                 pairwise_score_comps[genename_study2] = score\n",
    "#             pairwise_score_comps_df = pd.Series(pairwise_score_comps)\n",
    "#             pairwise_score[genename_study1] = pairwise_score_comps_df.sort_values(ascending=False).iloc[0:NN_top_scores]\n",
    "#             #print(pairwise_score[genename_YGOB])\n",
    "#             N = N - 1\n",
    "\n",
    "#     pairwise_scores[spec_orig_genome] =  pairwise_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Load pairwise scores\n",
    "\n",
    "pairwise_scores_fname = base_dir + os.sep + 'alphafold' + os.sep + 'pairwise_scores.pkl'\n",
    "\n",
    "# # Save pairwise_scores\n",
    "# with open(pairwise_scores_fname, 'wb') as f:\n",
    "#     pickle.dump(pairwise_scores, f)\n",
    "\n",
    "# Load pairwise_scores\n",
    "with open(pairwise_scores_fname, 'rb') as f: \n",
    "    pairwise_scores = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Show histogram of the pairwise scores and use it to choose thresholds: \n",
    "\n",
    "fig, axarr = plt.subplots(3,3, figsize=(15,15))\n",
    "\n",
    "#pairwise_scores = {}\n",
    "#for jj, spec in enumerate(['Ncas','Klac','Smik']):\n",
    "\n",
    "#pairwise_score_fname = data_processing_dir + os.path.normpath(\"ortholog_files_YGOB/AA_lists/\" + spec + \"_pairwise_score.pkl\")\n",
    "\n",
    "#with open(pairwise_score_fname,'rb') as file: \n",
    "#    pairwise_score= pickle.load(file)\n",
    "\n",
    "#pairwise_scores[spec] = pairwise_score\n",
    "#Use histogram to choose thresholds\n",
    "\n",
    "for jj, (spec_orig_genome, pairwise_score) in enumerate(pairwise_scores.items()):\n",
    "    ax = axarr.flatten()[jj]\n",
    "    maxvals = []\n",
    "    diff_top2 = []\n",
    "    for genename_shen, options in pairwise_score.items(): \n",
    "        options_sorted = options.sort_values()\n",
    "        #idxmax = options.idxmax()\n",
    "        maxval = options_sorted[-1]\n",
    "        diff_top2_val = maxval-options_sorted[-2]\n",
    "        diff_top2.append(diff_top2_val)\n",
    "        maxvals.append(maxval)\n",
    "        #if maxval<90: \n",
    "        #if (maxval>90) & (diff_top2_val<20):\n",
    "        #   print('{}, maxval: {}, difference: {}'.format(genename_shen,maxval,diff_top2_val))\n",
    "\n",
    "    \n",
    "    ax.scatter(maxvals, diff_top2)\n",
    "    ax.set_title(spec_orig_genome)\n",
    "    ax.set_xlabel('maximum similarity')\n",
    "    ax.set_ylabel('difference between top 2')\n",
    "    #ax.set_title(spec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if genome_name == 'saccharomyces_cerevisiae':\n",
    "#     #load gene_id map based on the species\n",
    "#     gene_lookup_spec_fname = y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/id_lookups/\" + genome_name + '.csv')\n",
    "#     gene_lookup_spec = pd.read_csv(gene_lookup_spec_fname, index_col='y1000_id')\n",
    "\n",
    "#     protein_fname = y1000plus_dir + os.path.normpath('shen_2018_data/0_332yeast_genomes/332_genome_annotations/Saccharomyces_cerevisiae_S288C_protein.fasta')\n",
    "\n",
    "#     seq_records = SeqIO.parse(protein_fname, \"fasta\")\n",
    "\n",
    "#     genes_lookup = gene_lookup_spec.loc[genes]\n",
    "#     genes_lookup_set = set(gene_lookup_spec.loc[genes,'gene_id'].values)  #SC specific\n",
    "\n",
    "#     for seq_record in seq_records:\n",
    "#         gene_id = seq_record.description.split()[0] #SC specific\n",
    "#         #print(gene_full)\n",
    "#         if (gene_id in genes_lookup_set):\n",
    "#             #find which y1000_id was matched\n",
    "#             y1000_rlookup = (genes_lookup['gene_id'] == gene_id)\n",
    "#             for gene, tf in y1000_rlookup.items(): \n",
    "#                 if tf:\n",
    "#                     y1000_id=gene\n",
    "#             protein_seq = seq_record.seq\n",
    "#             f.write('>' + genome_name + '_' + gene_id + ' y1000_id=' + y1000_id + ' gene_id=' + gene_id +\n",
    "#                     ' common_name=' + seq_record.description.split()[1] + #this adds in the cds from the original description\n",
    "#                     '\\n')\n",
    "#             f.write(str(protein_seq) + '\\n')  #I wonder why some of the bases were in lower case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get Orthogroup metadata\n",
    "#Cycle through each protein. \n",
    "\n",
    "#Record Metadata\n",
    "\n",
    "#OG Metadata: \n",
    "# SC Genenames, SP genenames, CA genenames\n",
    "# Total number\n",
    "# Percent of species covered\n",
    "#Possible future\n",
    "# Pathways\n",
    "# Enzyme numbers\n",
    "\n",
    "#{og: N_genes, pct_spec, og_genes, pathways, enzyme_numbers, sc_genenames, sc_names, sp_genename, ca_genename}\n",
    "\n",
    "scer_spec_id = 110\n",
    "\n",
    "#Build lookup table from S. pombe to S. cerevisiae\n",
    "sp_orth_map_fname = os.path.normpath('C:/Users/heineib/Google Drive/Crick_LMS/external_data/genomes/spom/cerevisiae-orthologs.txt')\n",
    "sp_orth_map = pd.read_table(sp_orth_map_fname, skiprows=1, header=None, names=['Spom','Scer'])\n",
    "scer_spom_lookup = {}\n",
    "none_list = []\n",
    "for row, (spom_gene, scer_genes) in sp_orth_map.iterrows():\n",
    "    if scer_genes=='NONE': \n",
    "        none_list.append(spom_gene)\n",
    "    else:\n",
    "        scer_genes_list = scer_genes.split('|')\n",
    "        for scer_gene in scer_genes_list: \n",
    "            scer_spom_lookup[scer_gene] = spom_gene\n",
    "\n",
    "#build C. albicans lookup table \n",
    "ca_orth_map_fname = os.path.normpath('C:/Users/heineib/Google Drive/Crick_LMS/external_data/genomes/calb/C_albicans_SC5314_S_cerevisiae_orthologs.txt')\n",
    "ca_orth_map = pd.read_table(ca_orth_map_fname, skiprows=8, header=None, names = ['cg_name', 'cgd_id', 'sc_genename', 'sc_name', 'sgd_id']) #, names=['Spom','Scer'])\n",
    "scer_calb_lookup = dict(zip(ca_orth_map['sc_genename'], ca_orth_map.index))\n",
    "\n",
    "            \n",
    "            \n",
    "og_out_data_full = {}\n",
    "jj = 0\n",
    "for (og, (N_genes, pct_spec, og_genes)) in og_out_data.items():\n",
    "    #print(og + ' ' + str(jj))\n",
    "    og_out_data_full_item = {}\n",
    "    og_out_data_full_item['N_genes'] = N_genes\n",
    "    og_out_data_full_item['pct_spec'] = pct_spec\n",
    "    og_out_data_full_item['og_genes'] = og_genes\n",
    "    if  pct_spec>pct_present_thresh:\n",
    "        og_out_data_full_item['pass_species_filter'] = True\n",
    "    else: \n",
    "        og_metadata_full_item['pass_species_filter'] = False\n",
    "        \n",
    "    #get_sc_genes\n",
    "    sc_genes = []\n",
    "    for og_gene in og_genes: \n",
    "        spec_id_check = int(og_gene.split('_')[0])\n",
    "        if spec_id_check==scer_spec_id:\n",
    "            sc_genes.append(scer_y1000_id_gene_id_lookup[og_gene])\n",
    "    \n",
    "    #get pathway and EC annotations\n",
    "    og_out_data_full_item['sc_genes'] = sc_genes\n",
    "    og_out_data_full_item['sc_names'] = yeast_esr_exp.SC_common_name_lookup(sc_genes)\n",
    "    og_out_data_full_item['pathways_id'] = list(sc_annotations_orf.loc[sc_genes]['Gene.pathways.identifier'])\n",
    "    og_out_data_full_item['pathways_name'] = list(sc_annotations_orf.loc[sc_genes]['Gene.pathways.name'])\n",
    "    og_out_data_full_item['EC_number'] = list(sc_annotations_orf.loc[sc_genes]['EC number'])\n",
    "    og_out_data_full_item['PDB_ID'] = list(sc_annotations_orf.loc[sc_genes]['PDB ID'])  #Need to convert NaN in these to 'NULL' to work with R's JSON import also see https://jsonlint.com/\n",
    "    \n",
    "    #get sp_genes\n",
    "    sp_genes = []\n",
    "    ca_genes = []\n",
    "    for sc_gene in sc_genes: \n",
    "        if sc_gene in scer_spom_lookup.keys(): \n",
    "            sp_genes.append(scer_spom_lookup[sc_gene])\n",
    "        else: \n",
    "            sp_genes.append('NONE')\n",
    "        \n",
    "        if sc_gene in scer_calb_lookup.keys(): \n",
    "            ca_genes.append(scer_calb_lookup[sc_gene])\n",
    "        else: \n",
    "            ca_genes.append('NONE')\n",
    "    \n",
    "    og_out_data_full_item['spom_genes'] = sp_genes\n",
    "    og_out_data_full_item['calb_genes'] = ca_genes\n",
    "    \n",
    "    \n",
    "    og_out_data_full[og] = og_out_data_full_item\n",
    "    jj = jj + 1\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save og_out_data_full (or metadata file)\n",
    "og_out_data_full_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_metadata.json'\n",
    "\n",
    "with open(og_out_data_full_fname, 'w') as f:\n",
    "    json.dump(og_out_data_full, f, indent=4 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load og_out_data_filt\n",
    "og_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_metadata.json'\n",
    "with open(og_fname, 'r') as f:\n",
    "    og_out_data_full = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Rebuild selected_proteins to fix source issue. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Build Fasta file\n",
    "\n",
    "fasta_out_fname = base_dir + os.sep + 'alphafold' + os.sep + 'selected_proteins.fasta'\n",
    "\n",
    "#the following information is in the fasta header for each protein\n",
    "#<spec_orig_genome>__<og>__<y1000_id>: you can use this to look up other metadata in the og_metadata.json file\n",
    "#source:  The source of the peptide sequence (either 'shen' or 'uniprot')\n",
    "#gene_full:  The full name of the gene from source peptide file\n",
    "#gene_full_shen: the full name of the gene from the shen peptide file\n",
    "#L: Length of the gene\n",
    "#sim_score_vs_shen: The similarity score between the shen peptide sequence and the uniprot peptide sequence.  If it is above the threshold, the uniprot sequence is used, below the threshold, we continue to use the shen sequence\n",
    "#sim_score_vs_shen_diff: The difference between similarity score of the top peptide and the second highest scoring peptide.  For genes with similarity score below a threshold, if this difference is too small, we keep the shen sequence\n",
    "\n",
    "shen_proteome_dir = os.path.normpath(y1000plus_dir + 'shen_2018_data/0_332yeast_genomes/332_genome_annotations/pep') + os.sep \n",
    "uniprot_proteome_dir = os.path.normpath('C:/Users/heineib/Google Drive/Crick_LMS/external_data/genomes/diverse_yeast/proteomes') + os.sep \n",
    "\n",
    "#(species genome name: (shen_fname, uniprot_fname)): \n",
    "proteome_fasta_dict = {'kluyveromyces_lactis': {'shen': shen_proteome_dir + 'kluyveromyces_lactis.max.pep', \n",
    "                                                'uniprot': uniprot_proteome_dir + 'Klac_uniprot-proteome_UP000000598.fasta'\n",
    "                                               }, \n",
    "                       'wickerhamomyces_anomalus': {'shen': shen_proteome_dir + 'wickerhamomyces_anomalus.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Wano_uniprot-proteome_UP000094112.fasta'\n",
    "                           }, \n",
    "                       'lachancea_thermotolerans': {'shen': shen_proteome_dir + 'lachancea_thermotolerans.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Lthe_uniprot-proteome_UP000002036.fasta'\n",
    "                           }, \n",
    "                       'komagataella_pastoris': {'shen': shen_proteome_dir + 'komagataella_pastoris.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Kpas_uniprot-proteome_UP000000314.fasta'\n",
    "                           }, \n",
    "                        'candida_tropicalis': {'shen': shen_proteome_dir + 'candida_tropicalis.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Ctro_uniprot-proteome_UP000002037.fasta'\n",
    "                           },\n",
    "                        'debaryomyces_hansenii': {'shen': shen_proteome_dir + 'debaryomyces_hansenii.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Dhan_uniprot-proteome_UP000000599.fasta'\n",
    "                           }, \n",
    "                        'yHMPu5000034957_hanseniaspora_osmophila_160519': {'shen': shen_proteome_dir + 'yHMPu5000034957_hanseniaspora_osmophila_160519.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Hosm_uniprot-proteome_UP000095728.fasta'\n",
    "                           }, \n",
    "                        'kluyveromyces_marxianus': {'shen': shen_proteome_dir + 'kluyveromyces_marxianus.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Kmar_uniprot-proteome_UP000065495.fasta'\n",
    "                           }, \n",
    "                        'zygosaccharomyces_rouxii': {'shen': shen_proteome_dir + 'zygosaccharomyces_rouxii.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Zrou_uniprot-proteome_UP000008536.fasta'\n",
    "                           }\n",
    "                      }\n",
    "\n",
    "\n",
    "sim_score_high_thresh = 75\n",
    "sim_score_low_thresh = 60\n",
    "sim_score_diff_thresh = 8\n",
    "\n",
    "\n",
    "with open(fasta_out_fname, 'w') as f_out: \n",
    "\n",
    "    #for each species: \n",
    "    for spec_id, spec_orig_genome in spec_id_map.items(): \n",
    "        \n",
    "        #spec_orig_genome = spec_orig_genome_lookup[spec_id]\n",
    "        print(spec_orig_genome)\n",
    "        \n",
    "        if spec_orig_genome in {'saccharomyces_cerevisiae', 'candida_albicans', 'schizosaccharomyces_pombe'}: \n",
    "            print('skipping ' + spec_orig_genome)\n",
    "            #Skip this species\n",
    "        elif spec_orig_genome in proteome_fasta_dict.keys():\n",
    "            #Attempt to convert to uniprot fasta  \n",
    "            print('checking to see if uniprot or shen et al sequences should be used')\n",
    "            source = 'uniprot'\n",
    "            \n",
    "            protein_fname_uniprot = proteome_fasta_dict[spec_orig_genome]['uniprot']\n",
    "            seq_records_uniprot = SeqIO.parse(protein_fname_uniprot, \"fasta\")\n",
    "            pairwise_scores_spec = pairwise_scores[spec_orig_genome]\n",
    "            \n",
    "            protein_fname_shen = proteome_fasta_dict[spec_orig_genome]['shen']\n",
    "            seq_records_shen = SeqIO.parse(protein_fname_shen, \"fasta\")\n",
    "        \n",
    "            spec_lookup_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/' + spec_orig_genome + '.csv')\n",
    "            spec_lookup = pd.read_csv(spec_lookup_fname, index_col=0)\n",
    "\n",
    "#             if spec_orig_genome == 'saccharomyces_cerevisiae':\n",
    "#                 #for S. cerevisiae the orf name is the 'gene_full'\n",
    "#                 gene_full_y1000_id_lookup = dict(zip(spec_lookup['y1000_id'], spec_lookup.index))\n",
    "#             else: \n",
    "            gene_full_y1000_id_lookup = dict(zip(spec_lookup['y1000_id'], spec_lookup['gene_full']))\n",
    "\n",
    "            #for each orthogroup:\n",
    "            #Identify whether the proteins will be uniprot or shen \n",
    "            seqs_to_get_uniprot = {} #gene_full_uniprot: (y1000_id, og, maxval, diff_top2_val, gene_full_shen)\n",
    "            seqs_to_get_shen = {}  #gene_full_shen: (y1000_id, og, maxval, diff_top2_val)\n",
    "            gene_full_set_uniprot = []\n",
    "            gene_full_set_shen = []\n",
    "            for og, (N_genes, pct_present, og_genes_out) in og_out_data.items():\n",
    "                for y1000_id in og_genes_out: \n",
    "                    spec_id_check = int(y1000_id.split('_')[0])\n",
    "                    if spec_id_check==spec_id:\n",
    "                        gene_full_shen = gene_full_y1000_id_lookup[y1000_id]\n",
    "                        pairwise_score_gene = pairwise_scores_spec[gene_full_shen]\n",
    "                        \n",
    "                        #Check to see if pairwise scores pass filters: \n",
    "                        options_sorted = pairwise_score_gene.sort_values(ascending=False)  #possible that they are already sorted and this is unnecessary\n",
    "                        maxval = options_sorted[0]\n",
    "                        diff_top2_val = maxval-options_sorted[1]\n",
    "                        \n",
    "                        if ((maxval>sim_score_high_thresh) | ((maxval>sim_score_low_thresh) & (diff_top2_val>sim_score_diff_thresh))): \n",
    "                            gene_full_uniprot = options_sorted.index[0]\n",
    "                            seqs_to_get_uniprot[gene_full_uniprot] = (y1000_id, og, maxval, diff_top2_val, gene_full_shen)\n",
    "                            gene_full_set_uniprot.append(gene_full_uniprot)\n",
    "                        else: \n",
    "                            seqs_to_get_shen[gene_full_shen] = (y1000_id, og, maxval, diff_top2_val)\n",
    "                            gene_full_set_shen.append(gene_full_shen)\n",
    "                                      \n",
    "            #Cycle through uniprot proteome and print to fasta if it is in the set. \n",
    "            dyt.seq_record_fasta_printout(seq_records_uniprot, f_out, gene_full_set_uniprot, seqs_to_get_uniprot, 'uniprot', spec_orig_genome)\n",
    "                 \n",
    "            #Cycle through shen proteome and print to fasta if it is in the set. \n",
    "            dyt.seq_record_fasta_printout(seq_records_shen, f_out, gene_full_set_shen, seqs_to_get_shen, 'shen', spec_orig_genome)\n",
    "\n",
    "        \n",
    "        \n",
    "        else: \n",
    "            print('using shen et al sequence')\n",
    "            dyt.fasta_extract_shen(f_out,shen_proteome_dir, spec_orig_genome, y1000plus_dir, og_out_data, spec_id)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Control peptide files\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Control peptides are \n",
    "#CDC19/PYK2\n",
    "#ERG11\n",
    "\n",
    "#These are already made in 20210826_met_orths\n",
    "#MET17\n",
    "#STE2/HSU1\n",
    "#IRC7\n",
    "#STR3\n",
    "#YH112C\n",
    "#CYS3\n",
    "\n",
    "y1000_species_subset = y1000_species[y1000_species['Major clade']!='outgroup']\n",
    "\n",
    "# y1000_plus_dir_ubuntu = '/home/heineike/GitHub/y1000plus_tools/data/'\n",
    "# metproj_data_dir = os.path.normpath('C:/Users/heineib/Google Drive/Crick_LMS/projects/jason_met_auxotrophy') + os.sep\n",
    "\n",
    "# aln_fname_base_win = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/proteins_og') + os.sep\n",
    "\n",
    "#Select only clades of interest: \n",
    "#clades_of_interest = {'Saccharomycetaceae','Saccharomycodaceae','Phaffomycetaceae','CUG-Ser2','CUG-Ser1','Pichiaceae','CUG-Ala'}\n",
    "#clades_of_interest = {'Saccharomycetaceae', 'Saccharomycodaceae'}\n",
    "#outgroups = [ 'Wickerhamomyces anomalus','Cyberlindnera jadinii', 'Ascoidea rubescens','Komagataella pastoris','Debaryomyces hansenii', 'Candida albicans', 'Yarrowia lipolytica']\n",
    "#outgroup = 'Yarrowia lipolytica'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#Met17: 'YLR303W', \n",
    "#Str2:  YJR130C\n",
    "#'YML082W', \n",
    "#Str3: 'YGL184C', \n",
    "#'YLL058W', \n",
    "#Cys3: 'YAL012W'\n",
    "#IRC7: 'YFR055W'\n",
    "\n",
    "\n",
    "#target_y1000_ids = scer_lookup.loc[['YLR303W', 'YML082W','YJR130C', 'YFR055W', 'YHR112C', 'YGL184C', 'YLL058W', 'YAL012W'], 'y1000_id']\n",
    "gois = ['CDC19', 'PYK2', 'ERG11', 'STR2']\n",
    "target_y1000_ids = scer_lookup.loc[yeast_esr_exp.SC_orf_lookup_by_name(gois), 'y1000_id']\n",
    "\n",
    "gois_y1000_id_lookup = dict(zip(gois, target_y1000_ids))\n",
    "\n",
    "\n",
    "y1000_id_list = list(target_y1000_ids)\n",
    "goi_og_lookup, og_genes_lookup = y1000plus_tools.make_og_genes_lookup(y1000_id_list, y1000_species_subset)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gois_y1000_id_lookup['CDC19']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gois_y1000_id_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goi_og_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_y1000_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_mapping = {'CDC19_PYK2': goi_og_lookup[gois_y1000_id_lookup['CDC19']],\n",
    "              'ERG11': goi_og_lookup[gois_y1000_id_lookup['ERG11']], \n",
    "              'STR2_HSU1':  goi_og_lookup[gois_y1000_id_lookup['STR2']]\n",
    "             }\n",
    "og_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Only need to do once\n",
    "\n",
    "for protein_fname, og in og_mapping.items():\n",
    "    og_genes = og_genes_lookup[og]\n",
    "    y1000plus_tools.extract_protein_seqs(og_genes, protein_fname, y1000_species_subset)\n",
    "    print(protein_fname + ' protein sequence extraction complete')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make control JSON files to pass to jacob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load og_out_data_full\n",
    "og_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_metadata.json'\n",
    "with open(og_fname, 'r') as f:\n",
    "    og_out_data_full = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_out_data_control = {}\n",
    "\n",
    "for protein_fname, og in og_mapping.items():\n",
    "    og_out_data_og = og_out_data_full[og]\n",
    "    og_genes = og_genes_lookup[og]\n",
    "    og_out_data_og['og_genes']=list(og_genes)\n",
    "    og_out_data_og['N_genes'] = len(og_genes)\n",
    "    \n",
    "    \n",
    "    del og_out_data_og['pct_spec']\n",
    "    del og_out_data_og['pass_species_filter']\n",
    "    og_out_data_control[og] = og_out_data_og\n",
    "    \n",
    "# Save og_out_data\n",
    "og_out_data_control_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_genes_control.json'\n",
    "\n",
    "with open(og_out_data_control_fname, 'w') as f:\n",
    "    json.dump(og_out_data_control, f, sort_keys=True, indent=4 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Possible fields include: \n",
    "\n",
    "N_genes: Number of genes in orthogroup\n",
    "og_genes: List of identifiers of genes in orthogroup from [1]\n",
    "sc_genes: Systematic name of S. cerevisiae genes present in orthogroup\n",
    "sc_names: Common Name of S. cerevisiae genes present in orthogroup\n",
    "pathways_id: Id number of pathways for each gene from S. cerevisiae\n",
    "pathways_name: Name of pathway for each gene from S. cerevisiae\n",
    "EC_number: EC number (if any) for each gene from S. cerevisiae\n",
    "PDB_ID: PDB ID of structure for each gene from S. cerevisaiae\n",
    "spom_genes: systematic name of S. pombe ortholog of the S. cerevisiae genes in the orthogroup. Based on Pombase file.\n",
    "calb_genes: systematic name of C. albicans ortholog of the S. cerevisiae genes in the orthogroup. Based on CGD file.\n",
    "\n",
    "Removed: \n",
    "pass_species_filter: true or false depending on whether pct_spec is above the threshold (currently 0.4)\n",
    "pct_spec: percentage of species from our species selection that are represented in the orthogroup\n",
    "\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load 25 species og mapping\n",
    "\n",
    "#Read in orthogroup listing from orthomcl.txt\n",
    "\n",
    "#Count og_genes and add in N_genes\n",
    "\n",
    "#Reset N_genes\n",
    "#Remove pct_spec, pass_species_filter\n",
    "\n",
    "#Add in new og_genes list.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_mapping = {'CDC19_PYK2': goi_og_lookup[gois_y1000_id_lookup['CDC19']],\n",
    "              'ERG11': goi_og_lookup[gois_y1000_id_lookup['ERG11']], \n",
    "              'STR2_HSU1':  goi_og_lookup[gois_y1000_id_lookup['STR2']]\n",
    "             }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 03 Feb 2022 Sandra noticed duplicate IDs in the protein file\n",
    "# I found the bug (was naming proteins with gene_full instead of gene_full_check), remade the files and sent her the corrected ones.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 03 Feb 2022 Sandra noticed duplicate IDs in the protein file\n",
    "# I found the bug (was naming proteins with gene_full instead of gene_full_check), remade the files and sent her the corrected ones.  \n",
    "\n",
    "selected_fasta = base_dir + os.sep + 'alphafold' + os.sep + 'selected_proteins.fasta'\n",
    "all_fasta = base_dir + os.sep + 'alphafold' + os.sep + 'all_proteins.fasta'\n",
    "cont_fasta = base_dir + os.sep + 'alphafold' + os.sep + 'STR2_YLL058W.fasta'\n",
    "\n",
    "seq_records_sel = SeqIO.parse(selected_fasta, \"fasta\")\n",
    "seq_records_all = SeqIO.parse(all_fasta, \"fasta\")\n",
    "seq_records_cont = SeqIO.parse(cont_fasta, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = []\n",
    "\n",
    "for record in seq_records_cont: \n",
    "    names.append(record.name)\n",
    "\n",
    "len(set(names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Counter(names).most_common()\n",
    "\n",
    "#Previously it found the following species had many duplicate ids.  The other species were fine.  The error effected only species that only had shen data: \n",
    "\n",
    "# ('alloascoidea_hylecoeti__OG5722__0_7605', 552),\n",
    "#  ('geotrichum_candidum__OG5722__43_1771', 510),\n",
    "#  ('vanderwaltozyma_polyspora__OG7278__136_23', 477),\n",
    "#  ('ascoidea_rubescens__OG5735__6_750', 461),\n",
    "#  ('kazachstania_naganishii__OG8784__49_4009', 460),\n",
    "#  ('cyberlindnera_jadinii__OG6065__35_2226', 454),\n",
    "#  ('yHMPu5000034604_sporopachydermia_lactativora_160519__OG5404__170_5570',\n",
    "#   450),\n",
    "#  ('yarrowia_lipolytica__OG5722__144_3506', 447),\n",
    "#  ('pachysolen_tannophilus__OG5621__106_2089', 445),\n",
    "#  ('torulaspora_delbrueckii__OG7278__135_3579', 430),\n",
    "#  ('lipomyces_starkeyi__OG5722__66_8498', 426),\n",
    "#  ('eremothecium_gossypii__OG7818__40_277', 401),\n",
    "#  ('ogataea_parapolymorpha__OG5096__104_1849', 397),\n",
    "#  ('candida_apicola__OG5441__13_3008', 393),\n",
    "#  ('tortispora_caseinolytica__OG5722__134_2655', 380),"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate Pathway list to send to Jacob Steenwyk 20220419"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_annotations_df = pd.read_csv(base_dir + os.sep + 'eggnog_mapper' + os.sep +  'annotation' + os.sep + 'yeastmine_annotation_summary_20211007.csv', index_col = 0)\n",
    "\n",
    "scer_orf_table = pd.read_table(base_dir + os.sep + 'eggnog_mapper' + os.sep +  'annotation' + os.sep + 'yeastmine_orf_list_20211025.tsv', index_col=1, names= ['SGD_id', 'species', 'Name', 'full_name', 'qualifier', 'status'])\n",
    "\n",
    "scer_orf_table_not_dub = scer_orf_table[scer_orf_table['qualifier'] != 'Dubious']\n",
    "\n",
    "sc_annotations_orf = sc_annotations_df.loc[list(set(sc_annotations_df.index) & set(scer_orf_table_not_dub.index ))]\n",
    "\n",
    "sc_annotations_orf['N_pathways'] = [len(ast.literal_eval(pway_list)) for pway_list in sc_annotations_orf['Gene.pathways.name']]\n",
    "\n",
    "sc_annotations_pway_subset = sc_annotations_orf[sc_annotations_orf['N_pathways']>0]\n",
    "\n",
    "og_metadata_filtered = {}\n",
    "\n",
    "for og, metadata in og_metadata.items():\n",
    "    if metadata['pass_species_filter']: \n",
    "        og_metadata_filtered[og] = metadata\n",
    "        \n",
    "print('Total ogs in pathways: {}, Ogs that meet species percentage threshold: {}'.format((len(og_metadata.keys())), len(og_metadata_filtered.keys())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make list of all possible pathways\n",
    "pway_full_list= []\n",
    "\n",
    "for pway_list in sc_annotations_pway_subset['Gene.pathways.name']:\n",
    "    pway_list = ast.literal_eval(pway_list)\n",
    "    for pway in pway_list: \n",
    "        pway_full_list.append(pway)\n",
    "\n",
    "print(len(pway_full_list))\n",
    "\n",
    "pway_unique_list = list(set(pway_full_list))\n",
    "\n",
    "print(len(pway_unique_list))\n",
    "    \n",
    "# Make dict of all pathways with list of their s_cer genes\n",
    "\n",
    "pway_genes = {pway: [] for pway in pway_unique_list}\n",
    "\n",
    "for gene, pway_list in sc_annotations_pway_subset['Gene.pathways.name'].iteritems():\n",
    "    #print(gene)\n",
    "    #print(pway_list)\n",
    "    pway_list = ast.literal_eval(pway_list)\n",
    "    for pway in pway_list: \n",
    "        #print(pway)\n",
    "#         if pway == 'PWY30-230': \n",
    "#             print(gene)\n",
    "#             print(pway_genes_out)\n",
    "        pway_genes_out = pway_genes[pway]\n",
    "        pway_genes_out.append(gene)\n",
    "\n",
    "        pway_genes[pway] = pway_genes_out\n",
    "\n",
    "\n",
    "        \n",
    "# Make data frame from that dict\n",
    "pway_genes_df = pd.DataFrame(pd.Series(pway_genes))\n",
    "\n",
    "pway_genes_df.rename(columns = {0:'sc_genes'}, inplace=True)\n",
    "\n",
    "# Add a column which is the gene name\n",
    "\n",
    "# Add a column which is a list of whether each gene is present or not in our list\n",
    "og_metadata_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_metadata.json'\n",
    "\n",
    "\n",
    "with open(og_metadata_fname, 'r') as f:\n",
    "    og_metadata = json.load(f) \n",
    "    \n",
    "all_present_genes = []\n",
    "pct_present_thresh = 0.4\n",
    "\n",
    "for og in og_metadata.keys():\n",
    "    if og_metadata[og]['pct_spec']>pct_present_thresh:\n",
    "        all_present_genes = all_present_genes + og_metadata[og]['sc_genes']\n",
    "\n",
    "all_present_genes_unique = set(all_present_genes)\n",
    "print('Number of genes present in orthogroups: {} \\nNumber of unique genes present in ortogroups: {}'.format(len(all_present_genes), len(all_present_genes_unique)))\n",
    "\n",
    "\n",
    "sc_genes_name = []\n",
    "sc_genes_presence = []\n",
    "for sc_genes in pway_genes_df['sc_genes']: \n",
    "    sc_names = yeast_esr_exp.SC_common_name_lookup(sc_genes)\n",
    "    sc_genes_name.append(sc_names)\n",
    "    sc_presence = [sc_gene in all_present_genes_unique for sc_gene in sc_genes]\n",
    "    sc_genes_presence.append(sc_presence)\n",
    "\n",
    "pway_genes_df['sc_genes_name'] = sc_genes_name\n",
    "pway_genes_df['sc_genes_presence'] = sc_genes_presence\n",
    "pway_genes_df.index.name = 'pathway'\n",
    "    \n",
    "#pway_genes_df.to_csv(base_dir + os.sep + 'alphafold' + os.sep + 'pathway_list.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtain fasta for Feiran of our S.cer, C.alb, and S.pom proteins of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Need to get IDs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spom: No official name, systematic_id or synonym for matmc_2 y1000_id=121_134\n",
      "Spom: No official name, systematic_id or synonym for SPCC622.05 y1000_id=121_2159\n",
      "Spom: No official name, systematic_id or synonym for SPAC4H3.12c y1000_id=121_2282\n",
      "Spom: No official name, systematic_id or synonym for meu1-1-1 y1000_id=121_2287\n",
      "Spom: No official name, systematic_id or synonym for sab14 y1000_id=121_2434\n",
      "Spom: No official name, systematic_id or synonym for SPBC36.13 y1000_id=121_2778\n",
      "Spom: No official name, systematic_id or synonym for SPAC1F8.09c.1 y1000_id=121_3125\n",
      "Spom: No official name, systematic_id or synonym for SPBC32F12.17 y1000_id=121_3368\n",
      "Spom: No official name, systematic_id or synonym for SPBC1685.12c y1000_id=121_3533\n",
      "Spom: No official name, systematic_id or synonym for SPCC417.04 y1000_id=121_3584\n",
      "Spom: No official name, systematic_id or synonym for SPMTR.04 y1000_id=121_3769\n",
      "Spom: No official name, systematic_id or synonym for SPMTR.03 y1000_id=121_3787\n",
      "Spom: No official name, systematic_id or synonym for SPCC188.05 y1000_id=121_3839\n",
      "Spom: No official name, systematic_id or synonym for chr3 y1000_id=121_3849\n",
      "Spom: No official name, systematic_id or synonym for ppr1.2 y1000_id=121_3929\n",
      "Spom: No official name, systematic_id or synonym for Tf2-10-pseudo y1000_id=121_4201\n",
      "Spom: No official name, systematic_id or synonym for SPAC27E2.13 y1000_id=121_4235\n",
      "Spom: No official name, systematic_id or synonym for matmc_1 y1000_id=121_426\n",
      "Spom: No official name, systematic_id or synonym for rgg8 y1000_id=121_4275\n",
      "Spom: No official name, systematic_id or synonym for ppr1.1 y1000_id=121_4515\n",
      "Spom: No official name, systematic_id or synonym for Tf2-13-pseudo y1000_id=121_4528\n",
      "Spom: No official name, systematic_id or synonym for mat1-m y1000_id=121_4547\n",
      "Spom: No official name, systematic_id or synonym for meu1-1-2 y1000_id=121_4559\n",
      "Spom: No official name, systematic_id or synonym for SPBC354.11c y1000_id=121_4585\n",
      "Spom: No official name, systematic_id or synonym for mat3-m y1000_id=121_4774\n",
      "Spom: No official name, systematic_id or synonym for sab10 y1000_id=121_4780\n",
      "Spom: No official name, systematic_id or synonym for chr1 y1000_id=121_4846\n",
      "Spom: No official name, systematic_id or synonym for SPCC1672.14 y1000_id=121_4989\n",
      "Spom: No official name, systematic_id or synonym for rpl35b y1000_id=121_521\n",
      "Spom: No official name, systematic_id or synonym for SPCC794.16 y1000_id=121_547\n",
      "Spom: No official name, systematic_id or synonym for SPAC1D4.07c y1000_id=121_640\n",
      "Spom: No official name, systematic_id or synonym for B22918-2 y1000_id=121_710\n",
      "gene_full not mapping to name that maps to systematic id SPAC110.05 systematic_id=SPAC110.06\n",
      "Spom: No official name, systematic_id or synonym for chr2 y1000_id=121_891\n",
      "Spom: No official name, systematic_id or synonym for tif35 y1000_id=121_937\n",
      "gene_full not mapping to name that maps to systematic id SPBC713.13 systematic_id=SPBC713.14c\n",
      "Scer\n",
      "Spom\n",
      "no y1000_id found for P87214 in Spom\n",
      "Calb\n"
     ]
    }
   ],
   "source": [
    "#Extract Protein Sequences directly from Oliver's PDB Files\n",
    "\n",
    "#(metadata_field,ref_status, species)\n",
    "model_fields = {'Scer': ( 'REF_','saccharomyces_cerevisiae'),\n",
    "                'Spom': ('','schizosaccharomyces_pombe'),\n",
    "                'Calb': ('','candida_albicans')\n",
    "               }\n",
    "          \n",
    "pdb_base = base_dir + os.sep + os.path.normpath('alphafold/selected_proteins/pdbs') + os.sep\n",
    "output_file = base_dir + os.sep + os.path.normpath('alphafold/selected_proteins/selected_proteins_modelorgs.fasta')\n",
    "\n",
    "swissprot_id_2_gene_id = dyt.load_model_swissprot_id_2_gene_id()\n",
    "gene_id_2_y1000_id = dyt.load_model_gene_id_2_y1000_id()\n",
    "model_og_lookup = dyt.load_model_og_lookup()\n",
    "\n",
    "#Make dictionary to lookup og from geneID in model species. \n",
    "\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")  #There was a warning about headers not existing in the PDBs\n",
    "    with open(output_file, 'w') as f_out:\n",
    "        for spec_abbrev, (ref_status, spec) in model_fields.items():\n",
    "            print(spec_abbrev)\n",
    "            swissprot_id_2_gene_id_spec = swissprot_id_2_gene_id[spec_abbrev]\n",
    "            gene_id_2_y1000_id_spec = gene_id_2_y1000_id[spec_abbrev]\n",
    "            model_og_lookup_spec = model_og_lookup[spec_abbrev]\n",
    "\n",
    "            pdb_base_spec = pdb_base + 'Alphafold_' + spec_abbrev + '_sel' + os.sep\n",
    "            for fname in os.listdir(pdb_base_spec):\n",
    "                swissprot_id = fname.split('-')[1]\n",
    "                if swissprot_id in swissprot_id_2_gene_id_spec.keys():\n",
    "                    gene_id = swissprot_id_2_gene_id_spec[swissprot_id]\n",
    "                    if gene_id in gene_id_2_y1000_id_spec.keys():\n",
    "                        y1000_id = gene_id_2_y1000_id_spec[gene_id]\n",
    "                    else: \n",
    "                        print('no y1000_id found for ' + swissprot_id + ' in ' + spec_abbrev)\n",
    "                        y1000_id = 'NONE'\n",
    "\n",
    "                    assert gene_id in model_og_lookup_spec.keys(), gene_id + ' not in og lookup'\n",
    "\n",
    "                    og = model_og_lookup_spec[gene_id]\n",
    "\n",
    "\n",
    "                else: \n",
    "                    print('no gene_id for ' + swissprot_id + ' in ' + spec_abbrev)\n",
    "                    gene_id = 'NONE'\n",
    "                    y1000_id = 'NONE'\n",
    "\n",
    "                header = '>' + gene_id + ' og=' + og + ' species=' + spec + ' uniprot_id=' + swissprot_id + ' af2_id=' + ref_status + spec_abbrev + '_' + fname.split('.')[0] + ' y1000_id=' + y1000_id + '\\n'\n",
    "                f_out.write(header)\n",
    "                pdb_file = SeqIO.parse(pdb_base_spec + fname, \"pdb-atom\")\n",
    "                for (jj,record) in enumerate(pdb_file):\n",
    "                    if jj==0: \n",
    "                        f_out.write(str(record.seq) + '\\n')    \n",
    "                    elif jj==1: \n",
    "                        print('Why is there more than one record? ' + fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pombe gene P87214 is HEM4 (SPAC31G4.08) in og 3856, but in the og list, there are no pombe genes (121_XXXX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## extraction of model species protein sequences and Ids (originally to give a file to Feiran of our S.cer species)\n",
    "\n",
    "# #(metadata_field, ref_status, y1000_id_present)\n",
    "# model_fields = {'Scer': ('sc_genes', 'REF_','Saccharomyces_cerevisiae'),\n",
    "#                 'Spom': ('spom_genes','','Schizosaccharomyces_pombe'),\n",
    "#                 'Calb': ('calb_genes','','Candida_albicans')\n",
    "#                }\n",
    "\n",
    "# gene_id_2_y1000_id = dyt.load_model_gene_id_2_y1000_id()\n",
    "# #swissprot_id_2_gene_id = dyt.load_model_swissprot_id_2_gene_id()\n",
    "# gene_id_2_swissprot_id = dyt.load_model_gene_id_2_swissprot_id()\n",
    "\n",
    "# ### Note: I sent Feiran Pombase sequences, but some of those are different from the Uniprot sequences used by Alphafold.  e.g. OG2645\n",
    "\n",
    "\n",
    "# model_protein_dict = {}\n",
    "# for spec_abbrev in ['Scer', 'Spom', 'Calb']: \n",
    "#     model_protein_dict[spec_abbrev] = dyt.load_model_protein_dict(spec_abbrev)\n",
    "    \n",
    "# #Load og_metadata\n",
    "# #note:  This is called og_out_data_full earlier in the script - should converge notation\n",
    "# og_metadata_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_metadata.json'\n",
    "\n",
    "# with open(og_metadata_fname, 'r') as f:\n",
    "#     og_metadata = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "\n",
    "# #For each OG, extract s.cer, s.pom, and c.alb genes\n",
    "# #Build a header which uses the same naming convention as we get from alphafold\n",
    "# # <primary id> swissprot_id= af2_id=<REF_><spec_abbrev>_AF-<swissprot_id>-F1-model_v2 y1000_id= \n",
    "\n",
    "# #(metadata_field, ref_status, y1000_id_present)\n",
    "# model_fields = {'Scer': ('sc_genes', 'REF_','Saccharomyces_cerevisiae'),\n",
    "#                 'Spom': ('spom_genes','','Schizosaccharomyces_pombe'),\n",
    "#                 'Calb': ('calb_genes','','Candida_albicans')\n",
    "#                }\n",
    "\n",
    "# with open(base_dir + os.sep + 'alphafold' + os.sep + 'selected_proteins_modelorgs.fasta', 'w') as f_out: \n",
    "#     for og, metadata in og_metadata.items(): \n",
    "#         for spec_abbrev, (metadata_field,ref_status, spec) in model_fields.items():\n",
    "       \n",
    "#             model_genes = metadata[metadata_field]\n",
    "#             model_protein_dict_spec = model_protein_dict[spec_abbrev]\n",
    "\n",
    "#             for gene_id in model_genes:\n",
    "#                 if gene_id!='NONE':\n",
    "                    \n",
    "#                     if gene_id in model_protein_dict_spec.keys():\n",
    "#                         if gene_id in gene_id_2_y1000_id[spec_abbrev].keys():\n",
    "#                             y1000_id = gene_id_2_y1000_id[spec_abbrev][gene_id] \n",
    "#                         else: \n",
    "#                             print('No y1000_id for ' + spec_abbrev + ' ' + gene_id)\n",
    "#                             y1000_id = 'None'\n",
    "#                         y1000_id_header = ' y1000_id=' + y1000_id\n",
    "#                         swissprot_id = gene_id_2_swissprot_id[spec_abbrev][gene_id]\n",
    "#                         header = '>' + gene_id + ' og=' + og + ' species=' + spec + ' uniprot_id=' + swissprot_id + ' af2_id=' + ref_status + spec_abbrev + '_AF-' + swissprot_id + '-F1-model_v2' + y1000_id_header + '\\n'\n",
    "\n",
    "#                         f_out.write(header)\n",
    "\n",
    "#                         f_out.write(model_protein_dict_spec[gene_id]+ '\\n')\n",
    "#                     else: \n",
    "#                         print('OG:' + og +  ' spec:'+ spec_abbrev +  ' gene:' + gene_id + ' not present in sequence data')\n",
    "\n",
    "\n",
    "#     #Build a fasta with all the files in "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract proteome and codon sequences for various species\n",
    "Shen et al already has this\n",
    "Uniprot genomes either use NCBI (linked from Uniprot page) or NCBI_genome page.  \n",
    "note:  Might be able to get all the files from the NCBI_genome page\n",
    "For S. pom, S. cer and C. alb use pombase, SGD, and CGD databases respectively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Uniprot genomes with contigs linked from Uniprot page\n",
    "\n",
    "# #Make folders for ncbi_extract.keys()\n",
    "\n",
    "# for spec in ncbi_extract.keys(): \n",
    "#     os.mkdir(genomes_base + os.sep + 'ncbi_coding_seq' + os.sep + spec)\n",
    "#     os.mkdir(genomes_base + os.sep + 'ncbi_protein_seq' + os.sep + spec)\n",
    "\n",
    "# Made json file listing all contigs for each organism from looking at the list on uniprot.  \n",
    "\n",
    "#Load ncbi chromosome list\n",
    "ncbi_extract_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ncbi_extract.json')\n",
    "with open(ncbi_extract_fname, 'r') as f:\n",
    "    ncbi_extract = json.load(f) \n",
    "\n",
    "for spec, spec_info in ncbi_extract.items(): \n",
    "    contig_file_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ncbi_coding_seq/' + spec + '/contig_list.txt')\n",
    "    with open(contig_file_fname,'w') as f_out:\n",
    "        for contig in spec_info['contigs']:\n",
    "            f_out.write(contig + '\\n')\n",
    "\n",
    "        \n",
    "#For each file used NCBI batch Entrez \n",
    "#https://www.ncbi.nlm.nih.gov/sites/batchentrez\n",
    "\n",
    "#Choose Nucleotide Database, \n",
    "#Upload contig_list.txt for the species of interest and hit retrieve\n",
    "#Check names of contigs\n",
    "#Select all contigs\n",
    "#Hit send to, and choose 'coding sequences', and download as fasta nucleotide and fasta protein. \n",
    "#Save in the appropriate folders as proteome.fasta and cds_raw.fasta\n",
    "#Proteomes: \n",
    "#Crick_LMS\\external_data\\genomes\\diverse_yeast\\ncbi_protein_seq\\<species>\\proteome.fasta\n",
    "#CDS: \n",
    "#Crick_LMS\\external_data\\genomes\\diverse_yeast\\ncbi_coding_seq\\<species>\\cds_raw.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make cds files with uniprot names as keys\n",
    "\n",
    "uniprot_ncbi_specs = [\n",
    "                 'zygosaccharomyces_rouxii',\n",
    "                 'debaryomyces_hansenii',\n",
    "                 'lachancea_thermotolerans',\n",
    "                 'kluyveromyces_marxianus',\n",
    "                 'kluyveromyces_lactis',\n",
    "                 'komagataella_pastoris',\n",
    "                 'schizosaccharomyces_pombe'\n",
    "                     ]\n",
    "\n",
    "km_id_map = pd.read_table(genomes_base + os.sep + os.path.normpath('diverse_yeast/proteomes/Kmar_uniprot_id_table_20221020.tsv'))\n",
    "\n",
    "#several k.mar proteins have more than one gene name for a single uniprot_id\n",
    "km_orf_2_uniprot = {}\n",
    "for orf,uniprot_id in zip(km_id_map['Gene Names (ORF)'], km_id_map['Entry']):\n",
    "    if len(orf.split(' ')) >1:\n",
    "        for orf_id in orf.split(' '):\n",
    "            km_orf_2_uniprot[orf_id] = uniprot_id\n",
    "    else: \n",
    "        km_orf_2_uniprot[orf] = uniprot_id\n",
    "\n",
    "for spec in uniprot_ncbi_specs:\n",
    "    cds_raw_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ncbi_coding_seq/' + spec + '/' + 'cds_raw.fasta')\n",
    "    cds_out_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ncbi_coding_seq/' + spec + '/' + 'cds.fasta')\n",
    "    \n",
    "    cds_raw = SeqIO.parse(cds_raw_fname, 'fasta')\n",
    "    \n",
    "    with open(cds_out_fname, 'w') as f_cds_out:\n",
    "    \n",
    "        for record in cds_raw: \n",
    "            record_metadata = {}\n",
    "            metadata_list = record.description.split('] [')\n",
    "            metadata_list[0] = metadata_list[0].split('[')[1]\n",
    "            metadata_list[-1] = metadata_list[-1].strip(']')\n",
    "                    \n",
    "            \n",
    "            for metadata_set in metadata_list:\n",
    "                (metadata_label, metadata) = metadata_set.split('=')\n",
    "                record_metadata[metadata_label] = metadata\n",
    "            \n",
    "            \n",
    "            if spec == 'kluyveromyces_marxianus':\n",
    "                \n",
    "                assert 'locus_tag' in record_metadata.keys(), 'in ' + spec + ' no locus tag for ' + record.description \n",
    "                secondary_id = record_metadata['locus_tag']\n",
    "                secondary_id_name = 'locus_tag'\n",
    "                                \n",
    "                primary_id = km_orf_2_uniprot[secondary_id]\n",
    "                \n",
    "            \n",
    "            else: \n",
    "            \n",
    "            \n",
    "\n",
    "                if 'locus_tag' in record_metadata.keys(): \n",
    "                    secondary_id = record_metadata['locus_tag']\n",
    "                    secondary_id_name = 'locus_tag'\n",
    "                elif 'gene' in record_metadata.keys(): \n",
    "                    secondary_id = record_metadata['gene']\n",
    "                    secondary_id_name = 'gene'\n",
    "                elif 'protein_id' in record_metadata.keys(): \n",
    "                    secondary_id = record_metadata['protein_id']\n",
    "                    secondary_id_name = 'protein_id'\n",
    "                else:\n",
    "                    raise ValueError(record.description)\n",
    "\n",
    "                if 'db_xref' in record_metadata.keys(): \n",
    "                    db_xref_dict = {}\n",
    "                    for db_xref_item in record_metadata['db_xref'].split(','):\n",
    "                        (xref_db, xref_id) = db_xref_item.split(':')\n",
    "                        db_xref_dict[xref_db] = xref_id\n",
    "\n",
    "                    if 'UniProtKB/TrEMBL' in db_xref_dict.keys():\n",
    "                        primary_id = db_xref_dict['UniProtKB/TrEMBL']\n",
    "                    elif 'UniProtKB/Swiss-Prot' in db_xref_dict.keys():\n",
    "                        primary_id = db_xref_dict['UniProtKB/Swiss-Prot']\n",
    "                    elif 'GOA' in db_xref_dict.keys():\n",
    "                        primary_id = db_xref_dict['GOA']\n",
    "                    else:     \n",
    "                        print('No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead')\n",
    "                        print(record.description)\n",
    "                        primary_id = secondary_id\n",
    "\n",
    "                else: \n",
    "                    print('No db_xref for ' + record.id + ' using ' + secondary_id_name + ' instead')\n",
    "                    primary_id = secondary_id\n",
    "            \n",
    "            f_cds_out.write('>' + primary_id + ' '+ secondary_id_name + '=' + secondary_id + ' cds_raw_id=' + record.id + '\\n')\n",
    "            f_cds_out.write(str(record.seq) + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Error file for all but Pombe: \n",
    "No db_xref for lcl|CU928176.1_cds_CAR27779.1_270 using locus_tag instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CU928179.1_cds_CAR29269.1_186 [locus_tag=ZYRO0G04708g] [db_xref=InterPro:IPR009737,InterPro:IPR012336] [protein=ZYRO0G04708p] [protein_id=CAR29269.1] [location=373979..374959] [gbkey=CDS]\n",
    "No db_xref for lcl|CU928179.1_cds_CAR29904.1_821 using locus_tag instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CU928179.1_cds_CAR29911.1_828 [locus_tag=ZYRO0G19932g] [db_xref=InterPro:IPR020485] [protein=ZYRO0G19932p] [protein_id=CAR29911.1] [location=1652252..1652536] [gbkey=CDS]\n",
    "No db_xref for lcl|CU928181.1_cds_CAR31070.1_368 using locus_tag instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58062.1_1 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58063.1_2 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58064.1_3 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58065.1_4 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58066.1_5 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58067.1_6 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58068.1_7 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58069.1_8 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58070.1_9 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58071.1_10 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58072.1_11 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58073.1_12 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58074.1_13 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58075.1_14 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58076.1_15 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58077.1_16 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58078.1_17 using gene instead\n",
    "No db_xref for lcl|DQ508940.1_cds_ABF58079.1_18 using gene instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382133.2_cds_CAG84876.2_553 [locus_tag=DEHA2A13112g] [db_xref=InterPro:IPR011021,InterPro:IPR011022] [protein=DEHA2A13112p] [protein_id=CAG84876.2] [location=complement(1102609..1104957)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382134.2_cds_CAG85501.2_537 [locus_tag=DEHA2B12474g] [db_xref=InterPro:IPR001810] [protein=DEHA2B12474p] [protein_id=CAG85501.2] [location=982449..984296] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382135.2_cds_CAG86143.1_394 [locus_tag=DEHA2C09086g] [db_xref=InterPro:IPR029071] [protein=DEHA2C09086p] [protein_id=CAG86143.1] [location=complement(816908..817129)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382137.2_cds_CAG88023.2_456 [locus_tag=DEHA2E10956g] [db_xref=InterPro:IPR009737,InterPro:IPR012336] [protein=DEHA2E10956p] [protein_id=CAG88023.2] [location=861415..862236] [gbkey=CDS]\n",
    "No db_xref for lcl|CR382137.2_cds_CAG88461.1_865 using locus_tag instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382139.2_cds_CAG91119.2_1045 [locus_tag=DEHA2G24442g] [db_xref=InterPro:IPR000717] [protein=DEHA2G24442p] [protein_id=CAG91119.2] [location=complement(1972263..1973579)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382138.2_cds_CAR66320.1_473 [locus_tag=DEHA2F10890g] [db_xref=InterPro:IPR011989] [protein=DEHA2F10890p] [protein_id=CAR66320.1] [location=complement(936927..939593)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382138.2_cds_CAG89184.2_474 [locus_tag=DEHA2F10912g] [db_xref=InterPro:IPR004328,InterPro:IPR025304] [protein=DEHA2F10912p] [protein_id=CAG89184.2] [location=941905..944205] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382138.2_cds_CAG89693.2_944 [locus_tag=DEHA2F21868g] [db_xref=InterPro:IPR001680,InterPro:IPR015943,InterPro:IPR017986] [protein=DEHA2F21868p] [protein_id=CAG89693.2] [location=1821602..1823356] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CU928167.1_cds_CAR21915.1_115 [locus_tag=KLTH0C02882g] [db_xref=InterPro:IPR009737,InterPro:IPR012336] [protein=KLTH0C02882p] [protein_id=CAR21915.1] [location=253835..254833] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CU928168.1_cds_CAR22480.1_186 [locus_tag=KLTH0D04510g] [db_xref=InterPro:IPR020485] [protein=KLTH0D04510p] [protein_id=CAR22480.1] [location=414739..415074] [gbkey=CDS]\n",
    "No db_xref for lcl|CU928168.1_cds_CAR22530.1_236 using locus_tag instead\n",
    "No db_xref for lcl|CU928171.1_cds_CAR24651.1_34 using locus_tag instead\n",
    "No db_xref for lcl|CU928171.1_cds_CAR25157.1_540 using locus_tag instead\n",
    "No db_xref for lcl|CR382126.1_cds_CAG98395.1_582 using locus_tag instead\n",
    "No db_xref for lcl|CR382126.1_cds_CAG98679.1_856 using locus_tag instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382125.1_cds_CAG99174.1_133 [locus_tag=KLLA0_E03147g] [db_xref=InterPro:IPR029052] [protein=KLLA0E03147p] [protein_id=CAG99174.1] [location=complement(290622..292358)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382125.1_cds_CAH00080.1_1001 [locus_tag=KLLA0_E23079g] [db_xref=InterPro:IPR001810] [protein=KLLA0E23079p] [protein_id=CAH00080.1] [location=complement(2056949..2059129)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382124.1_cds_CAH00274.1_99 [locus_tag=KLLA0_D02464g] [db_xref=InterPro:IPR002733,InterPro:IPR023473,InterPro:IPR027485] [protein=KLLA0D02464p] [protein_id=CAH00274.1] [location=complement(209705..210388)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382124.1_cds_CAH00586.1_394 [locus_tag=KLLA0_D09603g] [db_xref=InterPro:IPR026248] [protein=KLLA0D09603p] [protein_id=CAH00586.1] [location=811234..813696] [gbkey=CDS]\n",
    "No db_xref for lcl|CR382124.1_cds_CAH00778.2_586 using locus_tag instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382122.1_cds_CAH02487.1_520 [locus_tag=KLLA0_B12650g] [db_xref=InterPro:IPR020485] [protein=KLLA0B12650p] [protein_id=CAH02487.1] [location=complement(1104366..1104749)] [gbkey=CDS]\n",
    "No db_xref for lcl|CR382123.1_cds_CAH01062.1_14 using locus_tag instead\n",
    "No db_xref for lcl|CR382123.1_cds_CAH01376.1_313 using locus_tag instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382123.1_cds_CAH01377.1_314 [locus_tag=KLLA0_C07348g] [db_xref=InterPro:IPR004328,InterPro:IPR025304] [protein=KLLA0C07348p] [protein_id=CAH01377.1] [location=639956..641914] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382123.1_cds_CAH01428.1_365 [locus_tag=KLLA0_C08470g] [db_xref=InterPro:IPR000626,InterPro:IPR029071] [protein=KLLA0C08470p] [protein_id=CAH01428.1] [location=742530..742754] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382123.1_cds_CAH01563.1_495 [locus_tag=KLLA0_C11473g] [db_xref=InterPro:IPR011021,InterPro:IPR011022,InterPro:IPR014756] [protein=KLLA0C11473p] [protein_id=CAH01563.1] [location=complement(987546..989264)] [gbkey=CDS]\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382123.1_cds_CAH01790.1_716 [locus_tag=KLLA0_C16533g] [db_xref=InterPro:IPR001680,InterPro:IPR015943,InterPro:IPR017986,InterPro:IPR019775] [protein=KLLA0C16533p] [protein_id=CAH01790.1] [location=complement(1448866..1450404)] [gbkey=CDS]\n",
    "No db_xref for lcl|CR382121.1_cds_CAH02635.1_45 using locus_tag instead\n",
    "No db_xref for lcl|CR382121.1_cds_CAH02866.1_263 using locus_tag instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|CR382121.1_cds_CAH02950.1_345 [locus_tag=KLLA0_A08140g] [db_xref=InterPro:IPR009737] [protein=KLLA0A08140p] [protein_id=CAH02950.1] [location=complement(726017..726985)] [gbkey=CDS]\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64949.1_1 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64950.1_2 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64951.1_3 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64952.1_4 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64956.1_5 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64957.1_6 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64953.1_7 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64955.1_8 using gene instead\n",
    "No db_xref for lcl|AY654900.1_cds_AAT64954.1_9 using gene instead\n",
    "No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead\n",
    "lcl|FN392320.1_cds_CAY69769.1_1169 [locus_tag=PAS_chr2-2_0158] [db_xref=EnsemblGenomes-Gn:PAS_chr2-2_0158,EnsemblGenomes-Tr:CAY69769] [protein=Putative protein of unknown function] [protein_id=CAY69769.1] [location=complement(2091045..2091407)] [gbkey=CDS]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make cds files with uniprot names as keys\n",
    "\n",
    "uniprot_ncbi_genome_specs = [('wickerhamomyces_anomalus','GCF_001661255.1'),  # 'GCA_001661255.1'\n",
    "                ('candida_tropicalis','GCF_000006335.3'),   #'GCA_000006335.3'\n",
    "                ('yHMPu5000034957_hanseniaspora_osmophila_160519','GCA_001747045.1')   #'GCA_001747045.1'\n",
    "               ]\n",
    "\n",
    "for (spec,ncbi_genome_id) in uniprot_ncbi_genome_specs:\n",
    "#(spec,ncbi_genome_id) = uniprot_ncbi_genome_specs[0]\n",
    "\n",
    "    cds_raw_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ncbi_genomes_coding_seq/' + spec + '/ncbi_dataset/data/' + ncbi_genome_id + '/cds_from_genomic.fna')\n",
    "    cds_out_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ncbi_genomes_coding_seq/' + spec + '/cds.fasta')\n",
    "\n",
    "    uniprot_table_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ncbi_genomes_coding_seq/' + spec + '/uniprot_table.tsv')\n",
    "    uniprot_table = pd.read_table(uniprot_table_fname)\n",
    "    assert len(set(list(uniprot_table['Gene Names (ORF)'])))==len(list(uniprot_table['Gene Names (ORF)'])), 'Gene Names (ORF) not unique for ' + spec\n",
    "    uniprot_id_lookup = dict(zip(uniprot_table['Gene Names (ORF)'], uniprot_table['Entry'])) \n",
    "\n",
    "    cds_raw = SeqIO.parse(cds_raw_fname, 'fasta')\n",
    "\n",
    "    uniprot_ids = []\n",
    "    with open(cds_out_fname, 'w') as f_cds_out:\n",
    "\n",
    "        for record in cds_raw: \n",
    "            record_metadata = {}\n",
    "            metadata_list = record.description.split('] [')\n",
    "            metadata_list[0] = metadata_list[0].split('[')[1]\n",
    "            metadata_list[-1] = metadata_list[-1].strip(']')\n",
    "\n",
    "\n",
    "            for metadata_set in metadata_list:\n",
    "                (metadata_label, metadata) = metadata_set.split('=')\n",
    "                record_metadata[metadata_label] = metadata\n",
    "\n",
    "\n",
    "            assert 'locus_tag' in record_metadata.keys(), 'in ' + spec + ' no locus tag for ' + record.description \n",
    "            secondary_id = record_metadata['locus_tag']\n",
    "            secondary_id_name = 'locus_tag'\n",
    "\n",
    "            if secondary_id in uniprot_id_lookup.keys(): \n",
    "                primary_id = uniprot_id_lookup[secondary_id]\n",
    "                f_cds_out.write('>' + primary_id + ' '+ secondary_id_name + '=' + secondary_id + ' cds_raw_id=' + record.id + '\\n')\n",
    "                f_cds_out.write(str(record.seq) + '\\n')\n",
    "                uniprot_ids.append(primary_id)\n",
    "            else: \n",
    "                print(spec + ': no uniprot id for ' + secondary_id)\n",
    "\n",
    "    assert len(uniprot_ids)==len(set(uniprot_ids)), spec + ': duplicate uniprot ids'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "            \n",
    "# km_id_map = pd.read_table(genomes_base + os.sep + os.path.normpath('diverse_yeast/proteomes/Kmar_uniprot_id_table_20221020.tsv'))\n",
    "\n",
    "# #several k.mar proteins have more than one gene name for a single uniprot_id\n",
    "# km_orf_2_uniprot = {}\n",
    "# for orf,uniprot_id in zip(km_id_map['Gene Names (ORF)'], km_id_map['Entry']):\n",
    "#     if len(orf.split(' ')) >1:\n",
    "#         for orf_id in orf.split(' '):\n",
    "#             km_orf_2_uniprot[orf_id] = uniprot_id\n",
    "#     else: \n",
    "#         km_orf_2_uniprot[orf] = uniprot_id\n",
    "\n",
    "\n",
    "\n",
    "        #         if 'locus_tag' in record_metadata.keys(): \n",
    "        #             secondary_id = record_metadata['locus_tag']\n",
    "        #             secondary_id_name = 'locus_tag'\n",
    "        #         elif 'gene' in record_metadata.keys(): \n",
    "        #             secondary_id = record_metadata['gene']\n",
    "        #             secondary_id_name = 'gene'\n",
    "        #         elif 'protein_id' in record_metadata.keys(): \n",
    "        #             secondary_id = record_metadata['protein_id']\n",
    "        #             secondary_id_name = 'protein_id'\n",
    "        #         else:\n",
    "        #             raise ValueError(record.description)\n",
    "\n",
    "        #         if 'db_xref' in record_metadata.keys(): \n",
    "        #             db_xref_dict = {}\n",
    "        #             for db_xref_item in record_metadata['db_xref'].split(','):\n",
    "        #                 (xref_db, xref_id) = db_xref_item.split(':')\n",
    "        #                 db_xref_dict[xref_db] = xref_id\n",
    "\n",
    "        #             if 'UniProtKB/TrEMBL' in db_xref_dict.keys():\n",
    "        #                 primary_id = db_xref_dict['UniProtKB/TrEMBL']\n",
    "        #             elif 'UniProtKB/Swiss-Prot' in db_xref_dict.keys():\n",
    "        #                 primary_id = db_xref_dict['UniProtKB/Swiss-Prot']\n",
    "        #             elif 'GOA' in db_xref_dict.keys():\n",
    "        #                 primary_id = db_xref_dict['GOA']\n",
    "        #             else:     \n",
    "        #                 print('No UniProtKB/TrEMBL,UniProtKB/Swiss-Prot or GOA id in db_xref, using locus_tag instead')\n",
    "        #                 print(record.description)\n",
    "        #                 primary_id = secondary_id\n",
    "\n",
    "        #         else: \n",
    "        #             print('No db_xref for ' + record.id + ' using ' + secondary_id_name + ' instead')\n",
    "        #             primary_id = secondary_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cds for model species\n",
    "\n",
    "#C. albicans\n",
    "#http://www.candidagenome.org/download/sequence/C_albicans_SC5314/Assembly22/current/  \n",
    "# C_albicans_SC5314_A22_current_default_coding.fasta\n",
    "# C_albicans_SC5314_A22_current_default_protein.fasta\n",
    "#downloaded on 20221012 and stored in\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\candida_albicans\n",
    "\n",
    "\n",
    "#S. cerevisiae\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\saccharomyces_cerevisiae\\S288C_reference_genome_R64-2-1_20150113\n",
    "#orf_coding_all_R64-2-1_20150113.fasta\n",
    "#orf_trans_all_R64-2-1_20150113.fasta\n",
    "\n",
    "#S. pombe\n",
    "## The Pombase data didn't match uniprot and therefore didnt match alphafold.  Used Uniprot instead stored at\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\diverse_yeast\\ncbi_coding_seq\\schizosaccharomyces_pombe\n",
    "\n",
    "#From https://www.pombase.org/data/genome_sequence_and_features/feature_sequences/ on 20221012\n",
    "#cds.fa \n",
    "#peptide.fa  \n",
    "#both last modified 2022-10-12 03:26\n",
    "#Stored at\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\schizosaccharomyces_pombe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Do the model Species peptide sequences match the peptide sequences from SGD and CGD? \n",
    "\n",
    "spec_abbrev = 'Scer'\n",
    "\n",
    "#Make dict from uniprot to Length\n",
    "\n",
    "uniprot_proteome_fnames = {'Scer': genomes_base + os.sep + os.path.normpath('diverse_yeast/proteomes/Scer_uniprot-proteome_UP000002311.fasta'),\n",
    "                   'Calb': genomes_base + os.sep + os.path.normpath('diverse_yeast/proteomes/Calb_uniprot-proteome_UP000000559.fasta')\n",
    "                  }\n",
    "\n",
    "uniprot_proteome = SeqIO.parse(uniprot_proteome_fnames[spec_abbrev], 'fasta')\n",
    "\n",
    "uniprot_proteome_lengths = {}\n",
    "for record in uniprot_proteome: \n",
    "    uniprot_proteome_lengths[record.id.split('|')[1]] = len(record.seq)\n",
    "\n",
    "#Make dict from geneID to length\n",
    "model_proteome_fnames = {\n",
    "    'Scer': genomes_base + os.sep + os.path.normpath('saccharomyces_cerevisiae/S288C_reference_genome_R64-2-1_20150113/orf_trans_all_R64-2-1_20150113.fasta'),\n",
    "    'Calb': genomes_base + os.sep + os.path.normpath('candida_albicans/C_albicans_SC5314_A22_current_default_protein.fasta')\n",
    "}\n",
    "\n",
    "model_proteome = SeqIO.parse(model_proteome_fnames[spec_abbrev], 'fasta')\n",
    "\n",
    "model_proteome_lengths = {}\n",
    "for record in model_proteome: \n",
    "    model_proteome_lengths[record.id] = len(record.seq)\n",
    "\n",
    "\n",
    "model_gene_id_2_swissprot = dyt.load_model_gene_id_2_swissprot_id()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "discrepancy_list_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/proteomes/' + spec_abbrev + '_discrepancy_list.txt')\n",
    "\n",
    "with open(discrepancy_list_fname,'w') as f_out: \n",
    "\n",
    "    for gene_id, len_model in model_proteome_lengths.items(): \n",
    "        if gene_id in model_gene_id_2_swissprot[spec_abbrev].keys(): \n",
    "            uniprot_id = model_gene_id_2_swissprot[spec_abbrev][gene_id]\n",
    "\n",
    "            if uniprot_id in uniprot_proteome_lengths.keys(): \n",
    "                len_uniprot = uniprot_proteome_lengths[uniprot_id]\n",
    "\n",
    "\n",
    "                if len_model != len_uniprot+1: \n",
    "                    f_out.write('Length mismatch ' + gene_id + ': ' + str(len_model) +' ' + uniprot_id + ': ' + str(len_uniprot) + '\\n')\n",
    "            else: \n",
    "                if isinstance(uniprot_id,float): \n",
    "                    assert np.isnan(uniprot_id), gene_id + ': Float but not nan for uniprot ID'\n",
    "                    f_out.write(gene_id + ' has nan for uniprot_id\\n')\n",
    "                else: \n",
    "                    f_out.write(uniprot_id + '/' + gene_id + ' not in uniprot_proteome_lengths lookup\\n')\n",
    "        else: \n",
    "            f_out.write(gene_id + ' has no uniprot ID\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Both S. cereivisiae and C. albicans model databases include the stop codon, but that is not included in Uniprot.  \n",
    "#For the most part, both the databases match in terms of length, but there are some exceptions and some that are hard to look up in uniprot from S. cer. \n",
    "##Should probably use S.cer and C.alb uniprot data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Extra things"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Code to gather cds from Shen information I believe now code is in Documents/GitHub/diverse_yeast/20221012_assemble_peptide_cds_fastas.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Read in orthogroups from file\n",
    "# #Load og_out_data_filt\n",
    "# #STR2/HSU1: OG1390\n",
    "# #ERG11: OG1355\n",
    "# #PYK2: OG1299\n",
    "\n",
    "# #>Alloascoidea_hylecoeti|0_1963  identifier_proteome_id=-augustus-gene-10.125-mRNA-1_1\n",
    "# target_og = 'OG1390'\n",
    "    \n",
    "# # Gather CDS sequences into one fasta\n",
    "# cds_fasta_out_fname = base_dir + os.sep + 'alphafold' + os.sep + target_og + '_cds.fasta'\n",
    "\n",
    "# spec_orig_genome_lookup\n",
    "\n",
    "# og_genes = og_out_data_full[target_og]['og_genes']\n",
    "\n",
    "# # with open(cds_fasta_out_fname, 'w') as f_out: \n",
    "# #     for gene in og_genes: \n",
    "# #         #extract species number\n",
    "# #         #extract species name\n",
    "# #         #extract proteinid\n",
    "        \n",
    "# y1000plus_tools.extract_cds_seqs(og_genes, cds_fasta_out_fname, y1000_species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra things\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra things\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Extract CDS for all genes in orthogroup: \n",
    "\n",
    "# target_og = 'OG1390'\n",
    "    \n",
    "# # Gather CDS sequences into one fasta\n",
    "# cds_all_fasta_out_fname = base_dir + os.sep + 'alphafold' + os.sep + target_og + '_all_cds.fasta'\n",
    "\n",
    "# spec_orig_genome_lookup\n",
    "\n",
    "# og_genes = og_genes_lookup[target_og]\n",
    "\n",
    "# y1000plus_tools.extract_cds_seqs(og_genes, cds_all_fasta_out_fname, y1000_species)\n",
    "# print(protein_fname + ' cds sequence extraction complete')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# #the following information is in the fasta header for each protein\n",
    "# #<spec_orig_genome>__<og>__<y1000_id>: you can use this to look up other metadata in the og_metadata.json file\n",
    "# #source:  The source of the peptide sequence (either 'shen' or 'uniprot')\n",
    "# #gene_full:  The full name of the gene from source peptide file\n",
    "# #gene_full_shen: the full name of the gene from the shen peptide file\n",
    "# #L: Length of the gene\n",
    "# #sim_score_vs_shen: The similarity score between the shen peptide sequence and the uniprot peptide sequence.  If it is above the threshold, the uniprot sequence is used, below the threshold, we continue to use the shen sequence\n",
    "# #sim_score_vs_shen_diff: The difference between similarity score of the top peptide and the second highest scoring peptide.  For genes with similarity score below a threshold, if this difference is too small, we keep the shen sequence\n",
    "\n",
    "# shen_proteome_dir = os.path.normpath(y1000plus_dir + 'shen_2018_data/0_332yeast_genomes/332_genome_annotations/pep') + os.sep \n",
    "# uniprot_proteome_dir = os.path.normpath('C:/Users/heineib/Google Drive/Crick_LMS/external_data/genomes/diverse_yeast/proteomes') + os.sep \n",
    "\n",
    "# #(species genome name: (shen_fname, uniprot_fname)): \n",
    "# proteome_fasta_dict = {'kluyveromyces_lactis': {'shen': shen_proteome_dir + 'kluyveromyces_lactis.max.pep', \n",
    "#                                                 'uniprot': uniprot_proteome_dir + 'Klac_uniprot-proteome_UP000000598.fasta'\n",
    "#                                                }, \n",
    "#                        'wickerhamomyces_anomalus': {'shen': shen_proteome_dir + 'wickerhamomyces_anomalus.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Wano_uniprot-proteome_UP000094112.fasta'\n",
    "#                            }, \n",
    "#                        'lachancea_thermotolerans': {'shen': shen_proteome_dir + 'lachancea_thermotolerans.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Lthe_uniprot-proteome_UP000002036.fasta'\n",
    "#                            }, \n",
    "#                        'komagataella_pastoris': {'shen': shen_proteome_dir + 'komagataella_pastoris.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Kpas_uniprot-proteome_UP000000314.fasta'\n",
    "#                            }, \n",
    "#                         'candida_tropicalis': {'shen': shen_proteome_dir + 'candida_tropicalis.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Ctro_uniprot-proteome_UP000002037.fasta'\n",
    "#                            },\n",
    "#                         'debaryomyces_hansenii': {'shen': shen_proteome_dir + 'debaryomyces_hansenii.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Dhan_uniprot-proteome_UP000000599.fasta'\n",
    "#                            }, \n",
    "#                         'yHMPu5000034957_hanseniaspora_osmophila_160519': {'shen': shen_proteome_dir + 'yHMPu5000034957_hanseniaspora_osmophila_160519.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Hosm_uniprot-proteome_UP000095728.fasta'\n",
    "#                            }, \n",
    "#                         'kluyveromyces_marxianus': {'shen': shen_proteome_dir + 'kluyveromyces_marxianus.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Kmar_uniprot-proteome_UP000065495.fasta'\n",
    "#                            }, \n",
    "#                         'zygosaccharomyces_rouxii': {'shen': shen_proteome_dir + 'zygosaccharomyces_rouxii.max.pep',\n",
    "#                             'uniprot': uniprot_proteome_dir + 'Zrou_uniprot-proteome_UP000008536.fasta'\n",
    "#                            }\n",
    "#                       }\n",
    "\n",
    "\n",
    "# sim_score_high_thresh = 75\n",
    "# sim_score_low_thresh = 60\n",
    "# sim_score_diff_thresh = 8\n",
    "\n",
    "\n",
    "# with open(fasta_out_fname, 'w') as f_out: \n",
    "\n",
    "#     #for each species: \n",
    "#     for spec_id, spec_name in spec_id_map.items(): \n",
    "#         spec = spec_id_map[spec_id]\n",
    "\n",
    "#         spec_orig_genome = spec_orig_genome_lookup[spec_id]\n",
    "#         print(spec_orig_genome)\n",
    "        \n",
    "#         if spec_orig_genome in {'saccharomyces_cerevisiae', 'candida_albicans'}: \n",
    "#             print('skipping ' + spec_orig_genome)\n",
    "#             #Skip this species\n",
    "#         elif spec_orig_genome in proteome_fasta_dict.keys():\n",
    "#             #Attempt to convert to uniprot fasta  \n",
    "#             print('checking to see if uniprot or shen et al sequences should be used')\n",
    "#             source = 'uniprot'\n",
    "            \n",
    "#             protein_fname_uniprot = proteome_fasta_dict[spec_orig_genome]['uniprot']\n",
    "#             seq_records_uniprot = SeqIO.parse(protein_fname_uniprot, \"fasta\")\n",
    "#             pairwise_scores_spec = pairwise_scores[spec_orig_genome]\n",
    "            \n",
    "#             protein_fname_shen = proteome_fasta_dict[spec_orig_genome]['shen']\n",
    "#             seq_records_shen = SeqIO.parse(protein_fname_shen, \"fasta\")\n",
    "        \n",
    "#             spec_lookup_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/' + spec_orig_genome + '.csv')\n",
    "#             spec_lookup = pd.read_csv(spec_lookup_fname, index_col=0)\n",
    "\n",
    "# #             if spec_orig_genome == 'saccharomyces_cerevisiae':\n",
    "# #                 #for S. cerevisiae the orf name is the 'gene_full'\n",
    "# #                 gene_full_y1000_id_lookup = dict(zip(spec_lookup['y1000_id'], spec_lookup.index))\n",
    "# #             else: \n",
    "#             gene_full_y1000_id_lookup = dict(zip(spec_lookup['y1000_id'], spec_lookup['gene_full']))\n",
    "\n",
    "#             #for each orthogroup:\n",
    "#             #Identify whether the proteins will be unipro or shen \n",
    "#             seqs_to_get_uniprot = {} #gene_full_uniprot: (y1000_id, og, maxval, diff_top2_val, gene_full_shen)\n",
    "#             seqs_to_get_shen = {}  #gene_full_shen: (y1000_id, og, maxval, diff_top2_val)\n",
    "#             gene_full_set_uniprot = []\n",
    "#             gene_full_set_shen = []\n",
    "#             for og, (N_genes, pct_present, og_genes_out) in og_out_data.items():\n",
    "#                 for y1000_id in og_genes_out: \n",
    "#                     spec_id_check = int(y1000_id.split('_')[0])\n",
    "#                     if spec_id_check==spec_id:\n",
    "#                         gene_full_shen = gene_full_y1000_id_lookup[y1000_id]\n",
    "#                         pairwise_score_gene = pairwise_scores_spec[gene_full_shen]\n",
    "                        \n",
    "#                         #Check to see if pairwise scores pass filters: \n",
    "#                         options_sorted = pairwise_score_gene.sort_values(ascending=False)  #possible that they are already sorted and this is unnecessary\n",
    "#                         maxval = options_sorted[0]\n",
    "#                         diff_top2_val = maxval-options_sorted[1]\n",
    "                        \n",
    "#                         if ((maxval>sim_score_high_thresh) | ((maxval>sim_score_low_thresh) & (diff_top2_val>sim_score_diff_thresh))): \n",
    "#                             gene_full_uniprot = options_sorted.index[0]\n",
    "#                             seqs_to_get_uniprot[gene_full_uniprot] = (y1000_id, og, maxval, diff_top2_val, gene_full_shen)\n",
    "#                             gene_full_set_uniprot.append(gene_full_uniprot)\n",
    "#                         else: \n",
    "#                             seqs_to_get_shen[gene_full_shen] = (y1000_id, og, maxval, diff_top2_val)\n",
    "#                             gene_full_set_shen.append(gene_full_shen)\n",
    "                                      \n",
    "#             #Cycle through uniprot proteome and print to fasta if it is in the set. \n",
    "#             seq_record_fasta_printout(seq_records_uniprot, f_out, gene_full_set_uniprot, seqs_to_get_uniprot, 'uniprot', spec_orig_genome)\n",
    "                              \n",
    "#             #Cycle through shen proteome and print to fasta if it is in the set. \n",
    "#             seq_record_fasta_printout(seq_records_shen, f_out, gene_full_set_shen, seqs_to_get_shen, 'shen', spec_orig_genome)\n",
    "\n",
    "        \n",
    "        \n",
    "#         else: \n",
    "#             print('using shen et al sequence')\n",
    "#             fasta_extract_shen(f,shen_proteome_dir, spec_orig_genome, y1000plus_dir, og_out_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fname = cds_fasta_out_fname\n",
    "# y1000_species_subset = y1000_species\n",
    "\n",
    "# # def extract_cds_seqs(og_genes, fname, y1000_species_subset): \n",
    "# #     #Looks up coding sequences for given list of orthogroup genes \n",
    "# #     #\n",
    "# #     ## Does not work for outgroup species\n",
    "\n",
    "# #     #os.mkdir(y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/proteins_og/' + fname))\n",
    "# #     #proteins_og_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/proteins_og/' + fname + '/' + fname + '.fasta')\n",
    "\n",
    "# #     genome_name_lookup = dict(zip(y1000_species_subset['spec_og_id'],y1000_species_subset['original_genome_id']))\n",
    "\n",
    "# #     with open(fname,'w') as f: \n",
    "# #         # group by species\n",
    "# #         og_genes_specs = {spec_gene.split('_')[0]: [] for spec_gene in og_genes}\n",
    "# #         for y1000_id in og_genes: \n",
    "# #             spec_og_id = y1000_id.split('_')[0]\n",
    "# #             og_genes_specs[spec_og_id].append(y1000_id)\n",
    "\n",
    "# #         for spec_og_id, genes in og_genes_specs.items(): \n",
    "# #             genome_name = genome_name_lookup[int(spec_og_id)]\n",
    "# #             print(genome_name)\n",
    "# #             #if S.Cer or C. Albicans, do slightly different routine\n",
    "\n",
    "# #             if genome_name == 'saccharomyces_cerevisiae':\n",
    "# #                 #load gene_id map based on the species\n",
    "# #                 gene_lookup_spec_fname = y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/id_lookups/\" + genome_name + '.csv')\n",
    "# #                 gene_lookup_spec = pd.read_csv(gene_lookup_spec_fname, index_col='y1000_id')\n",
    "\n",
    "# #                 cds_fname = y1000plus_dir + os.path.normpath('shen_2018_data/0_332yeast_genomes/332_genome_annotations/Saccharomyces_cerevisiae_S288C_coding.fasta')\n",
    "\n",
    "# #                 seq_records = SeqIO.parse(cds_fname, \"fasta\")\n",
    "\n",
    "# #                 genes_lookup = gene_lookup_spec.loc[genes]\n",
    "# #                 genes_lookup_set = set(gene_lookup_spec.loc[genes,'gene_id'].values)  #SC specific\n",
    "\n",
    "# #                 for seq_record in seq_records:\n",
    "# #                     gene_id = seq_record.description.split()[0] #SC specific\n",
    "# #                     #print(gene_full)\n",
    "# #                     if (gene_id in genes_lookup_set):\n",
    "# #                         #find which y1000_id was matched\n",
    "# #                         y1000_rlookup = (genes_lookup['gene_id'] == gene_id)\n",
    "# #                         for gene, tf in y1000_rlookup.items(): \n",
    "# #                             if tf:\n",
    "# #                                 y1000_id=gene\n",
    "# #                         cds_seq = seq_record.seq\n",
    "# #                         f.write('>' + genome_name + '_' + gene_id + ' y1000_id=' + y1000_id + ' gene_id=' + gene_id +\n",
    "# #                                 ' common_name=' + seq_record.description.split()[1] + #this adds in the cds from the original description\n",
    "# #                                 '\\n')\n",
    "# #                         f.write(str(cds_seq) + '\\n')  \n",
    "# #             elif genome_name == 'candida_albicans':\n",
    "# #                 #load gene_id map based on the species\n",
    "# #                 gene_lookup_spec_fname = y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/id_lookups/\" + genome_name + '.csv')\n",
    "# #                 gene_lookup_spec = pd.read_csv(gene_lookup_spec_fname, index_col='y1000_id')\n",
    "\n",
    "# #                 cds_fname = y1000plus_dir + os.path.normpath('shen_2018_data/0_332yeast_genomes/332_genome_annotations/Candida_albicans_SC5314_A22_current_default_coding.fasta')\n",
    "# #                 seq_records = SeqIO.parse(cds_fname, \"fasta\")\n",
    "\n",
    "# #                 genes_lookup = gene_lookup_spec.loc[genes]\n",
    "# #                 genes_lookup_set = set(gene_lookup_spec.loc[genes,'gene_id'].values)  \n",
    "\n",
    "# #                 for seq_record in seq_records:\n",
    "# #                     gene_id = seq_record.description \n",
    "\n",
    "# #                     if (gene_id in genes_lookup_set):\n",
    "# #                         #find which y1000_id was matched\n",
    "# #                         y1000_rlookup = (genes_lookup['gene_id'] == gene_id)\n",
    "# #                         for gene, tf in y1000_rlookup.items(): \n",
    "# #                             if tf:\n",
    "# #                                 y1000_id=gene\n",
    "# #                         cds_seq = seq_record.seq\n",
    "# #                         f.write('>' + genome_name + '_' + gene_id + ' y1000_id=' + y1000_id + ' gene_id=' + gene_id +'\\n')\n",
    "# #                         f.write(str(cds_seq) + '\\n')  #I wonder why some of the bases were in lower case\n",
    "\n",
    "# #             else:   #if not(genome_name in missing_specs):    could also include a missing species list\n",
    "# #                 #load gene_id map based on the species\n",
    "# #                 gene_lookup_spec_fname = y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/id_lookups/\" + genome_name + '.csv')\n",
    "# #                 gene_lookup_spec = pd.read_csv(gene_lookup_spec_fname, index_col='y1000_id')\n",
    "\n",
    "# #                 #Extract peptide sequences from peptide fasta from genome\n",
    "# #                 cds_dir = os.path.normpath(y1000plus_dir + 'shen_2018_data/0_332yeast_genomes/332_genome_annotations/cds') + os.sep \n",
    "\n",
    "# #                 cds_fname = cds_dir + genome_name + '.max.cds'\n",
    "\n",
    "# #                 seq_records = SeqIO.parse(cds_fname, \"fasta\")\n",
    "\n",
    "# #                 genes_lookup = gene_lookup_spec.loc[genes]\n",
    "# #                 genes_lookup_set = set(gene_lookup_spec.loc[genes,'gene_full'].values)\n",
    "\n",
    "# #                 for seq_record in seq_records:\n",
    "# #                     #gene_full = 'augustus_masked-Deha2C-processed-gene-4.36'\n",
    "# #                     gene_full = seq_record.description.split()[1].split('=')[1]\n",
    "# #                     #print(gene_full)\n",
    "# #                     if (gene_full in genes_lookup_set):\n",
    "# #                         #find which y1000_id was matched\n",
    "# #                         y1000_rlookup = genes_lookup['gene_full'] == gene_full\n",
    "# #                         for gene, tf in y1000_rlookup.items(): \n",
    "# #                             if tf:\n",
    "# #                                 y1000_id=gene\n",
    "# #                         gene_id = genes_lookup.loc[y1000_id, 'gene_id']\n",
    "# #                         cds_seq = seq_record.seq\n",
    "# #                         f.write('>' + genome_name + '_' + gene_id + ' y1000_id=' + y1000_id + ' gene_full=' + gene_full +\n",
    "# #                                 ' ' + seq_record.description.split()[2] + #this adds in the cds from the original description\n",
    "# #                                 '\\n')\n",
    "# #                         f.write(str(cds_seq) + '\\n')  #I wonder why some of the bases were in lower case\n",
    "\n",
    "# #     return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ideally change to something like this: \n",
    "#>Alloascoidea_hylecoeti|0_1963  identifier_proteome_id=-augustus-gene-10.125-mRNA-1_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# ## Make Codon based genes sets, align them and make phylogenetic trees\n",
    "# og_fname = base_dir + os.sep + 'alphafold' + os.sep + 'og_metadata.json'\n",
    "# with open(og_fname, 'r') as f:\n",
    "#     og_out_data_full = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast.Applications import NcbimakeblastdbCommandline\n",
    "from Bio.Blast.Applications import NcbiblastpCommandline\n",
    "\n",
    "cline = NcbimakeblastdbCommandline(dbtype=\"prot\",\n",
    "                                   input_file=\"NC_005816.faa\")\n",
    "cline\n",
    "NcbimakeblastdbCommandline(cmd='makeblastdb', dbtype='prot', input_file='NC_005816.faa')\n",
    "print(cline)\n",
    "makeblastdb -dbtype prot -in NC_005816.faa\n",
    "\n",
    "\n",
    "cline = NcbiblastpCommandline(query=\"rosemary.pro\", db=\"nr\",\n",
    "                              evalue=0.001, remote=True, ungapped=True)\n",
    "cline\n",
    "NcbiblastpCommandline(cmd='blastp', query='rosemary.pro', db='nr', evalue=0.001, remote=True, ungapped=True)\n",
    "print(cline)\n",
    "blastp -query rosemary.pro -db nr -evalue 0.001 -remote -ungapped\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#     if check_gene: \n",
    "#         print(str(N) + ' ' + genename_study1)\n",
    "#         AA_seqs_study2 = SeqIO.parse(AA_seqs_fnames[study2], \"fasta\")\n",
    "#         pairwise_score_comps = {}\n",
    "#         for seq_study2 in AA_seqs_study2: \n",
    "#             genename_study2 = seq_study2.id\n",
    "#             score = pairwise2.align.globalms(seq_study2,seq_study1, match_points, mismatch_points, gap_open, gap_extension, penalize_end_gaps = True, score_only = True )\n",
    "#             pairwise_score_comps[genename_study2] = score\n",
    "#         pairwise_score_comps_df = pd.Series(pairwise_score_comps)\n",
    "#         pairwise_score[genename_study1] = pairwise_score_comps_df.sort_values(ascending=False).iloc[0:NN_top_scores]\n",
    "#         #print(pairwise_score[genename_YGOB])\n",
    "#         N = N - 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#scer_y1000_id_gene_id_lookup = dict(zip(spec_lookup['y1000_id'], scer_lookup.index))\n",
    "    \n",
    "#Check 'quality' of sequences\n",
    "    # GC content\n",
    "    # Length of each species\n",
    "\n",
    "#Map seq from shen genomes to our genomes. \n",
    "\n",
    "#Make Metadata for each orthogroup\n",
    "# SC Genenames, SP genenames, CA genenames\n",
    "# Total number\n",
    "# Percent of species covered\n",
    "# Pathways\n",
    "# Enzyme numbers\n",
    "\n",
    "#Fasta with Sc, Sp, and Calb removed\n",
    "\n",
    "# maybe later\n",
    "# PCT Conservation (if I do MSA for each?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Quantify information for SGD annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For each of the 11 species\n",
    "\n",
    "spec_source = 'Calb'\n",
    "spec_target = 'Scer'\n",
    "\n",
    "eggnog_path = 'eggnog_mapper' + os.sep + 'proteomics_set' + os.sep\n",
    "\n",
    "eggnog_orths = pd.read_csv(base_dir + os.sep + eggnog_path + spec_source + '_' + spec_target + '_annot.csv')\n",
    "\n",
    "#for both genomes\n",
    "\n",
    "#Show how many map to something, how many map to S.cer\n",
    "\n",
    "#EC percentage for each one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ENA Species - this code is not used anymore as I was able to get the cds' from ncbi\n",
    "\n",
    "#With Nir's Help I ran a routine to use emboss extract_features on the embl files \n",
    "\n",
    "#First installed emboss on WSL\n",
    "#sudo apt-get emboss\n",
    "#Needed to run it with some updates because it didn't install the first time.\n",
    "#\n",
    "#Then made a script, ena_cds_extract.py to extract and provide Uniprot names\n",
    "#\n",
    "#ran with\n",
    "#python3 ./ena_cds_extract.py from the folder with the scrip\n",
    "#\n",
    "#The files were originally downloaded in Oct 2022 from:\n",
    "\n",
    "# W. anomalus: https://www.ebi.ac.uk/ena/browser/view/GCA_001661255.1?show=blobtoolkit\n",
    "# H. osmophila: https://www.ebi.ac.uk/ena/browser/view/GCA_001747045.1?show=blobtoolkit\n",
    "# C. tropicalis: https://www.ebi.ac.uk/ena/browser/view/GCA_000006335.3?show=blobtoolkit\n",
    "\n",
    "#Also extracted the genome (.fasta and embl version - changed the extension of the embl version from .txt to .embl) and put it into the appropriate folder\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #make cds files with uniprot names as keys\n",
    "# ena_speclist = [('wickerhamomyces_anomalus','GCA_001661255.1'),\n",
    "#                 ('candida_tropicalis','GCA_000006335.3'),\n",
    "#                 ('yHMPu5000034957_hanseniaspora_osmophila_160519','GCA_001747045.1')\n",
    "#                ]\n",
    "\n",
    "# for spec, gca_name in ena_speclist:\n",
    "#     cds_raw_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ena/' + spec + '/' + 'cds_raw.fasta')\n",
    "#     cds_out_fname = genomes_base + os.sep + os.path.normpath('diverse_yeast/ena/' + spec + '/' + 'cds.fasta')\n",
    "    \n",
    "#     cds_raw = SeqIO.parse(cds_raw_fname, 'fasta')\n",
    "    \n",
    "#     with open(cds_out_fname, 'w') as f_cds_out:\n",
    "    \n",
    "#         for record in cds_raw: \n",
    "#             db_xref_list = record.description.split('(')[1].split(')')[0].split(', ')\n",
    "#             db_xref_dict = {}\n",
    "#             for db_xref_item in db_xref_list: \n",
    "#                 (xref_db, xref_id) = db_xref_item.split('=')[1].strip('\"').split(':')\n",
    "#                 db_xref_dict[xref_db] = xref_id\n",
    "            \n",
    "#             if 'UniProtKB/TrEMBL' in db_xref_dict.keys():\n",
    "#                 id_field = 'UniProtKB/TrEMBL'\n",
    "#             elif 'UniProtKB/Swiss-Prot' in db_xref_dict.keys():\n",
    "#                 id_field =  'UniProtKB/Swiss-Prot'\n",
    "#             else:     \n",
    "#                 print(db_xref_dict.keys())\n",
    "                \n",
    "#             if 'EnsemblGenomes-Gn' in db_xref_dict.keys():\n",
    "#                 ena_id = db_xref_dict['EnsemblGenomes-Gn']\n",
    "#             else: \n",
    "#                 ena_id = 'None'\n",
    "                \n",
    "            \n",
    "#             f_cds_out.write('>' + db_xref_dict[id_field] + ' ena_id=' + ena_id + ' cds_raw_id=' + record.id + '\\n')\n",
    "#             f_cds_out.write(str(record.seq) + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The following code didn't work to use biopython to extract ids from .embl files\n",
    "#Kept getting Unexpected content after SQ or CO line: 'XX'\n",
    "\n",
    "# embl_in = genomes_base + os.sep + os.path.normpath('diverse_yeast/ena/candida_tropicalis/GCA_000006335.3.embl')\n",
    "# #genomes_base + os.sep + os.path.normpath('ena/wickerhamomyces_anomalus/GCA_001661255.1.embl')\n",
    "\n",
    "\n",
    "# cds_out = genomes_base + os.sep + os.path.normpath('diverse_yeast/ena/wickerhamomyces_anomalus/GCA_001661255.1.cds.uniprot_ids.fasta')\n",
    "# prot_out = genomes_base + os.sep + os.path.normpath('diverse_yeast/ena/wickerhamomyces_anomalus/GCA_001661255.1.prot.uniprot_ids.fasta')\n",
    "\n",
    "# with open(embl_in, 'r') as embl_handle:\n",
    "#     with open(cds_out, 'w') as f_cds: \n",
    "#         with open(prot_out,'w') as f_prot: \n",
    "#             for record in SeqIO.parse(embl_handle, \"embl\"):\n",
    "#                 q = 1+1\n",
    "\n",
    "# #Emboss installation https://emboss.sourceforge.net/download/\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
