{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Use y1000plus_tools functions? (Y/n) Y\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "y1000plus_tools dir is C:\\Users\\heineib\\Documents\\GitHub\\y1000plus_tools\\\n",
      "y1000plus data dir is C:\\Users\\heineib\\Documents\\GitHub\\y1000plus_tools\\data\\\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pickle \n",
    "import sys\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "# if yeast_esr_flag =='Y':\n",
    "yeast_esr_exp_path = os.path.normpath('C:/Users/heineib/Documents/GitHub/yeast_esr_expression_analysis') + os.sep\n",
    "#io_library_path_core = io_library_path + 'core' + os.sep\n",
    "if not(yeast_esr_exp_path in sys.path):\n",
    "    sys.path.append(yeast_esr_exp_path)\n",
    "    print(\"Added \" + yeast_esr_exp_path + \" to path\" )\n",
    "\n",
    "import yeast_esr_exp\n",
    "yeast_esr_exp.base_dir = yeast_esr_exp_path\n",
    "yeast_esr_exp.data_processing_dir = yeast_esr_exp_path + os.path.normpath('expression_data') + os.sep\n",
    "\n",
    "\n",
    "y1000plus_tools_flag = input(\"Use y1000plus_tools functions? (Y/n)\")\n",
    "\n",
    "# if y1000plus_tools_flag =='Y':\n",
    "y1000plus_tools_path = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools') + os.sep\n",
    "#io_library_path_core = io_library_path + 'core' + os.sep\n",
    "if not(y1000plus_tools_path in sys.path):\n",
    "    sys.path.append(y1000plus_tools_path)\n",
    "    print(\"Added \" + y1000plus_tools_path + \" to path\" )\n",
    "\n",
    "import y1000plus_tools\n",
    "y1000plus_tools.base_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools') + os.sep\n",
    "print(\"y1000plus_tools dir is \" + y1000plus_tools.base_dir )                      \n",
    "y1000plus_tools.y1000plus_dir = y1000plus_tools.base_dir + 'data' + os.sep \n",
    "print(\"y1000plus data dir is \" + y1000plus_tools.y1000plus_dir)\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "y1000plus_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools/data') + os.sep\n",
    "#base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts')\n",
    "\n",
    "\n",
    "#Load y1000 species table\n",
    "y1000_species = pd.read_csv(y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/y1000_species_table.csv\"), index_col=0)\n",
    "\n",
    "\n",
    "#D. hansenii and H. osmophilia  missing from 220811_Diverse_species_SF08_6h_CN1_woSlowGrowth.csv?\n",
    "spec_abbrev_map = {'BY': 'saccharomyces_cerevisiae', \n",
    "                   'SC': 'saccharomyces_cerevisiae', \n",
    "                   'CA': 'candida_albicans', \n",
    "                   'CT': 'candida_tropicalis', \n",
    "                   'KL': 'kluyveromyces_lactis', \n",
    "                   'KM': 'kluyveromyces_marxianus', \n",
    "                   'LT': 'lachancea_thermotolerans',\n",
    "                   'PP': 'komagataella_pastoris', \n",
    "                   'WA': 'wickerhamomyces_anomalus', \n",
    "                   'ZR': 'zygosaccharomyces_rouxii'\n",
    "                  }\n",
    "\n",
    "\n",
    "##Need to add something for S. cer\n",
    "shen_proteome_dir = os.path.normpath(y1000plus_dir + 'shen_2018_data/0_332yeast_genomes/332_genome_annotations/pep') + os.sep \n",
    "uniprot_proteome_dir = os.path.normpath('G:/My Drive/Crick_LMS/external_data/genomes/diverse_strains/proteomes') + os.sep \n",
    "\n",
    "#(species genome name: (shen_fname, uniprot_fname)): \n",
    "proteome_fasta_dict = {'kluyveromyces_lactis': {'shen': shen_proteome_dir + 'kluyveromyces_lactis.max.pep', \n",
    "                                                'uniprot': uniprot_proteome_dir + 'Klac_uniprot-proteome_UP000000598.fasta'\n",
    "                                               }, \n",
    "                       'wickerhamomyces_anomalus': {'shen': shen_proteome_dir + 'wickerhamomyces_anomalus.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Wano_uniprot-proteome_UP000094112.fasta'\n",
    "                           }, \n",
    "                       'lachancea_thermotolerans': {'shen': shen_proteome_dir + 'lachancea_thermotolerans.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Lthe_uniprot-proteome_UP000002036.fasta'\n",
    "                           }, \n",
    "                       'komagataella_pastoris': {'shen': shen_proteome_dir + 'komagataella_pastoris.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Kpas_uniprot-proteome_UP000000314.fasta'\n",
    "                           }, \n",
    "                        'candida_tropicalis': {'shen': shen_proteome_dir + 'candida_tropicalis.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Ctro_uniprot-proteome_UP000002037.fasta'\n",
    "                           },\n",
    "                        'debaryomyces_hansenii': {'shen': shen_proteome_dir + 'debaryomyces_hansenii.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Dhan_uniprot-proteome_UP000000599.fasta'\n",
    "                           }, \n",
    "                        'yHMPu5000034957_hanseniaspora_osmophila_160519': {'shen': shen_proteome_dir + 'yHMPu5000034957_hanseniaspora_osmophila_160519.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Hosm_uniprot-proteome_UP000095728.fasta'\n",
    "                           }, \n",
    "                        'kluyveromyces_marxianus': {'shen': shen_proteome_dir + 'kluyveromyces_marxianus.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Kmar_uniprot-proteome_UP000065495.fasta'\n",
    "                           }, \n",
    "                        'zygosaccharomyces_rouxii': {'shen': shen_proteome_dir + 'zygosaccharomyces_rouxii.max.pep',\n",
    "                            'uniprot': uniprot_proteome_dir + 'Zrou_uniprot-proteome_UP000008536.fasta'\n",
    "                           }, \n",
    "                        'saccharomyces_cerevisiae': {'uniprot': uniprot_proteome_dir + 'Scer_uniprot-proteome_UP000002311.fasta'}\n",
    "                      }\n",
    "\n",
    "\n",
    "\n",
    "base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts')\n",
    "\n",
    "pairwise_scores_fname = base_dir + os.sep + 'alphafold' + os.sep + 'pairwise_scores.pkl'\n",
    "# Load pairwise_scores\n",
    "with open(pairwise_scores_fname, 'rb') as f: \n",
    "    pairwise_scores = pickle.load(f)\n",
    "\n",
    "    \n",
    "#Load Dataset\n",
    "combined_data = pd.read_csv(base_dir + os.sep + os.path.normpath('proteomics/220811/220811_Diverse_species_SF08_6h_CN1_woSlowGrowth.csv'), index_col=0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#spec_orig_genome_lookup = dict(zip(y1000_species['spec_og_id'], y1000_species['original_genome_id']))\n",
    "\n",
    "\n",
    "# #Load S.Cer lookup table: \n",
    "# scer_lookup_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/saccharomyces_cerevisiae.csv')\n",
    "\n",
    "# scer_lookup = pd.read_csv(scer_lookup_fname, index_col=0)\n",
    "# scer_y1000_id_gene_id_lookup = dict(zip(scer_lookup['y1000_id'], scer_lookup.index))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save mapping files for each species in the diverse yeast proteomics dataset.  Maps from \n",
    "#Also saves log file for numbers of ids found and lists multimappers. \n",
    "\n",
    "fname_log = base_dir + os.sep + os.path.normpath('proteomics/220811/uniprot_2_y1000id_log.txt')\n",
    "\n",
    "with open(fname_log,'w') as f_log: \n",
    "    for spec_abbrev, spec in spec_abbrev_map.items():\n",
    "        #Skipping S. cer because I don't have pairwise scores for that\n",
    "        if not(spec_abbrev in {'SC', 'BY', 'CA'}): \n",
    "               #spec_abbrev = 'KL' \n",
    "            #spec = spec_abbrev_map[spec_abbrev]\n",
    "            fname_out = base_dir + os.sep + os.path.normpath('proteomics/220811/uniprot_2_y1000id_' + spec_abbrev + '.csv')\n",
    "\n",
    "\n",
    "            #Load map from full name to shen identifier \n",
    "\n",
    "            spec_lookup_fname = y1000plus_dir +  os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/' + spec + '.csv')\n",
    "            spec_lookup = pd.read_csv(spec_lookup_fname, index_col=0)\n",
    "\n",
    "            shen_full_2_y1000_id = dict(zip(spec_lookup['gene_full'], spec_lookup['y1000_id']))\n",
    "\n",
    "            #Make Dictionary from pairwise scores\n",
    "            pairwise_scores_spec = pairwise_scores[spec]\n",
    "\n",
    "            sim_score_high_thresh = 75\n",
    "            sim_score_low_thresh = 60\n",
    "            sim_score_diff_thresh = 8\n",
    "\n",
    "            shen_full_2_uniprot = {}\n",
    "\n",
    "            f_log.write('Species: ' + spec + '\\n')\n",
    "\n",
    "            for shen_full, scores in pairwise_scores_spec.items(): \n",
    "\n",
    "                #shen_full = 'augustus_masked-CR382121-processed-gene-0.43'\n",
    "                #scores = pairwise_scores_spec[shen_full]\n",
    "\n",
    "                scores_sorted = scores.sort_values(ascending=False)\n",
    "                maxval = scores_sorted[0]\n",
    "                diff_top2_val = maxval - scores_sorted[1]\n",
    "\n",
    "                if ((maxval>sim_score_high_thresh) | ((maxval>sim_score_low_thresh) & (diff_top2_val>sim_score_diff_thresh))): \n",
    "                    uniprot_id = scores_sorted.index[0].split('|')[1]  #The id looks like tr|<ID>|<name_specabbrev>, so we only extract the middle id\n",
    "\n",
    "                else: \n",
    "                    f_log.write('No clear uniprot match: ' + shen_full)\n",
    "                    uniprot_id = None\n",
    "\n",
    "                shen_full_2_uniprot[shen_full] = uniprot_id\n",
    "\n",
    "\n",
    "\n",
    "            f_log.write('Shen IDs in selected OGs: ' + str(len(pairwise_scores_spec)) + '\\n')\n",
    "            shen_mapped = [value for value in shen_full_2_uniprot.values() if value != None]\n",
    "            f_log.write('Shen IDs mapped to uniprot: ' + str(len(shen_mapped)))\n",
    "            if len(shen_mapped)==len(set(shen_mapped)):\n",
    "                f_log.write('One shen ID mapped to multiple Uniprot IDs\\n')\n",
    "                f_log.write('Duplicate Uniprot IDs: \\n')\n",
    "                for name, count in Counter(shen_mapped).items(): \n",
    "                    if count>1: \n",
    "                        f_log.write(name + '\\n')\n",
    "\n",
    "            #check how many shen_ids are missing\n",
    "            f_log.write('total Shen IDs: ' + str(len(spec_lookup))+ '\\n')\n",
    "            assert len(set(shen_full_2_uniprot.keys()) - set(spec_lookup['gene_full']))==0, 'Some unidentifed shen ids'\n",
    "            f_log.write('Shen IDs not mapped from og score matrix: ' + str(len(set(spec_lookup['gene_full']) - set(shen_full_2_uniprot.keys())))+ '\\n')\n",
    "\n",
    "            #Check how many uniprot ids are missing\n",
    "            uniprot_ids = []\n",
    "            uniprot_fasta = SeqIO.parse(proteome_fasta_dict[spec]['uniprot'], \"fasta\")\n",
    "            for seq in uniprot_fasta: \n",
    "                name = seq.name\n",
    "                uniprot_ids.append(name.split('|')[1])\n",
    "\n",
    "            f_log.write('Total Uniprot IDs: ' +  str(len(uniprot_ids))+ '\\n')\n",
    "            assert len(set(shen_full_2_uniprot.values()) - (set(uniprot_ids) | {None}))==0, 'Some unidentifed Uniprot IDs'\n",
    "            f_log.write('Uniprot IDs not mapped from og score matrix: ' + str(len(set(uniprot_ids) - set(shen_full_2_uniprot.values())))+ '\\n')\n",
    "\n",
    "            #Check how many uniprot ids are missing from Federica's data (assert none)\n",
    "            data_uniprot_ids = combined_data.loc[combined_data['Species']==spec_abbrev, :]['Protein.Ids'] \n",
    "            f_log.write('Uniprot IDs in data: ' + str(len(set(data_uniprot_ids)))+ '\\n')\n",
    "\n",
    "            #Check how many from federica's data set map to shen ids\n",
    "            f_log.write('Uniprot IDs in data that map to OG proteins: ' + str(len(set(data_uniprot_ids) & set(shen_full_2_uniprot.values())  ))+ '\\n')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            #Make dictionary of Uniprot id to shen identifer \n",
    "\n",
    "            uniprot_id_2_y1000_id_spec = {}\n",
    "\n",
    "            with open(fname_out, 'w') as f_out: \n",
    "                for key, value in shen_full_2_uniprot.items(): \n",
    "                    if value != None: \n",
    "                        uniprot_id_2_y1000_id_spec[value] = shen_full_2_y1000_id[key]\n",
    "                        f_out.write(value + ', ' + shen_full_2_y1000_id[key] +'\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
