{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import json\n",
    "from Bio import SeqIO\n",
    "from Bio import AlignIO   #, Align\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import diverse_yeast_tools as dyt\n",
    "\n",
    "base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts/alphafold')\n",
    "divyeast_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/diverse_yeast')\n",
    "y1000plus_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools/data') + os.sep\n",
    "genomes_dir = os.path.normpath('G:/My Drive/Crick_LMS/external_data/genomes')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load main analysis file\n",
    "struct_analysis = pickle.load(open(base_dir + os.sep + os.path.normpath('Output/data/Filter_clusters.pkl'), 'rb'))\n",
    "\n",
    "#Filter_clusters was made using Nir's clustering data.  \n",
    "# Oliver said \"we have to drop C0 in Nir's data. It affects OG1380 Q07732 and OG4312 P38280 in S.cer (REF)\"\n",
    "#The old analysis file had slightly different clusters to drop. \n",
    "#struct_analysis = pickle.load(open(base_dir + os.sep + os.path.normpath('Output/data/Analysis_new_02.pkl'), 'rb'))\n",
    "#OG1004__Scer_AF-P20095-F1-model_v2 and OG1004__Scer_AF-P15938-F1-model_v2 are examples where the orthogroup divides into smaller groups. \n",
    "\n",
    "#Load original sequence file, and make dictionary of gene_id to fasta header and gene_id to peptide sequence\n",
    "#Read in Sequence File, extract all names make dict of seq_alignment sequences\n",
    "selected_proteins = SeqIO.parse(base_dir +os.sep +  'selected_proteins.fasta', 'fasta')\n",
    "selected_proteins_headers = {}\n",
    "selected_proteins_seqs = {}\n",
    "for record in selected_proteins: \n",
    "    selected_proteins_headers[record.id] = record.description\n",
    "    selected_proteins_seqs[record.id] = str(record.seq)\n",
    "\n",
    "#Load peptide sequences for model species, make dictionary from gene id to peptide sequence\n",
    "model_protein_dict = {}\n",
    "for spec_abbrev in ['Scer', 'Spom', 'Calb']: \n",
    "    model_protein_dict[spec_abbrev] = dyt.load_model_protein_dict(spec_abbrev)\n",
    "    \n",
    "\n",
    "#ID Mapping for model species\n",
    "#Lookup from gene_id to y1000_id \n",
    "gene_id_2_y1000_id = dyt.load_model_gene_id_2_y1000_id()\n",
    "\n",
    "#swissprot_id_2_gene_id = dyt.load_model_swissprot_id_2_gene_id()\n",
    "swissprot_id_2_gene_id = dyt.load_model_swissprot_id_2_gene_id()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Calb_AF-A0A1D8PMC5-F1-model_v2',\n",
       " 'REF_Scer_AF-P38840-F1-model_v2',\n",
       " 'Spom_AF-O14192-F1-model_v2',\n",
       " 'alloascoidea_hylecoeti__OG3639__0_2924',\n",
       " 'ascoidea_rubescens__OG3639__6_326',\n",
       " 'ascoidea_rubescens__OG3639__6_5407',\n",
       " 'candida_tropicalis__OG3639__30_2724',\n",
       " 'cyberlindnera_jadinii__OG3639__35_5353',\n",
       " 'cyberlindnera_jadinii__OG3639__35_5427',\n",
       " 'kazachstania_naganishii__OG3639__49_4799',\n",
       " 'kluyveromyces_lactis__OG3639__50_1047',\n",
       " 'kluyveromyces_marxianus__OG3639__51_1738',\n",
       " 'komagataella_pastoris__OG3639__52_2093',\n",
       " 'lachancea_thermotolerans__OG3639__64_114',\n",
       " 'ogataea_parapolymorpha__OG3639__104_1542',\n",
       " 'pachysolen_tannophilus__OG3639__106_4781',\n",
       " 'torulaspora_delbrueckii__OG3639__135_1978',\n",
       " 'vanderwaltozyma_polyspora__OG3639__136_4279',\n",
       " 'wickerhamomyces_anomalus__OG3639__139_4915',\n",
       " 'wickerhamomyces_anomalus__OG3639__139_5127',\n",
       " 'yHMPu5000034957_hanseniaspora_osmophila_160519__OG3639__247_2276',\n",
       " 'yarrowia_lipolytica__OG3639__144_5496',\n",
       " 'zygosaccharomyces_rouxii__OG3639__342_3711']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = list(struct_analysis.keys())[0]\n",
    "a = struct_analysis[a]['Files to be included']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter original structural alignments to remove bad structures\n",
    "#Also include metadata from original species selection fasta and add metadata for model species. \n",
    "# base_dir/selected_proteins.fasta\n",
    "#\n",
    "#\n",
    "# Make a protein fasta for each alignment from orignal sequence and with model species sequences\n",
    "#\n",
    "# Make a dictionary of lists of each protein present in each species - save as a .json\n",
    "# This json is keyed on species and has elements\n",
    "#[name_orig,gene_id,y1000_id,swissprot_id,source]\n",
    "\n",
    "species_table = pd.read_csv(base_dir + os.sep + 'species_selection.csv')\n",
    "specs = list(species_table.loc[species_table['Load']=='Y']['original_genome_id'])\n",
    "proteins_present_by_spec = {spec: [] for spec in specs}\n",
    "\n",
    "modelspec_params = {'Scer':('saccharomyces_cerevisiae'), \n",
    "                    'Calb':('candida_albicans'), \n",
    "                    'Spom':('schizosaccharomyces_pombe')\n",
    "                   }\n",
    "proteins_present_by_spec_fname = base_dir + os.sep + 'selected_protein_ids.json'\n",
    "\n",
    "\n",
    "for fasta_fname in os.listdir(base_dir + os.sep +  os.path.normpath('msas\\structural\\FASTA') + os.sep  ):\n",
    "    #fasta_fname = 'OG1004_REF_Scer_AF-P15938-F1-model_v2.FASTA'\n",
    "    fasta_fname_base = fasta_fname.split('.')[0]\n",
    "\n",
    "    seqs_to_remove = struct_analysis[fasta_fname.split('.')[0]]['Files to be excluded']\n",
    "\n",
    "    fname_struct_aln_orig = base_dir + os.sep +  os.path.normpath('msas\\structural\\FASTA\\\\' + fasta_fname)   # os.path.normpath('msas\\FILES_ogs_pep_aligned\\\\' + og + '.mfaa.mafft')\n",
    "    fname_struct_aln_filt_out = base_dir + os.sep + os.path.normpath('msas\\structural\\\\fasta_filt\\\\' + fasta_fname_base + '.struct_filt.fasta')\n",
    "    fname_proteome = base_dir + os.sep + os.path.normpath('og_sequences\\proteome\\\\' + fasta_fname_base + '.pep.fasta')\n",
    "\n",
    "    #Read in dictionary of ref name map to sequence name\n",
    "    struct_align = AlignIO.read(fname_struct_aln_orig, 'fasta')\n",
    "    with open(fname_struct_aln_filt_out , 'w') as f_out_filt:\n",
    "        with open(fname_proteome, 'w') as f_out_prot: \n",
    "            for record in struct_align: \n",
    "                structure_imported = False\n",
    "                name_orig = record.id    \n",
    "\n",
    "                #Check if name in ref sequence (S.cer, C.alb, and S. pom) + convert\n",
    "                if name_orig.split('_')[0] in set(['Scer', 'REF', 'Calb', 'Spom']):\n",
    "                    structure_imported = True\n",
    "                    if name_orig.split('_')[0] == 'REF': \n",
    "                        spec_abbrev = name_orig.split('_')[1]\n",
    "                        swissprot_id = name_orig.split('_')[2].split('-')[1]\n",
    "                    else: \n",
    "                        spec_abbrev = name_orig.split('_')[0]\n",
    "                        swissprot_id = name_orig.split('_')[1].split('-')[1]\n",
    "                    \n",
    "                    (spec) = modelspec_params[spec_abbrev]\n",
    "                    gene_id = swissprot_id_2_gene_id[spec_abbrev][swissprot_id]\n",
    "                    if spec_abbrev == 'Spom':\n",
    "                        y1000_id = 'None'\n",
    "                    else: \n",
    "                        #spec_old = spec_abbrev_dict[spec_abbrev]\n",
    "                        y1000_id = gene_id_2_y1000_id[spec_abbrev][gene_id]\n",
    "\n",
    "                else: \n",
    "                    (spec, og, y1000_id) = name_orig.split('__')\n",
    "\n",
    "\n",
    "                #Filter out seqs with bad alignments\n",
    "                if not(name_orig in seqs_to_remove): \n",
    "                    \n",
    "                    #Extract sequence from original peptide fasta\n",
    "                    \n",
    "                    if structure_imported: \n",
    "                        prot_seq = model_protein_dict[spec_abbrev][gene_id]\n",
    "                        L = len(prot_seq)\n",
    "                        header = '>' + name_orig + ' source=af2  gene_full=' + gene_id +' y1000_id=' + y1000_id + ' L=' + str(L) + '\\n' \n",
    "                        \n",
    "                        proteins_present_by_spec[spec].append((name_orig,gene_id,y1000_id,swissprot_id,'af2'))\n",
    "                    \n",
    "                        \n",
    "                    else: \n",
    "                        header_dict = {}\n",
    "                        header_raw =selected_proteins_headers[name_orig] \n",
    "                        for item in header_raw.split(' ')[1:]: \n",
    "                            key,val = item.split('=')\n",
    "                            header_dict[key] = val\n",
    "                        \n",
    "                        source = header_dict['source']\n",
    "                        if source=='uniprot':\n",
    "                            swissprot_id = header_dict['gene_full'].split('|')[1]\n",
    "                        elif source=='shen':\n",
    "                            swissprot_id = None\n",
    "\n",
    "                        proteins_present_by_spec[spec].append((name_orig,None,y1000_id,swissprot_id, source))\n",
    "            \n",
    "                        prot_seq = selected_proteins_seqs[name_orig]\n",
    "                        header = '>' + header_raw + '\\n'  \n",
    "                        \n",
    "\n",
    "                    f_out_prot.write(header)\n",
    "                    f_out_prot.write(prot_seq + '\\n')\n",
    "\n",
    "                    f_out_filt.write(header)\n",
    "                    f_out_filt.write(str(record.seq) + '\\n')\n",
    "                    \n",
    "\n",
    "\n",
    "\n",
    "# Save proteins_present_by_spec\n",
    "\n",
    "#remove_duplicates\n",
    "for spec in specs: \n",
    "    proteins_present_by_spec[spec] = list(set(proteins_present_by_spec[spec]))\n",
    "\n",
    "with open(proteins_present_by_spec_fname, 'w') as f:\n",
    "    json.dump(proteins_present_by_spec, f, sort_keys=True, indent=4 )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identifies source of coding sequences for each species\n",
    "cds_params = { \n",
    "                 'saccharomyces_cerevisiae': {'af2': genomes_dir + os.sep + os.path.normpath('saccharomyces_cerevisiae/S288C_reference_genome_R64-2-1_20150113/orf_coding_all_R64-2-1_20150113.fasta') },\n",
    "                 'schizosaccharomyces_pombe': {'af2': genomes_dir + os.sep + os.path.normpath('schizosaccharomyces_pombe/cds.fa')},\n",
    "                 'candida_albicans' : {'af2': genomes_dir + os.sep + os.path.normpath('candida_albicans/C_albicans_SC5314_A22_current_default_coding.fasta')}\n",
    "             }\n",
    "             \n",
    "    \n",
    "#shen_only                 \n",
    "shen_specs = [\n",
    "                 'kazachstania_naganishii',\n",
    "                 'geotrichum_candidum',\n",
    "                 'vanderwaltozyma_polyspora',                              \n",
    "                 'eremothecium_gossypii',\n",
    "                 'ascoidea_rubescens',\n",
    "                 'cyberlindnera_jadinii',\n",
    "                 'ogataea_parapolymorpha',\n",
    "                 'pachysolen_tannophilus',\n",
    "                 'yHMPu5000034604_sporopachydermia_lactativora_160519',\n",
    "                 'alloascoidea_hylecoeti',\n",
    "                 'candida_apicola',\n",
    "                 'yarrowia_lipolytica',\n",
    "                 'tortispora_caseinolytica',\n",
    "                 'lipomyces_starkeyi',\n",
    "                 #when/uniprot-ENA\n",
    "                 'wickerhamomyces_anomalus',\n",
    "                 'yHMPu5000034957_hanseniaspora_osmophila_160519',\n",
    "                 'candida_tropicalis',\n",
    "                 #shen/uniprot-NCBI\n",
    "                 'zygosaccharomyces_rouxii',\n",
    "                 'debaryomyces_hansenii',\n",
    "                 'lachancea_thermotolerans',\n",
    "                 'kluyveromyces_marxianus',\n",
    "                 'kluyveromyces_lactis',\n",
    "                 'komagataella_pastoris',\n",
    "                 'torulaspora_delbrueckii'\n",
    "                  ]\n",
    "\n",
    "for spec in shen_specs: \n",
    "    cds_params[spec] = {'shen':y1000plus_dir + os.sep + os.path.normpath('shen_2018_data/0_332yeast_genomes/332_genome_annotations/cds/' + spec +'.max.cds')}\n",
    "\n",
    "#uniprot-ENA\n",
    "uniprot_ena_specs = [\n",
    "                 'wickerhamomyces_anomalus',\n",
    "                 'yHMPu5000034957_hanseniaspora_osmophila_160519',\n",
    "                 'candida_tropicalis'\n",
    "                  ]\n",
    "\n",
    "for spec in uniprot_ena_specs: \n",
    "    cds_params[spec]['uniprot_ena'] = genomes_dir + os.sep + os.path.normpath('diverse_strains/ena/' + spec + '/cds.fasta')\n",
    "\n",
    "    \n",
    "#uniprot-NCBI\n",
    "uniprot_ncbi_specs = [\n",
    "                 'zygosaccharomyces_rouxii',\n",
    "                 'debaryomyces_hansenii',\n",
    "                 'lachancea_thermotolerans',\n",
    "                 'kluyveromyces_marxianus',\n",
    "                 'kluyveromyces_lactis',\n",
    "                 'komagataella_pastoris'\n",
    "                     ]\n",
    "\n",
    "for spec in uniprot_ncbi_specs: \n",
    "    cds_params[spec]['uniprot_ncbi'] = genomes_dir + os.sep + os.path.normpath('diverse_strains/ncbi_coding_seq/'+ spec + '/cds.fasta')\n",
    "                 \n",
    "#     }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load proteins_present_by_spec\n",
    "\n",
    "proteins_present_by_spec_fname = base_dir + os.sep + 'selected_protein_ids.json'\n",
    "    \n",
    "with open(proteins_present_by_spec_fname, 'r') as f:\n",
    "    proteins_present_by_spec = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#proteins_present_by_spec['kluyveromyces_lactis']\n",
    "#['REF_Scer_AF-P16661-F1-model_v2', 'YBR110W', '110_477', 'P16661']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load cds file\n",
    "#load peptide file\n",
    "#Extract appropriate proteins from cds file\n",
    "#print into a new cds file with protein name from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "af2\n",
      "saccharomyces_cerevisiae\n",
      "schizosaccharomyces_pombe\n",
      "candida_albicans\n",
      "kazachstania_naganishii\n",
      "geotrichum_candidum\n",
      "vanderwaltozyma_polyspora\n",
      "eremothecium_gossypii\n",
      "ascoidea_rubescens\n",
      "cyberlindnera_jadinii\n",
      "ogataea_parapolymorpha\n",
      "pachysolen_tannophilus\n",
      "yHMPu5000034604_sporopachydermia_lactativora_160519\n",
      "alloascoidea_hylecoeti\n",
      "candida_apicola\n",
      "yarrowia_lipolytica\n",
      "tortispora_caseinolytica\n",
      "lipomyces_starkeyi\n",
      "wickerhamomyces_anomalus\n",
      "yHMPu5000034957_hanseniaspora_osmophila_160519\n",
      "candida_tropicalis\n",
      "zygosaccharomyces_rouxii\n",
      "debaryomyces_hansenii\n",
      "lachancea_thermotolerans\n",
      "kluyveromyces_marxianus\n",
      "kluyveromyces_lactis\n",
      "komagataella_pastoris\n",
      "torulaspora_delbrueckii\n",
      "shen\n",
      "saccharomyces_cerevisiae\n",
      "schizosaccharomyces_pombe\n",
      "candida_albicans\n",
      "kazachstania_naganishii\n",
      "geotrichum_candidum\n",
      "vanderwaltozyma_polyspora\n",
      "eremothecium_gossypii\n",
      "ascoidea_rubescens\n",
      "cyberlindnera_jadinii\n",
      "ogataea_parapolymorpha\n",
      "pachysolen_tannophilus\n",
      "yHMPu5000034604_sporopachydermia_lactativora_160519\n",
      "alloascoidea_hylecoeti\n",
      "candida_apicola\n",
      "yarrowia_lipolytica\n",
      "tortispora_caseinolytica\n",
      "lipomyces_starkeyi\n",
      "wickerhamomyces_anomalus\n",
      "yHMPu5000034957_hanseniaspora_osmophila_160519\n",
      "candida_tropicalis\n",
      "zygosaccharomyces_rouxii\n",
      "debaryomyces_hansenii\n",
      "lachancea_thermotolerans\n",
      "kluyveromyces_marxianus\n",
      "kluyveromyces_lactis\n",
      "komagataella_pastoris\n",
      "torulaspora_delbrueckii\n",
      "uniprot_ncbi\n",
      "saccharomyces_cerevisiae\n",
      "schizosaccharomyces_pombe\n",
      "candida_albicans\n",
      "kazachstania_naganishii\n",
      "geotrichum_candidum\n",
      "vanderwaltozyma_polyspora\n",
      "eremothecium_gossypii\n",
      "ascoidea_rubescens\n",
      "cyberlindnera_jadinii\n",
      "ogataea_parapolymorpha\n",
      "pachysolen_tannophilus\n",
      "yHMPu5000034604_sporopachydermia_lactativora_160519\n",
      "alloascoidea_hylecoeti\n",
      "candida_apicola\n",
      "yarrowia_lipolytica\n",
      "tortispora_caseinolytica\n",
      "lipomyces_starkeyi\n",
      "wickerhamomyces_anomalus\n",
      "yHMPu5000034957_hanseniaspora_osmophila_160519\n",
      "candida_tropicalis\n",
      "zygosaccharomyces_rouxii\n",
      "debaryomyces_hansenii\n",
      "lachancea_thermotolerans\n",
      "kluyveromyces_marxianus\n",
      "kluyveromyces_lactis\n",
      "komagataella_pastoris\n",
      "torulaspora_delbrueckii\n",
      "uniprot_ena\n",
      "saccharomyces_cerevisiae\n",
      "schizosaccharomyces_pombe\n",
      "candida_albicans\n",
      "kazachstania_naganishii\n",
      "geotrichum_candidum\n",
      "vanderwaltozyma_polyspora\n",
      "eremothecium_gossypii\n",
      "ascoidea_rubescens\n",
      "cyberlindnera_jadinii\n",
      "ogataea_parapolymorpha\n",
      "pachysolen_tannophilus\n",
      "yHMPu5000034604_sporopachydermia_lactativora_160519\n",
      "alloascoidea_hylecoeti\n",
      "candida_apicola\n",
      "yarrowia_lipolytica\n",
      "tortispora_caseinolytica\n",
      "lipomyces_starkeyi\n",
      "wickerhamomyces_anomalus\n",
      "yHMPu5000034957_hanseniaspora_osmophila_160519\n",
      "candida_tropicalis\n",
      "zygosaccharomyces_rouxii\n",
      "debaryomyces_hansenii\n",
      "lachancea_thermotolerans\n",
      "kluyveromyces_marxianus\n",
      "kluyveromyces_lactis\n",
      "komagataella_pastoris\n",
      "torulaspora_delbrueckii\n"
     ]
    }
   ],
   "source": [
    "for cds_source in ['af2', 'shen','uniprot_ncbi', 'uniprot_ena']: \n",
    "    print(cds_source)\n",
    "    for spec, cds_params_spec in cds_params.items():\n",
    "        print(spec)\n",
    "        if cds_source in cds_params_spec.keys():\n",
    "            cds_out_fname = base_dir + os.sep + os.path.normpath('selected_proteins/cds/' + spec+'__'+ cds_source + '.fasta')\n",
    "            #Load CDS file\n",
    "            cds_fasta = SeqIO.parse(cds_params_spec[cds_source], 'fasta')\n",
    "            \n",
    "            #make dictionary keyed on ID of cds seq\n",
    "            cds_fasta_dict = {}\n",
    "            for record in cds_fasta:\n",
    "                cds_fasta_dict[record.id] = str(record.seq)\n",
    " \n",
    "            #for shen genomes, make lookup for y1000_id\n",
    "            #Load y1000_id gene_full lookup\n",
    "            y1000_id_table = pd.read_csv(y1000plus_dir + os.sep + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/' + spec + '.csv'))\n",
    "            y1000_id_2_gene_full = dict(zip(y1000_id_table['y1000_id'],y1000_id_table['gene_full']))            \n",
    "\n",
    "            proteins_present = proteins_present_by_spec[spec]\n",
    "            \n",
    "            with open(cds_out_fname,'w') as f_cds_out: \n",
    "                for (name_orig,gene_id,y1000_id,uniprot_id,source) in proteins_present:\n",
    "                    if cds_source=='af2': \n",
    "                        try:\n",
    "                            f_cds_out.write('>' + name_orig+ ' source=' + source + ' source_id=' + gene_id + '\\n')\n",
    "                            f_cds_out.write(cds_fasta_dict[gene_id] + '\\n')\n",
    "                        except KeyError:\n",
    "                            print(gene_id + 'not found in ' + spec + ' ' + cds_source + ' cds file')\n",
    "                    \n",
    "                    if cds_source=='shen': \n",
    "                        if source=='shen':\n",
    "                            try:\n",
    "                                gene_full_cds = y1000_id_2_gene_full[y1000_id] + '-mRNA-1'\n",
    "                                f_cds_out.write('>' + name_orig+ ' source=' + source + ' source_id=' + gene_full_cds + '\\n')\n",
    "                                f_cds_out.write(cds_fasta_dict[gene_full_cds] + '\\n')\n",
    "                            except KeyError:\n",
    "                                print(gene_full +  ' not found in ' + spec + ' ' + cds_source + ' cds file')\n",
    "                    \n",
    "                    if cds_source=='uniprot_ena':\n",
    "                        if source=='uniprot':\n",
    "                            try:\n",
    "                                f_cds_out.write('>' + name_orig + ' source=' + source + ' source_id=' + uniprot_id + '\\n')\n",
    "                                f_cds_out.write(cds_fasta_dict[uniprot_id] + '\\n')\n",
    "                            except KeyError:\n",
    "                                print(gene_full +  ' not found in ' + spec + ' ' + cds_source + ' cds file')\n",
    "\n",
    "                    if cds_source=='uniprot_ncbi':\n",
    "                        if source=='uniprot':\n",
    "                            try:\n",
    "                                f_cds_out.write('>' + name_orig + ' source=' + source + ' source_id=' + uniprot_id + '\\n')\n",
    "                                f_cds_out.write(cds_fasta_dict[uniprot_id] + '\\n')\n",
    "                            except KeyError:\n",
    "                                print(gene_full +  ' not found in ' + spec + ' ' + cds_source + ' cds file')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Build a big dictionary for selected proteins and for selected proteins\n",
    "\n",
    "#Proteins is made up of selected_proteins.fasta and selected_proteins_modelorgs.fasta\n",
    "#\n",
    "#CDS is made up of all files in \n",
    "#G:\\My Drive\\Crick_LMS\\projects\\diverse_yeasts\\alphafold\\selected_proteins\\cds\n",
    "\n",
    "# #Load original sequence file, and make dictionary of gene_id to fasta header and gene_id to peptide sequence\n",
    "# #Read in Sequence File, extract all names make dict of seq_alignment sequences\n",
    "# selected_proteins = SeqIO.parse(base_dir +os.sep +  'selected_proteins.fasta', 'fasta')\n",
    "# #selected_proteins_headers = {}\n",
    "# selected_proteins_seqs = {}\n",
    "# for record in selected_proteins: \n",
    "#     #selected_proteins_headers[record.id] = record.description\n",
    "#     selected_proteins_seqs[record.id] = str(record.seq)\n",
    "    \n",
    "# all_selected_proteins_seqs = selected_proteins_seqs.copy()\n",
    "\n",
    "# #Load selected_proteins for  model species, include in dictionary\n",
    "# model_selected_proteins = SeqIO.parse(base_dir +os.sep +  'selected_proteins_modelorgs.fasta', 'fasta')\n",
    "\n",
    "# for record in model_selected_proteins: \n",
    "#     record_metadata = record.description.split(' ')[1:]\n",
    "\n",
    "#     metadata_dict = {}\n",
    "#     for metadata_line in record_metadata: \n",
    "#         metadata_label, metadata = metadata_line.split('=')\n",
    "#         metadata_dict[metadata_label]=metadata\n",
    "        \n",
    "#     all_selected_proteins_seqs[metadata_dict['af2_id']] = str(record.seq)\n",
    "\n",
    "#Make a dictionary of coding sequences\n",
    "cds_dir = base_dir + os.sep + os.path.normpath('selected_proteins/cds') + os.sep\n",
    "\n",
    "selected_cds_seqs = {}\n",
    "\n",
    "for cds_file in os.listdir(cds_dir):\n",
    "    cds_fasta = SeqIO.parse(cds_dir + cds_file, 'fasta')\n",
    "    for record in cds_fasta: \n",
    "        selected_cds_seqs[record.id] = str(record.seq)\n",
    "\n",
    "#Iterate through each msa and make a codon and protein sequence fasta (unaligned)\n",
    "\n",
    "seq_dir = base_dir + os.sep + 'og_sequences'\n",
    "\n",
    "for protein_file in os.listdir(seq_dir + os.sep + 'proteome' ):\n",
    "    cds_file_out = seq_dir + os.sep + 'cds' + os.sep + protein_file.split('.')[0] + '.cds.fasta'\n",
    "    \n",
    "    protein_file_full = seq_dir + os.sep+ 'proteome' + os.sep + protein_file    \n",
    "    protein_file_fasta = SeqIO.parse(protein_file_full, 'fasta')\n",
    "    \n",
    "    with open(cds_file_out,'w') as f_out_cds: \n",
    "        for record in protein_file_fasta: \n",
    "            f_out_cds.write(record.id + '\\n')\n",
    "            f_out_cds.write(selected_cds_seqs[record.id] + '\\n')\n",
    "            \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Model species info\n",
    "\n",
    "#C. albicans\n",
    "#http://www.candidagenome.org/download/sequence/C_albicans_SC5314/Assembly22/current/  \n",
    "# C_albicans_SC5314_A22_current_default_coding.fasta\n",
    "# C_albicans_SC5314_A22_current_default_protein.fasta\n",
    "#downloaded on 20221012 and stored in\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\Candida_albicans\n",
    "\n",
    "\n",
    "#S. cerevisiae\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\Saccharomyces_cerevisiae\\S288C_reference_genome_R64-2-1_20150113\n",
    "#orf_coding_all_R64-2-1_20150113.fasta\n",
    "#orf_trans_all_R64-2-1_20150113.fasta\n",
    "\n",
    "#S. pombe\n",
    "#From https://www.pombase.org/data/genome_sequence_and_features/feature_sequences/ on 20221012\n",
    "#cds.fa \n",
    "#peptide.fa  \n",
    "#both last modified 2022-10-12 03:26\n",
    "#Stored at\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\Schizosaccharomyces_pombe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make a Fasta of structures for S.pom, S.cer, and C. alb that we need Kcat/Km from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import og_metadata"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
