{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import json\n",
    "from Bio import SeqIO\n",
    "from Bio import AlignIO   #, Align\n",
    "import shutil\n",
    "from ete3 import Tree, SeqMotifFace, TreeStyle, add_face_to_node, RectFace, NodeStyle, TextFace, AttrFace\n",
    "from collections import Counter\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import diverse_yeast_tools as dyt\n",
    "\n",
    "base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts/alphafold')\n",
    "divyeast_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/diverse_yeast')\n",
    "y1000plus_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools/data') + os.sep\n",
    "genomes_dir = os.path.normpath('G:/My Drive/Crick_LMS/external_data/genomes')\n",
    "aln_dir = base_dir + os.sep + os.path.normpath('msas/structural/tm_align_4')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spom: No official name, systematic_id or synonym for matmc_2 y1000_id=121_134\n",
      "Spom: No official name, systematic_id or synonym for SPCC622.05 y1000_id=121_2159\n",
      "Spom: No official name, systematic_id or synonym for SPAC4H3.12c y1000_id=121_2282\n",
      "Spom: No official name, systematic_id or synonym for meu1-1-1 y1000_id=121_2287\n",
      "Spom: No official name, systematic_id or synonym for sab14 y1000_id=121_2434\n",
      "Spom: No official name, systematic_id or synonym for SPBC36.13 y1000_id=121_2778\n",
      "Spom: No official name, systematic_id or synonym for SPAC1F8.09c.1 y1000_id=121_3125\n",
      "Spom: No official name, systematic_id or synonym for SPBC32F12.17 y1000_id=121_3368\n",
      "Spom: No official name, systematic_id or synonym for SPBC1685.12c y1000_id=121_3533\n",
      "Spom: No official name, systematic_id or synonym for SPCC417.04 y1000_id=121_3584\n",
      "Spom: No official name, systematic_id or synonym for SPMTR.04 y1000_id=121_3769\n",
      "Spom: No official name, systematic_id or synonym for SPMTR.03 y1000_id=121_3787\n",
      "Spom: No official name, systematic_id or synonym for SPCC188.05 y1000_id=121_3839\n",
      "Spom: No official name, systematic_id or synonym for chr3 y1000_id=121_3849\n",
      "Spom: No official name, systematic_id or synonym for ppr1.2 y1000_id=121_3929\n",
      "Spom: No official name, systematic_id or synonym for Tf2-10-pseudo y1000_id=121_4201\n",
      "Spom: No official name, systematic_id or synonym for SPAC27E2.13 y1000_id=121_4235\n",
      "Spom: No official name, systematic_id or synonym for matmc_1 y1000_id=121_426\n",
      "Spom: No official name, systematic_id or synonym for rgg8 y1000_id=121_4275\n",
      "Spom: No official name, systematic_id or synonym for ppr1.1 y1000_id=121_4515\n",
      "Spom: No official name, systematic_id or synonym for Tf2-13-pseudo y1000_id=121_4528\n",
      "Spom: No official name, systematic_id or synonym for mat1-m y1000_id=121_4547\n",
      "Spom: No official name, systematic_id or synonym for meu1-1-2 y1000_id=121_4559\n",
      "Spom: No official name, systematic_id or synonym for SPBC354.11c y1000_id=121_4585\n",
      "Spom: No official name, systematic_id or synonym for mat3-m y1000_id=121_4774\n",
      "Spom: No official name, systematic_id or synonym for sab10 y1000_id=121_4780\n",
      "Spom: No official name, systematic_id or synonym for chr1 y1000_id=121_4846\n",
      "Spom: No official name, systematic_id or synonym for SPCC1672.14 y1000_id=121_4989\n",
      "Spom: No official name, systematic_id or synonym for rpl35b y1000_id=121_521\n",
      "Spom: No official name, systematic_id or synonym for SPCC794.16 y1000_id=121_547\n",
      "Spom: No official name, systematic_id or synonym for SPAC1D4.07c y1000_id=121_640\n",
      "Spom: No official name, systematic_id or synonym for B22918-2 y1000_id=121_710\n",
      "gene_full not mapping to name that maps to systematic id SPAC110.05 systematic_id=SPAC110.06\n",
      "Spom: No official name, systematic_id or synonym for chr2 y1000_id=121_891\n",
      "Spom: No official name, systematic_id or synonym for tif35 y1000_id=121_937\n",
      "gene_full not mapping to name that maps to systematic id SPBC713.13 systematic_id=SPBC713.14c\n"
     ]
    }
   ],
   "source": [
    "#Load main analysis file\n",
    "struct_analysis = pickle.load(open(base_dir + os.sep + os.path.normpath('Output/data/Filter_clusters.pkl'), 'rb'))\n",
    "\n",
    "#Filter_clusters was made using Nir's clustering data.  \n",
    "# Oliver said \"we have to drop C0 in Nir's data. It affects OG1380 Q07732 and OG4312 P38280 in S.cer (REF)\"\n",
    "#The old analysis file had slightly different clusters to drop. \n",
    "#struct_analysis = pickle.load(open(base_dir + os.sep + os.path.normpath('Output/data/Analysis_new_02.pkl'), 'rb'))\n",
    "#OG1004__Scer_AF-P20095-F1-model_v2 and OG1004__Scer_AF-P15938-F1-model_v2 are examples where the orthogroup divides into smaller groups. 1603 \n",
    "#While making trees, found that OG1603 P25632 only had 3 sequences (A pombe and Cerevisiae sequence)\n",
    "\n",
    "\n",
    "\n",
    "#Load original sequence file, and make dictionary of gene_id to fasta header and gene_id to peptide sequence\n",
    "#Read in Sequence File, extract all names make dict of seq_alignment sequences\n",
    "selected_proteins = SeqIO.parse(base_dir +os.sep +  'selected_proteins.fasta', 'fasta')\n",
    "selected_proteins_headers = {}\n",
    "selected_proteins_seqs = {}\n",
    "for record in selected_proteins: \n",
    "    selected_proteins_headers[record.id] = record.description\n",
    "    selected_proteins_seqs[record.id] = str(record.seq)\n",
    "\n",
    "#Load peptide sequences for model species, make dictionary from gene id to peptide sequence\n",
    "model_protein_dict = {}\n",
    "for spec_abbrev in ['Scer', 'Spom', 'Calb']: \n",
    "    model_protein_dict[spec_abbrev] = dyt.load_model_protein_dict(spec_abbrev)\n",
    "    \n",
    "\n",
    "#ID Mapping for model species\n",
    "#Lookup from gene_id to y1000_id \n",
    "gene_id_2_y1000_id = dyt.load_model_gene_id_2_y1000_id()\n",
    "\n",
    "#swissprot_id_2_gene_id = dyt.load_model_swissprot_id_2_gene_id()\n",
    "swissprot_id_2_gene_id = dyt.load_model_swissprot_id_2_gene_id()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We want to filter these out:   \n",
    "#OG1380_REF_Scer_AF-Q07732-F1-model_v2\n",
    "\n",
    "This is from OG1380 which had two reference structures\n",
    "\n",
    "110_4710, YDL238C, Q07729, GUD1\n",
    "110_1500, YDL239C, Q07732, ADY3\n",
    "\n",
    "110_1500 (Q07732) didn't align to any other structures in the orthogroup and I think that one was wrongly assigned.  Also it's paralog, YNL225C, CNM67, 110_3851 wasn't include in the orthogroup which is another red flag.  Also a bit strange that the two genes are next to each other on the genome, and that they are separated in the list of orthoMCL output (see below)\n",
    "\n",
    "we keep: OG1380_REF_Scer_AF-Q07729-F1-model_v2\n",
    "\n",
    "\n",
    "#OG4312_REF_Scer_AF-P38280-F1-model_v2\n",
    "\n",
    "This is a similar situation: \n",
    "\n",
    "110_5464, YBR148W, P38280,  YSW1 \n",
    "110_2373, YBR149W, P38115,  ARA1\n",
    "\n",
    "YS1 also has a paralog, (SPO21, \n",
    "YOL091W,110_5868) which doesn't appear in the orthogroup.  \n",
    "\n",
    "We keep  OG4312_REF_Scer_AF-P38115-F1-model_v2\n",
    "\n",
    "Here are the full orthogroups: \n",
    "\n",
    "\n",
    "OrthoMCL output \n",
    "OG1380: 0_7979 100_4253 101_3817 102_6062 103_5437 104_348 105_1180 106_5111 107_3454 108_2387 109_4239 10_2835 110_4710 111_1387 112_1650 113_718 114_4120 115_4109 116_4617 117_3524 118_3724 119_4044 11_2137 120_5835 121_2481 122_4884 123_5112 124_2101 125_3038 126_2404 127_4095 128_109 129_1560 12_1581 130_3367 131_1184 133_455 134_1588 135_3929 136_2960 137_2455 138_145 139_5439 13_2807 140_498 141_201 142_6286 143_5626 144_4382 145_4265 146_727 147_505 148_1666 149_2724 14_4430 150_1853 151_1750 152_1866 153_4807 154_5081 155_4631 156_4181 157_57 158_4406 159_3477 15_3825 160_1725 161_1345 162_4652 162_7685 163_3 164_1136 165_502 166_1085 167_3652 168_554 169_1185 16_442 170_976 171_2915 172_569 173_3649 174_2546 175_880 176_2935 177_4190 178_1430 179_1349 17_4470 180_1014 181_1221 182_1868 183_3094 184_2142 185_2479 186_7321 187_2997 188_3855 189_2713 18_4780 190_3057 191_11 192_1922 193_1953 194_5564 195_6999 196_3875 197_1074 198_1333 199_1654 19_2192 1_2320 200_5575 201_5015 202_3724 203_3378 204_3646 205_1851 206_1466 207_4698 208_866 209_1593 20_5171 210_5311 211_280 212_7439 213_1477 214_4903 215_6975 216_7879 217_1061 218_2201 219_612 21_3250 220_3150 221_1087 222_4980 223_3153 223_4919 224_3212 225_989 226_988 227_1061 228_288 229_3297 22_1871 230_5045 231_1857 232_1487 233_3318 233_643 234_1770 235_4216 236_1754 237_765 238_2593 238_536 239_3845 23_1723 240_4053 241_2332 242_3797 243_1657 244_422 245_1161 245_5322 246_12140 246_1257 246_1672 247_3550 248_946 249_3487 24_4840 250_1941 251_2697 252_2887 253_4804 254_109 255_4340 256_224 257_7103 258_36 259_2658 25_413 260_4371 261_2906 262_3266 263_356 264_2359 265_5094 266_3343 267_5453 268_1634 269_1392 26_2796 270_551 271_868 272_2471 273_2713 274_1637 275_1926 276_2162 277_3080 278_1254 279_5551 27_4285 280_5432 281_1196 283_4835 284_1046 285_3881 286_3199 287_1328 288_4607 289_4174 28_4523 290_626 292_4358 293_2170 294_3509 295_4648 296_4410 297_2892 298_1950 299_720 29_362 2_1802 300_2514 301_4895 302_1651 303_870 304_4414 305_3871 306_4543 307_2904 308_2445 309_218 30_5659 310_5143 311_1972 312_5610 313_347 314_3162 315_69 316_4515 317_5489 318_4838 319_1755 31_2778 31_3568 320_3083 321_2765 322_2192 323_86 324_707 325_5308 326_5838 327_1459 328_1140 329_3023 32_296 330_4719 331_3826 332_6309 333_2654 334_2779 335_627 336_4011 337_583 338_1331 339_4690 33_5843 340_5308 341_3165 342_1249 34_2841 35_4891 36_2559 37_2926 3_5729 42_1786 43_4147 44_1489 45_1613 46_3255 47_3991 49_2833 4_1857 50_1856 51_3157 52_80 53_47 54_2330 55_972 56_2062 57_1797 58_2309 59_4800 5_4710 60_906 61_920 62_4929 63_3584 64_2292 65_4669 66_7500 67_2789 69_664 6_1150 6_3153 6_5015 70_4894 71_1931 72_559 73_1395 74_2269 75_1603 76_3820 77_3910 78_5100 78_5836 79_3935 80_2013 81_3193 82_3677 83_2710 84_966 85_4866 86_5853 87_4990 88_3268 89_3822 8_6737 90_4378 91_3509 92_4498 93_1615 94_902 95_3852 96_3202 97_1100 98_3151 99_2311 9_4542 22_1437 290_403 172_187 15_1499 228_234 230_650 234_1916 103_4451 160_4492 189_1353 226_883 229_5558 231_2640 233_5082 235_2552 236_3821 237_637 238_4636 241_4669 284_3996 53_1753 53_328 110_1500 114_2713 232_3250 190_5056\n",
    "\n",
    "OG4312: 105_2440 15_5119 193_1462 233_2728 234_2443 235_973 240_5182 243_3685 277_2923 313_3735 100_271 101_2850 103_1114 106_5303 107_3045 109_1518 110_2373 111_676 112_1530 113_1707 113_2131 114_3654 115_4038 116_2995 119_2420 11_2509 120_3094 123_2102 123_2865 124_6051 127_5480 128_3721 12_3502 137_1571 139_6065 140_3429 146_1919 147_4561 148_1207 149_5134 14_4455 150_870 151_3741 152_537 153_2758 154_1802 155_5253 156_1145 158_34 159_71 160_126 161_5979 162_6883 163_3900 164_1861 165_2447 166_1617 168_457 16_1928 170_3358 172_4044 173_4599 176_3544 17_3425 180_509 181_1646 182_3615 183_4858 184_480 185_3280 187_1410 189_2923 18_147 191_4898 194_2814 198_5467 1_972 208_880 217_814 219_2709 224_3967 225_2967 226_1456 226_4544 228_1171 229_4717 230_5059 231_1276 232_1991 233_480 235_6 236_2319 237_715 23_3815 241_1711 244_4654 245_5483 246_9337 247_3900 24_2220 250_2452 251_4498 252_2172 254_2371 255_5001 256_5031 257_845 258_3257 259_195 260_1573 261_2122 262_1620 262_737 263_4387 264_4299 265_3620 266_1967 267_2120 268_1817 269_1055 270_2837 271_3037 273_2193 273_4190 274_1228 275_3604 276_828 27_1445 280_1353 283_1624 284_1717 285_3498 285_5154 286_3837 286_4325 287_2223 288_900 289_2400 28_3994 290_5258 293_3141 294_1897 295_988 299_4107 29_604 2_1778 2_74 304_4300 306_5473 307_3035 308_1569 309_990 30_5633 311_2849 312_5969 314_5133 315_3625 316_4563 317_2510 318_4080 319_1452 320_4268 321_3993 323_3201 325_4065 326_5211 327_4234 328_122 329_5516 32_2325 330_277 334_4167 335_864 336_3847 338_4948 340_3541 341_2212 342_2290 34_1272 35_3004 36_2114 37_2895 44_2343 45_2204 46_371 47_3261 48_2133 49_2789 50_199 51_4613 53_621 55_2664 56_4914 57_2283 59_1088 60_2628 61_144 62_590 63_58 64_1023 65_846 67_3819 91_4700 92_1006 99_1227 9_369 139_1772 142_2792 144_2332 172_1273 196_4330 201_442 241_2921 244_4204 257_4200 292_3284 292_735 296_5543 297_3403 340_1334 140_4847 160_2176 162_6036 162_6042 162_6389 230_840 234_522 245_4850 246_6737 246_9401 268_2639 53_3165 53_3434 201_3037 162_2583 245_6684 239_5185 258_3763 336_5209 46_2479 58_1410 99_412 127_4388 203_3682 245_923 27_3713 140_4831 43_605 109_2478 110_5464 111_4365 112_1416 143_5447 277_6609 202_1342 126_6065 124_2222 233_2217 9_2186 144_2569 253_2721 199_2141 162_7100 226_5057\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###This step seems unnecessary - could just go from original fasta file? Also filtering of those OGs is done at the clustering step.  \n",
    "\n",
    "##Instead of cycling through the proteins that came out of Oliver's runs, can cycle over the original FASTAs I provided. \n",
    "#This will mean that there will be some included whose sequences did not get calculated, but that will get filtered out after the alignment step. \n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['saccharomyces_cerevisiae',\n",
       " 'debaryomyces_hansenii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'lachancea_thermotolerans',\n",
       " 'kazachstania_naganishii',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'yHMPu5000034957_hanseniaspora_osmophila_160519',\n",
       " 'geotrichum_candidum',\n",
       " 'zygosaccharomyces_rouxii',\n",
       " 'candida_albicans',\n",
       " 'candida_tropicalis',\n",
       " 'kluyveromyces_lactis',\n",
       " 'komagataella_pastoris',\n",
       " 'schizosaccharomyces_pombe',\n",
       " 'vanderwaltozyma_polyspora',\n",
       " 'torulaspora_delbrueckii',\n",
       " 'eremothecium_gossypii',\n",
       " 'ascoidea_rubescens',\n",
       " 'cyberlindnera_jadinii',\n",
       " 'ogataea_parapolymorpha',\n",
       " 'pachysolen_tannophilus',\n",
       " 'yHMPu5000034604_sporopachydermia_lactativora_160519',\n",
       " 'alloascoidea_hylecoeti',\n",
       " 'candida_apicola',\n",
       " 'yarrowia_lipolytica',\n",
       " 'tortispora_caseinolytica',\n",
       " 'lipomyces_starkeyi']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make proteins_present_by_spec JSON file listing proteins we want to consider for each species. \n",
    "# This json is keyed on species and has elements\n",
    "#[name_orig,gene_id,y1000_id,swissprot_id,source]\n",
    "\n",
    "species_table = pd.read_csv(base_dir + os.sep + os.path.normpath('selected_proteins/species_selection/species_selection.csv'))\n",
    "specs = list(species_table.loc[species_table['Load']=='Y']['original_genome_id'])\n",
    "proteins_present_by_spec = {spec: [] for spec in specs}\n",
    "\n",
    "modelspec_params = {'Scer':('saccharomyces_cerevisiae'), \n",
    "                    'Calb':('candida_albicans'), \n",
    "                    'Spom':('schizosaccharomyces_pombe')\n",
    "                   }\n",
    "\n",
    "proteins_present_by_spec_fname = base_dir + os.sep + os.path.normpath('selected_proteins/selected_protein_ids.json')\n",
    "\n",
    "\n",
    "#Cycle through selected_proteins_modelorgs.fasta: \n",
    "\n",
    "selected_proteins = SeqIO.parse(base_dir + os.sep + os.path.normpath('selected_proteins/selected_proteins.fasta'), \"fasta\")\n",
    "\n",
    "for record in selected_proteins: \n",
    "\n",
    "    (spec, og, y1000_id) = record.id.split('__')\n",
    "\n",
    "    header_dict = {}\n",
    "    for item in record.description.split(' ')[1:]: \n",
    "        key,val = item.split('=')\n",
    "        header_dict[key] = val\n",
    "\n",
    "    source = header_dict['source']\n",
    "    if source=='uniprot':\n",
    "        swissprot_id = header_dict['gene_full'].split('|')[1]\n",
    "    elif source=='shen':\n",
    "        swissprot_id = None\n",
    "\n",
    "    proteins_present_by_spec[spec].append((record.id,None,y1000_id,swissprot_id, source))\n",
    "    \n",
    "selected_proteins_modelorgs = SeqIO.parse(base_dir + os.sep + os.path.normpath('selected_proteins/selected_proteins_modelorgs.fasta'), \"fasta\")\n",
    "\n",
    "for record in selected_proteins_modelorgs: \n",
    "\n",
    "    gene_id = record.id\n",
    "    header_dict = {}\n",
    "    for item in record.description.split(' ')[1:]: \n",
    "        key,val = item.split('=')\n",
    "        header_dict[key] = val\n",
    "    \n",
    "    proteins_present_by_spec[header_dict['species']].append(( header_dict['af2_id'],gene_id,header_dict['y1000_id'], header_dict['uniprot_id'],'af2'))\n",
    "\n",
    "\n",
    "with open(proteins_present_by_spec_fname, 'w') as f:\n",
    "    json.dump(proteins_present_by_spec, f, sort_keys=True, indent=4 )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'og': 'OG2365',\n",
       " 'species': 'candida_albicans',\n",
       " 'uniprot_id': 'Q9Y872',\n",
       " 'af2_id': 'Calb_AF-Q9Y872-F1-model_v2',\n",
       " 'y1000_id': '12_1859'}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This might be useful to do Oliver's structural alignments, but not useful for the proteins_present_by_spec json, as that is calculated from original data above\n",
    "\n",
    "# #Filter original structural alignments to remove poorly aligned structures.  \n",
    "# #Also include metadata from original species selection fasta and add metadata for model species. \n",
    "# # base_dir/selected_proteins.fasta\n",
    "# #\n",
    "# #\n",
    "# # Make a protein fasta for each alignment from orignal sequence and with model species sequences\n",
    "# #\n",
    "# # Make a dictionary of lists of each protein present in each species - save as a .json\n",
    "# # This json is keyed on species and has elements\n",
    "# #[name_orig,gene_id,y1000_id,swissprot_id,source]\n",
    "\n",
    "# #Filter based on above errors to Orthogroup Assignment\n",
    "# og_ref_filter_list = ['OG1380_REF_Scer_AF-Q07732-F1-model_v2','OG4312_REF_Scer_AF-P38280-F1-model_v2' ]\n",
    "\n",
    "# species_table = pd.read_csv(base_dir + os.sep + 'species_selection.csv')\n",
    "# specs = list(species_table.loc[species_table['Load']=='Y']['original_genome_id'])\n",
    "# proteins_present_by_spec = {spec: [] for spec in specs}\n",
    "\n",
    "# modelspec_params = {'Scer':('saccharomyces_cerevisiae'), \n",
    "#                     'Calb':('candida_albicans'), \n",
    "#                     'Spom':('schizosaccharomyces_pombe')\n",
    "#                    }\n",
    "# proteins_present_by_spec_fname = base_dir + os.sep + 'selected_protein_ids.json'\n",
    "\n",
    "\n",
    "# for fasta_fname in os.listdir(base_dir + os.sep +  os.path.normpath('msas\\structural\\FASTA') + os.sep  ):\n",
    "#     #fasta_fname = 'OG1004_REF_Scer_AF-P15938-F1-model_v2.FASTA'\n",
    "    \n",
    "#     if not(fasta_fname in [name + '.FASTA' for name in og_ref_filter_list]): \n",
    "    \n",
    "#         fasta_fname_base = fasta_fname.split('.')[0]\n",
    "\n",
    "#         seqs_to_include = struct_analysis[fasta_fname.split('.')[0]]['Files to be included']\n",
    "\n",
    "#         fname_struct_aln_orig = base_dir + os.sep +  os.path.normpath('msas\\structural\\FASTA\\\\' + fasta_fname)   # os.path.normpath('msas\\FILES_ogs_pep_aligned\\\\' + og + '.mfaa.mafft')\n",
    "#         fname_struct_aln_filt_out = base_dir + os.sep + os.path.normpath('msas\\structural\\\\fasta_filt\\\\' + fasta_fname_base + '.struct_filt.fasta')\n",
    "#         fname_proteome = base_dir + os.sep + os.path.normpath('og_sequences\\proteome\\\\' + fasta_fname_base + '.pep.fasta')\n",
    "\n",
    "#         #Read in dictionary of ref name map to sequence name\n",
    "#         struct_align = AlignIO.read(fname_struct_aln_orig, 'fasta')\n",
    "#         with open(fname_struct_aln_filt_out , 'w') as f_out_filt:\n",
    "#             with open(fname_proteome, 'w') as f_out_prot: \n",
    "#                 for record in struct_align: \n",
    "#                     structure_imported = False\n",
    "#                     name_orig = record.id    \n",
    "\n",
    "#                     #Check if name in ref sequence (S.cer, C.alb, and S. pom) + convert\n",
    "#                     if name_orig.split('_')[0] in set(['Scer', 'REF', 'Calb', 'Spom']):\n",
    "#                         structure_imported = True\n",
    "#                         if name_orig.split('_')[0] == 'REF': \n",
    "#                             spec_abbrev = name_orig.split('_')[1]\n",
    "#                             swissprot_id = name_orig.split('_')[2].split('-')[1]\n",
    "#                         else: \n",
    "#                             spec_abbrev = name_orig.split('_')[0]\n",
    "#                             swissprot_id = name_orig.split('_')[1].split('-')[1]\n",
    "\n",
    "#                         (spec) = modelspec_params[spec_abbrev]\n",
    "#                         gene_id = swissprot_id_2_gene_id[spec_abbrev][swissprot_id]\n",
    "#                         if spec_abbrev == 'Spom':\n",
    "#                             y1000_id = 'None'\n",
    "#                         else: \n",
    "#                             #spec_old = spec_abbrev_dict[spec_abbrev]\n",
    "#                             y1000_id = gene_id_2_y1000_id[spec_abbrev][gene_id]\n",
    "\n",
    "#                     else: \n",
    "#                         (spec, og, y1000_id) = name_orig.split('__')\n",
    "\n",
    "\n",
    "#                     #Filter by structural clusters within OGs based on Nir's analysis\n",
    "#                     if name_orig in seqs_to_include: \n",
    "\n",
    "#                         #Extract sequence from original peptide fasta\n",
    "\n",
    "#                         if structure_imported: \n",
    "#                             prot_seq = model_protein_dict[spec_abbrev][gene_id]\n",
    "#                             L = len(prot_seq)\n",
    "#                             header = '>' + name_orig + ' source=af2  gene_full=' + gene_id +' y1000_id=' + y1000_id + ' L=' + str(L) + '\\n' \n",
    "\n",
    "#                             proteins_present_by_spec[spec].append((name_orig,gene_id,y1000_id,swissprot_id,'af2'))\n",
    "\n",
    "\n",
    "#                         else: \n",
    "#                             header_dict = {}\n",
    "#                             header_raw =selected_proteins_headers[name_orig] \n",
    "#                             for item in header_raw.split(' ')[1:]: \n",
    "#                                 key,val = item.split('=')\n",
    "#                                 header_dict[key] = val\n",
    "\n",
    "#                             source = header_dict['source']\n",
    "#                             if source=='uniprot':\n",
    "#                                 swissprot_id = header_dict['gene_full'].split('|')[1]\n",
    "#                             elif source=='shen':\n",
    "#                                 swissprot_id = None\n",
    "\n",
    "#                             proteins_present_by_spec[spec].append((name_orig,None,y1000_id,swissprot_id, source))\n",
    "\n",
    "#                             prot_seq = selected_proteins_seqs[name_orig]\n",
    "#                             header = '>' + header_raw + '\\n'  \n",
    "\n",
    "\n",
    "#                         f_out_prot.write(header)\n",
    "#                         f_out_prot.write(prot_seq + '\\n')\n",
    "\n",
    "#                         f_out_filt.write(header)\n",
    "#                         f_out_filt.write(str(record.seq) + '\\n')\n",
    "                    \n",
    "\n",
    "\n",
    "\n",
    "# # Save proteins_present_by_spec\n",
    "\n",
    "# #remove_duplicates\n",
    "# for spec in specs: \n",
    "#     proteins_present_by_spec[spec] = list(set(proteins_present_by_spec[spec]))\n",
    "\n",
    "# with open(proteins_present_by_spec_fname, 'w') as f:\n",
    "#     json.dump(proteins_present_by_spec, f, sort_keys=True, indent=4 )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identifies source of coding sequences for each species\n",
    "cds_params = { \n",
    "                 'saccharomyces_cerevisiae': {'modeldb': genomes_dir + os.sep + os.path.normpath('saccharomyces_cerevisiae/S288C_reference_genome_R64-2-1_20150113/orf_coding_all_R64-2-1_20150113.fasta') },\n",
    "                 #'schizosaccharomyces_pombe': {'af2': genomes_dir + os.sep + os.path.normpath('schizosaccharomyces_pombe/cds.fa')},\n",
    "                 'candida_albicans' : {'modeldb': genomes_dir + os.sep + os.path.normpath('candida_albicans/C_albicans_SC5314_A22_current_default_coding.fasta')}\n",
    "             }\n",
    "             \n",
    "    \n",
    "#shen_only                 \n",
    "shen_specs = [\n",
    "                 'kazachstania_naganishii',\n",
    "                 'geotrichum_candidum',\n",
    "                 'vanderwaltozyma_polyspora',                              \n",
    "                 'eremothecium_gossypii',\n",
    "                 'ascoidea_rubescens',\n",
    "                 'cyberlindnera_jadinii',\n",
    "                 'ogataea_parapolymorpha',\n",
    "                 'pachysolen_tannophilus',\n",
    "                 'yHMPu5000034604_sporopachydermia_lactativora_160519',\n",
    "                 'alloascoidea_hylecoeti',\n",
    "                 'candida_apicola',\n",
    "                 'yarrowia_lipolytica',\n",
    "                 'tortispora_caseinolytica',\n",
    "                 'lipomyces_starkeyi',\n",
    "                 #shen/uniprot_ncbi_genome\n",
    "                 'wickerhamomyces_anomalus',\n",
    "                 'yHMPu5000034957_hanseniaspora_osmophila_160519',\n",
    "                 'candida_tropicalis',\n",
    "                 #shen/uniprot_ncbi\n",
    "                 'zygosaccharomyces_rouxii',\n",
    "                 'debaryomyces_hansenii',\n",
    "                 'lachancea_thermotolerans',\n",
    "                 'kluyveromyces_marxianus',\n",
    "                 'kluyveromyces_lactis',\n",
    "                 'komagataella_pastoris',\n",
    "                 'torulaspora_delbrueckii'\n",
    "                  ]\n",
    "\n",
    "for spec in shen_specs: \n",
    "    cds_params[spec] = {'shen':y1000plus_dir + os.sep + os.path.normpath('shen_2018_data/0_332yeast_genomes/332_genome_annotations/cds/' + spec +'.max.cds')}\n",
    "\n",
    "#uniprot_ncbi_genome\n",
    "uniprot_ncbi_genome_specs = [\n",
    "                 'wickerhamomyces_anomalus',\n",
    "                 'yHMPu5000034957_hanseniaspora_osmophila_160519',\n",
    "                 'candida_tropicalis'\n",
    "                  ]\n",
    "\n",
    "for spec in uniprot_ncbi_genome_specs: \n",
    "    cds_params[spec]['uniprot_ncbi_genome'] = genomes_dir + os.sep + os.path.normpath('diverse_yeast/ncbi_genomes_coding_seq/' + spec + '/cds.fasta')\n",
    "\n",
    "    \n",
    "#uniprot_ncbi\n",
    "uniprot_ncbi_specs = [\n",
    "                 'zygosaccharomyces_rouxii',\n",
    "                 'debaryomyces_hansenii',\n",
    "                 'lachancea_thermotolerans',\n",
    "                 'kluyveromyces_marxianus',\n",
    "                 'kluyveromyces_lactis',\n",
    "                 'komagataella_pastoris', \n",
    "                 'schizosaccharomyces_pombe'\n",
    "                     ]\n",
    "\n",
    "for spec in uniprot_ncbi_specs: \n",
    "    if spec in cds_params.keys():\n",
    "        cds_params[spec]['uniprot_ncbi'] = genomes_dir + os.sep + os.path.normpath('diverse_yeast/ncbi_coding_seq/'+ spec + '/cds.fasta')\n",
    "    else: \n",
    "        cds_params[spec] = {'uniprot_ncbi': genomes_dir + os.sep + os.path.normpath('diverse_yeast/ncbi_coding_seq/'+ spec + '/cds.fasta')}\n",
    "#     }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load proteins_present_by_spec\n",
    "\n",
    "proteins_present_by_spec_fname = base_dir + os.sep + os.path.normpath('selected_proteins/selected_protein_ids.json')\n",
    "    \n",
    "with open(proteins_present_by_spec_fname, 'r') as f:\n",
    "    proteins_present_by_spec = json.load(f) \n",
    "    \n",
    "#proteins_present_by_spec['kluyveromyces_lactis']\n",
    "#['REF_Scer_AF-P16661-F1-model_v2', 'YBR110W', '110_477', 'P16661']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "modeldb\n",
      "saccharomyces_cerevisiae\n",
      "candida_albicans\n",
      "shen\n",
      "kazachstania_naganishii\n",
      "geotrichum_candidum\n",
      "vanderwaltozyma_polyspora\n",
      "eremothecium_gossypii\n",
      "ascoidea_rubescens\n",
      "cyberlindnera_jadinii\n",
      "ogataea_parapolymorpha\n",
      "pachysolen_tannophilus\n",
      "yHMPu5000034604_sporopachydermia_lactativora_160519\n",
      "alloascoidea_hylecoeti\n",
      "candida_apicola\n",
      "yarrowia_lipolytica\n",
      "tortispora_caseinolytica\n",
      "lipomyces_starkeyi\n",
      "wickerhamomyces_anomalus\n",
      "yHMPu5000034957_hanseniaspora_osmophila_160519\n",
      "candida_tropicalis\n",
      "zygosaccharomyces_rouxii\n",
      "debaryomyces_hansenii\n",
      "lachancea_thermotolerans\n",
      "kluyveromyces_marxianus\n",
      "kluyveromyces_lactis\n",
      "komagataella_pastoris\n",
      "torulaspora_delbrueckii\n",
      "uniprot_ncbi\n",
      "zygosaccharomyces_rouxii\n",
      "debaryomyces_hansenii\n",
      "lachancea_thermotolerans\n",
      "kluyveromyces_marxianus\n",
      "kluyveromyces_lactis\n",
      "komagataella_pastoris\n",
      "schizosaccharomyces_pombe\n",
      "uniprot_ncbi_genome\n",
      "wickerhamomyces_anomalus\n",
      "A0A1E3P784 not found in wickerhamomyces_anomalus uniprot_ncbi_genome cds file\n",
      "yHMPu5000034957_hanseniaspora_osmophila_160519\n",
      "candida_tropicalis\n"
     ]
    }
   ],
   "source": [
    "#Print CDS for selected proteins from each source by species\n",
    "\n",
    "for cds_source in ['modeldb', 'shen','uniprot_ncbi', 'uniprot_ncbi_genome']: \n",
    "\n",
    "    print(cds_source)\n",
    "    #cds_source = 'uniprot_ncbi'\n",
    "\n",
    "    for spec, cds_params_spec in cds_params.items():\n",
    "        #spec = 'schizosaccharomyces_pombe'\n",
    "        #cds_params_spec = cds_params[spec]\n",
    "\n",
    "        if cds_source in cds_params_spec.keys():\n",
    "            print(spec)\n",
    "            cds_out_fname = base_dir + os.sep + os.path.normpath('selected_proteins/cds/' + spec+'__'+ cds_source + '.fasta')\n",
    "            #Load CDS file\n",
    "            cds_fasta = SeqIO.parse(cds_params_spec[cds_source], 'fasta')\n",
    "\n",
    "            #make dictionary keyed on ID of cds seq\n",
    "            cds_fasta_dict = {}\n",
    "            for record in cds_fasta:\n",
    "                cds_fasta_dict[record.id] = str(record.seq)\n",
    "\n",
    "            #for shen genomes, make lookup for y1000_id\n",
    "            #Load y1000_id gene_full lookup\n",
    "            y1000_id_table = pd.read_csv(y1000plus_dir + os.sep + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/' + spec + '.csv'))\n",
    "            y1000_id_2_gene_full = dict(zip(y1000_id_table['y1000_id'],y1000_id_table['gene_full']))            \n",
    "\n",
    "            proteins_present = proteins_present_by_spec[spec]\n",
    "\n",
    "            with open(cds_out_fname,'w') as f_cds_out: \n",
    "                for (name_orig,gene_id,y1000_id,uniprot_id,source) in proteins_present:\n",
    "                    if cds_source=='modeldb': \n",
    "                        try:\n",
    "                            f_cds_out.write('>' + name_orig+ ' source=' + cds_source + 'source_id=' + gene_id + '\\n')\n",
    "                            f_cds_out.write(cds_fasta_dict[gene_id] + '\\n')\n",
    "                        except KeyError:\n",
    "                            print(gene_id + 'not found in ' + spec + ' ' + cds_source + ' cds file')\n",
    "\n",
    "                    if cds_source=='shen': \n",
    "                        if source=='shen':\n",
    "                            try:\n",
    "                                gene_full_cds = y1000_id_2_gene_full[y1000_id] + '-mRNA-1'\n",
    "                                f_cds_out.write('>' + name_orig+ ' source=' + cds_source + ' source_id=' + gene_full_cds + '\\n')\n",
    "                                f_cds_out.write(cds_fasta_dict[gene_full_cds] + '\\n')\n",
    "                            except KeyError:\n",
    "                                print(gene_full +  ' not found in ' + spec + ' ' + cds_source + ' cds file')\n",
    "\n",
    "                    if cds_source=='uniprot_ncbi_genome':\n",
    "                        if source=='uniprot':\n",
    "                            try:\n",
    "                                f_cds_out.write('>' + name_orig + ' source=' + cds_source + ' source_id=' + uniprot_id + '\\n')\n",
    "                                f_cds_out.write(cds_fasta_dict[uniprot_id] + '\\n')\n",
    "                            except KeyError:\n",
    "                                print(uniprot_id +  ' not found in ' + spec + ' ' + cds_source + ' cds file')\n",
    "\n",
    "                    if cds_source=='uniprot_ncbi':\n",
    "                        if ((source=='uniprot') | ((source=='af2') & (spec=='schizosaccharomyces_pombe'))):\n",
    "                            try:\n",
    "                                f_cds_out.write('>' + name_orig + ' source=' + cds_source + ' source_id=' + uniprot_id + '\\n')\n",
    "                                f_cds_out.write(cds_fasta_dict[uniprot_id] + '\\n')\n",
    "                            except KeyError:\n",
    "                                print(uniprot_id +  ' not found in ' + spec + ' ' + cds_source + ' cds file')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rename files for Nir's clusters to match main names\n",
    "## Build CDS alignments for Nir's clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Decided not to go with clusters.pkl that Oliver sent because that didn't match my file.  \n",
    "#To load that data, do:\n",
    "#clusters = pickle.load(open(base_dir + os.sep + os.path.normpath('msas/structural/tm_align/clustering/clusters.pkl'), 'rb'))\n",
    "#Of the small clusters there were the 14 C0 clusters.  It looks like all these were removed from this original file. \n",
    "\n",
    "#Instead starting from documents that Nir left at: https://charitede-my.sharepoint.com/personal/nir_cohen_charite_de/_layouts/15/onedrive.aspx?ga=1&id=%2Fpersonal%2Fnir%5Fcohen%5Fcharite%5Fde%2FDocuments%2Fortho%5Faligment%2Fsub%5Fortho%5Fre%5Fclustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nir's routine took 435 orthogroups as inputs\n"
     ]
    }
   ],
   "source": [
    "clusters_df = pd.read_csv(aln_dir + os.sep + os.path.normpath('clustering/clusters_0.15.csv'), index_col=0)\n",
    "#assemble dictionary of OG_cluster to list of proteins\n",
    "\n",
    "print(\"Nir's routine took \" + str(len(set(clusters_df['ortho_group']))) + \" orthogroups as inputs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ortho_group</th>\n",
       "      <th>cluster</th>\n",
       "      <th>avg_TM_score</th>\n",
       "      <th>avg_RMSD</th>\n",
       "      <th>std_TM_score</th>\n",
       "      <th>std_RMSD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG1004</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.26820</td>\n",
       "      <td>4.16</td>\n",
       "      <td>0.145947</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>OG1310</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.21235</td>\n",
       "      <td>6.32</td>\n",
       "      <td>0.096944</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>OG1380</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.16530</td>\n",
       "      <td>7.06</td>\n",
       "      <td>0.017395</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>OG1714</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.19750</td>\n",
       "      <td>6.14</td>\n",
       "      <td>0.116955</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>OG1897</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.19990</td>\n",
       "      <td>8.09</td>\n",
       "      <td>0.018950</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>OG2147</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.22440</td>\n",
       "      <td>6.35</td>\n",
       "      <td>0.122612</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>OG2646</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.19735</td>\n",
       "      <td>5.54</td>\n",
       "      <td>0.097934</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>OG2901</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.21970</td>\n",
       "      <td>6.58</td>\n",
       "      <td>0.064205</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>OG3000</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.16190</td>\n",
       "      <td>6.29</td>\n",
       "      <td>0.073963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>OG3074</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.26185</td>\n",
       "      <td>3.89</td>\n",
       "      <td>0.132300</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>OG3313</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.28240</td>\n",
       "      <td>4.14</td>\n",
       "      <td>0.141846</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>OG4312</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.20410</td>\n",
       "      <td>7.54</td>\n",
       "      <td>0.007354</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>OG5006</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.28555</td>\n",
       "      <td>3.21</td>\n",
       "      <td>0.161574</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>OG5314</td>\n",
       "      <td>C0</td>\n",
       "      <td>0.23870</td>\n",
       "      <td>4.52</td>\n",
       "      <td>0.018668</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ortho_group cluster  avg_TM_score  avg_RMSD  std_TM_score  std_RMSD\n",
       "2        OG1004      C0       0.26820      4.16      0.145947       0.0\n",
       "60       OG1310      C0       0.21235      6.32      0.096944       0.0\n",
       "77       OG1380      C0       0.16530      7.06      0.017395       0.0\n",
       "147      OG1714      C0       0.19750      6.14      0.116955       0.0\n",
       "179      OG1897      C0       0.19990      8.09      0.018950       0.0\n",
       "222      OG2147      C0       0.22440      6.35      0.122612       0.0\n",
       "284      OG2646      C0       0.19735      5.54      0.097934       0.0\n",
       "313      OG2901      C0       0.21970      6.58      0.064205       0.0\n",
       "326      OG3000      C0       0.16190      6.29      0.073963       0.0\n",
       "342      OG3074      C0       0.26185      3.89      0.132300       0.0\n",
       "364      OG3313      C0       0.28240      4.14      0.141846       0.0\n",
       "433      OG4312      C0       0.20410      7.54      0.007354       0.0\n",
       "459      OG5006      C0       0.28555      3.21      0.161574       0.0\n",
       "463      OG5314      C0       0.23870      4.52      0.018668       0.0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_df.loc[clusters_df['cluster']=='C0']["
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "435"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#N"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reference sequence removed from OG1004_C0 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG1310_C0 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG1819_C2 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG2290_C1 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG2901_C0 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG3000_C0 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG3074_C0 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG3313_C0 because it was too short. Replaced ref with first remaining sequence.\n",
      "Reference sequence removed from OG5006_C0 because it was too short. Replaced ref with first remaining sequence.\n",
      "149 sequences removed because they were less than 0.8 of the median sequence length in their cluster\n"
     ]
    }
   ],
   "source": [
    "clusters_df = pd.read_csv(aln_dir + os.sep + os.path.normpath('clustering/clusters_0.15.csv'), index_col=0)\n",
    "#assemble dictionary of OG_cluster to list of proteins\n",
    "\n",
    "cluster_msa_base = aln_dir + os.sep + os.path.normpath('fasta') + os.sep\n",
    "\n",
    "clusters = {}\n",
    "clusters_ref = {}\n",
    "clusters_len_norm = {}\n",
    "clusters_len_filt = {}\n",
    "len_filt_ratio = 0.8  #If sequences are less than 0.8 of the median length, filter them out.  \n",
    "filtered_seqs = {}\n",
    "\n",
    "for OG, clust in zip(clusters_df['ortho_group'], clusters_df['cluster']): \n",
    "    cluster_seqs = SeqIO.parse(cluster_msa_base + OG + clust + '.fa', 'fasta')\n",
    "    seq_list = []\n",
    "    N_refs = 0\n",
    "    cluster_len = []\n",
    "    for record in cluster_seqs: \n",
    "        short_seq = dyt.seq_squeeze(str(record.seq),all_outputs=False)\n",
    "        cluster_len.append(len(short_seq))\n",
    "        \n",
    "        seq_list.append(record.id)\n",
    "        if len(record.description.split('*'))==2: #The line with the * is assigned by tm_align as the reference for the alignment\n",
    "            clusters_ref[OG + '_' + clust] = record.id\n",
    "            N_refs = N_refs+1\n",
    "    clusters[OG + '_' + clust] = seq_list\n",
    "    \n",
    "    cluster_len_array = np.array(cluster_len)\n",
    "    med_len = np.median(cluster_len_array)\n",
    "    clusters_len_norm[OG + '_' + clust] = (med_len, cluster_len_array/med_len)\n",
    "    \n",
    "    \n",
    "    filtered_seqs[OG + '_' + clust] = list(np.array(clusters[OG + '_' + clust])[(clusters_len_norm[OG + '_' + clust][1]<len_filt_ratio)])\n",
    "    \n",
    "    clusters_len_filt[OG + '_' + clust] = list(np.array(clusters[OG + '_' + clust])[(clusters_len_norm[OG + '_' + clust][1]>len_filt_ratio)])\n",
    "    \n",
    "    if clusters_ref[OG + '_' + clust] in filtered_seqs[OG + '_' + clust]:\n",
    "        clusters_ref[OG+'_' + clust] = clusters_len_filt[OG+'_'+clust][0]\n",
    "        print('Reference sequence removed from ' + OG + '_' + clust + ' because it was too short. Replaced ref with first remaining sequence.')\n",
    "    \n",
    "    assert N_refs==1, 'More than one reference cluster ' + OG + '_' + clust\n",
    "\n",
    "N_removed = 0\n",
    "\n",
    "with open(aln_dir + os.sep + os.path.normpath('clustering/len_filtered_seqs.txt'), 'w') as f_out: \n",
    "    f_out.write('# Sequences filtered out because the sequence was less than ' + str(len_filt_ratio) + ' of the median length in the cluster \\n')\n",
    "    for og_clust, filtered_seq_list in filtered_seqs.items():\n",
    "        N_removed = N_removed + len(filtered_seq_list)\n",
    "        f_out.write('>' + og_clust + '\\n')\n",
    "        for filtered_seq in filtered_seq_list: \n",
    "            f_out.write(filtered_seq + '\\n')\n",
    "\n",
    "print(str(N_removed) + ' sequences removed because they were less than ' + str(len_filt_ratio) + ' of the median sequence length in their cluster')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Calb_AF-A0A1D8PP74-F1-model_v2.pdb',\n",
       " 'Calb_AF-A0A1D8PS92-F1-model_v2.pdb',\n",
       " 'Calb_AF-Q5AJA5-F1-model_v2.pdb',\n",
       " 'Calb_AF-Q5ANN5-F1-model_v2.pdb',\n",
       " 'REF_Scer_AF-P15938-F1-model_v2.pdb',\n",
       " 'REF_Scer_AF-P20095-F1-model_v2.pdb',\n",
       " 'REF_Scer_AF-P24384-F1-model_v2.pdb',\n",
       " 'REF_Scer_AF-P53131-F1-model_v2.pdb',\n",
       " 'Spom_AF-O42643-F1-model_v2.pdb',\n",
       " 'Spom_AF-Q10752-F1-model_v2.pdb',\n",
       " 'Spom_AF-Q9P774-F1-model_v2.pdb',\n",
       " 'alloascoidea_hylecoeti__OG1004__0_887.pdb',\n",
       " 'alloascoidea_hylecoeti__OG1004__0_8918.pdb',\n",
       " 'alloascoidea_hylecoeti__OG1004__0_9006.pdb',\n",
       " 'alloascoidea_hylecoeti__OG1004__0_9446.pdb',\n",
       " 'ascoidea_rubescens__OG1004__6_4087.pdb',\n",
       " 'ascoidea_rubescens__OG1004__6_423.pdb',\n",
       " 'ascoidea_rubescens__OG1004__6_5493.pdb',\n",
       " 'candida_apicola__OG1004__13_946.pdb',\n",
       " 'candida_tropicalis__OG1004__30_158.pdb',\n",
       " 'candida_tropicalis__OG1004__30_3713.pdb',\n",
       " 'candida_tropicalis__OG1004__30_4380.pdb',\n",
       " 'cyberlindnera_jadinii__OG1004__35_1002.pdb',\n",
       " 'cyberlindnera_jadinii__OG1004__35_3955.pdb',\n",
       " 'cyberlindnera_jadinii__OG1004__35_5605.pdb',\n",
       " 'debaryomyces_hansenii__OG1004__36_1637.pdb',\n",
       " 'debaryomyces_hansenii__OG1004__36_1671.pdb',\n",
       " 'debaryomyces_hansenii__OG1004__36_206.pdb',\n",
       " 'debaryomyces_hansenii__OG1004__36_3659.pdb',\n",
       " 'eremothecium_gossypii__OG1004__40_2061.pdb',\n",
       " 'eremothecium_gossypii__OG1004__40_551.pdb',\n",
       " 'eremothecium_gossypii__OG1004__40_654.pdb',\n",
       " 'geotrichum_candidum__OG1004__43_1445.pdb',\n",
       " 'geotrichum_candidum__OG1004__43_4224.pdb',\n",
       " 'geotrichum_candidum__OG1004__43_4338.pdb',\n",
       " 'geotrichum_candidum__OG1004__43_6234.pdb',\n",
       " 'kazachstania_naganishii__OG1004__49_1489.pdb',\n",
       " 'kazachstania_naganishii__OG1004__49_2188.pdb',\n",
       " 'kluyveromyces_lactis__OG1004__50_1890.pdb',\n",
       " 'kluyveromyces_lactis__OG1004__50_4675.pdb',\n",
       " 'kluyveromyces_lactis__OG1004__50_560.pdb',\n",
       " 'kluyveromyces_lactis__OG1004__50_947.pdb',\n",
       " 'kluyveromyces_marxianus__OG1004__51_2472.pdb',\n",
       " 'kluyveromyces_marxianus__OG1004__51_2999.pdb',\n",
       " 'kluyveromyces_marxianus__OG1004__51_3204.pdb',\n",
       " 'kluyveromyces_marxianus__OG1004__51_4492.pdb',\n",
       " 'komagataella_pastoris__OG1004__52_305.pdb',\n",
       " 'komagataella_pastoris__OG1004__52_4447.pdb',\n",
       " 'komagataella_pastoris__OG1004__52_626.pdb',\n",
       " 'lachancea_thermotolerans__OG1004__64_1120.pdb',\n",
       " 'lachancea_thermotolerans__OG1004__64_1612.pdb',\n",
       " 'lachancea_thermotolerans__OG1004__64_415.pdb',\n",
       " 'lachancea_thermotolerans__OG1004__64_555.pdb',\n",
       " 'lipomyces_starkeyi__OG1004__66_2299.pdb',\n",
       " 'lipomyces_starkeyi__OG1004__66_2719.pdb',\n",
       " 'lipomyces_starkeyi__OG1004__66_5023.pdb',\n",
       " 'lipomyces_starkeyi__OG1004__66_5186.pdb',\n",
       " 'ogataea_parapolymorpha__OG1004__104_1852.pdb',\n",
       " 'ogataea_parapolymorpha__OG1004__104_2422.pdb',\n",
       " 'ogataea_parapolymorpha__OG1004__104_304.pdb',\n",
       " 'ogataea_parapolymorpha__OG1004__104_3759.pdb',\n",
       " 'pachysolen_tannophilus__OG1004__106_1032.pdb',\n",
       " 'pachysolen_tannophilus__OG1004__106_1046.pdb',\n",
       " 'pachysolen_tannophilus__OG1004__106_1205.pdb',\n",
       " 'pachysolen_tannophilus__OG1004__106_3036.pdb',\n",
       " 'tortispora_caseinolytica__OG1004__134_1167.pdb',\n",
       " 'tortispora_caseinolytica__OG1004__134_2049.pdb',\n",
       " 'tortispora_caseinolytica__OG1004__134_2172.pdb',\n",
       " 'tortispora_caseinolytica__OG1004__134_3721.pdb',\n",
       " 'torulaspora_delbrueckii__OG1004__135_168.pdb',\n",
       " 'torulaspora_delbrueckii__OG1004__135_2867.pdb',\n",
       " 'torulaspora_delbrueckii__OG1004__135_323.pdb',\n",
       " 'torulaspora_delbrueckii__OG1004__135_3739.pdb',\n",
       " 'vanderwaltozyma_polyspora__OG1004__136_1186.pdb',\n",
       " 'vanderwaltozyma_polyspora__OG1004__136_2813.pdb',\n",
       " 'vanderwaltozyma_polyspora__OG1004__136_5461.pdb',\n",
       " 'vanderwaltozyma_polyspora__OG1004__136_5617.pdb',\n",
       " 'wickerhamomyces_anomalus__OG1004__139_3846.pdb',\n",
       " 'wickerhamomyces_anomalus__OG1004__139_4353.pdb',\n",
       " 'wickerhamomyces_anomalus__OG1004__139_4666.pdb',\n",
       " 'yHMPu5000034604_sporopachydermia_lactativora_160519__OG1004__170_2393.pdb',\n",
       " 'yHMPu5000034604_sporopachydermia_lactativora_160519__OG1004__170_2638.pdb',\n",
       " 'yHMPu5000034604_sporopachydermia_lactativora_160519__OG1004__170_4824.pdb',\n",
       " 'yHMPu5000034604_sporopachydermia_lactativora_160519__OG1004__170_5251.pdb',\n",
       " 'yHMPu5000034957_hanseniaspora_osmophila_160519__OG1004__247_1238.pdb',\n",
       " 'yHMPu5000034957_hanseniaspora_osmophila_160519__OG1004__247_2465.pdb',\n",
       " 'yHMPu5000034957_hanseniaspora_osmophila_160519__OG1004__247_3014.pdb',\n",
       " 'zygosaccharomyces_rouxii__OG1004__342_1656.pdb',\n",
       " 'zygosaccharomyces_rouxii__OG1004__342_2886.pdb',\n",
       " 'zygosaccharomyces_rouxii__OG1004__342_3828.pdb',\n",
       " 'zygosaccharomyces_rouxii__OG1004__342_790.pdb']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_len_filt['OG1004_C1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'count')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkQAAAGwCAYAAABIC3rIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAreklEQVR4nO3deXRUZb6v8W+ZkBAgKZmSUE2AoJHBIDI1BOmGI/NM2wdQMA0HRGnGMDTDcUJPmwhHwKNoBFoBBxrXbcGhxUDohhwGGYykZUYZJDYJwTapAMYEkn3/cLHvLYIQQ1I78D6ftWotau+3qn4bhzxr166Ky7IsSwAAAAa7zekBAAAAnEYQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4gU4PcLMoKSnR6dOnFRoaKpfL5fQ4AACgDCzL0rlz5+TxeHTbbT99HoggKqPTp08rKirK6TEAAEA5ZGZmqmHDhj+5nyAqo9DQUEk//oWGhYU5PA0AACiL/Px8RUVF2T/HfwpBVEaX3yYLCwsjiAAAuMlc73IXLqoGAADGI4gAAIDxCCIAAGA8gggAABiPIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGC8QKcHQNXTZM7H111z8vn+fpgEAAD/4AwRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4jgbRpUuX9MQTTyg6OlohISFq2rSpnn32WZWUlNhrLMvSvHnz5PF4FBISom7duunAgQM+z1NYWKjJkyerXr16qlmzpgYNGqRvvvnGZ01ubq7i4+PldrvldrsVHx+vvLw8fxwmAACo4hwNovnz5+u1117TkiVLdOjQIS1YsED//d//rZdfftles2DBAi1atEhLlizRnj17FBkZqZ49e+rcuXP2moSEBK1bt05r1qzRtm3bdP78eQ0YMEDFxcX2mhEjRigjI0MpKSlKSUlRRkaG4uPj/Xq8AACganJZlmU59eIDBgxQRESEXn/9dXvbb3/7W9WoUUNvvfWWLMuSx+NRQkKCZs+eLenHs0ERERGaP3++HnvsMXm9XtWvX19vvfWWhg8fLkk6ffq0oqKitH79evXu3VuHDh1Sy5YttXPnTnXs2FGStHPnTsXFxenw4cNq1qzZdWfNz8+X2+2W1+tVWFhYJfxtVB1N5nx83TUnn+/vh0kAALgxZf357egZoi5duuhvf/ubjh49Kkn6xz/+oW3btqlfv36SpBMnTig7O1u9evWyHxMcHKyuXbtqx44dkqT09HRdvHjRZ43H41FsbKy95tNPP5Xb7bZjSJI6deokt9ttr7lSYWGh8vPzfW4AAODWFOjki8+ePVter1fNmzdXQECAiouL9dxzz+mhhx6SJGVnZ0uSIiIifB4XERGhr7/+2l4TFBSk2rVrl1pz+fHZ2dkKDw8v9frh4eH2mislJSXpmWeeubEDBAAANwVHzxC9++67evvtt7V69Wp9/vnnWrVqlV544QWtWrXKZ53L5fK5b1lWqW1XunLN1dZf63nmzp0rr9dr3zIzM8t6WAAA4Cbj6BmiP/zhD5ozZ44efPBBSVKrVq309ddfKykpSaNGjVJkZKSkH8/wNGjQwH5cTk6OfdYoMjJSRUVFys3N9TlLlJOTo86dO9trzpw5U+r1z549W+rs02XBwcEKDg6umAMFAABVmqNniL7//nvddpvvCAEBAfbH7qOjoxUZGanU1FR7f1FRkdLS0uzYadeunapVq+azJisrS/v377fXxMXFyev1avfu3faaXbt2yev12msAAIC5HD1DNHDgQD333HNq1KiR7r77bu3du1eLFi3SmDFjJP34NldCQoISExMVExOjmJgYJSYmqkaNGhoxYoQkye12a+zYsZoxY4bq1q2rOnXqaObMmWrVqpV69OghSWrRooX69OmjcePGaenSpZKkRx99VAMGDCjTJ8wAAMCtzdEgevnll/Xkk09qwoQJysnJkcfj0WOPPaannnrKXjNr1iwVFBRowoQJys3NVceOHbVx40aFhobaaxYvXqzAwEANGzZMBQUF6t69u1auXKmAgAB7zTvvvKMpU6bYn0YbNGiQlixZ4r+DBQAAVZaj30N0M+F7iHzxPUQAgJvBTfE9RAAAAFUBQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwnuNB9M9//lMPP/yw6tatqxo1aujee+9Venq6vd+yLM2bN08ej0chISHq1q2bDhw44PMchYWFmjx5surVq6eaNWtq0KBB+uabb3zW5ObmKj4+Xm63W263W/Hx8crLy/PHIQIAgCrO0SDKzc3Vfffdp2rVqumTTz7RwYMHtXDhQt1+++32mgULFmjRokVasmSJ9uzZo8jISPXs2VPnzp2z1yQkJGjdunVas2aNtm3bpvPnz2vAgAEqLi6214wYMUIZGRlKSUlRSkqKMjIyFB8f78/DBQAAVZTLsizLqRefM2eOtm/frq1bt151v2VZ8ng8SkhI0OzZsyX9eDYoIiJC8+fP12OPPSav16v69evrrbfe0vDhwyVJp0+fVlRUlNavX6/evXvr0KFDatmypXbu3KmOHTtKknbu3Km4uDgdPnxYzZo1K/XahYWFKiwstO/n5+crKipKXq9XYWFhFf1XUaU0mfPxddecfL6/HyYBAODG5Ofny+12X/fnt6NniD788EO1b99eQ4cOVXh4uNq0aaPly5fb+0+cOKHs7Gz16tXL3hYcHKyuXbtqx44dkqT09HRdvHjRZ43H41FsbKy95tNPP5Xb7bZjSJI6deokt9ttr7lSUlKS/faa2+1WVFRUhR47AACoOhwNouPHjys5OVkxMTHasGGDxo8frylTpujNN9+UJGVnZ0uSIiIifB4XERFh78vOzlZQUJBq1659zTXh4eGlXj88PNxec6W5c+fK6/Xat8zMzBs7WAAAUGUFOvniJSUlat++vRITEyVJbdq00YEDB5ScnKzf/e539jqXy+XzOMuySm270pVrrrb+Ws8THBys4ODgMh8LAAC4eTl6hqhBgwZq2bKlz7YWLVro1KlTkqTIyEhJKnUWJycnxz5rFBkZqaKiIuXm5l5zzZkzZ0q9/tmzZ0udfQIAAOZxNIjuu+8+HTlyxGfb0aNH1bhxY0lSdHS0IiMjlZqaau8vKipSWlqaOnfuLElq166dqlWr5rMmKytL+/fvt9fExcXJ6/Vq9+7d9ppdu3bJ6/XaawAAgLkcfcts2rRp6ty5sxITEzVs2DDt3r1by5Yt07JlyyT9+DZXQkKCEhMTFRMTo5iYGCUmJqpGjRoaMWKEJMntdmvs2LGaMWOG6tatqzp16mjmzJlq1aqVevToIenHs059+vTRuHHjtHTpUknSo48+qgEDBlz1E2YAAMAsjgZRhw4dtG7dOs2dO1fPPvusoqOj9eKLL2rkyJH2mlmzZqmgoEATJkxQbm6uOnbsqI0bNyo0NNRes3jxYgUGBmrYsGEqKChQ9+7dtXLlSgUEBNhr3nnnHU2ZMsX+NNqgQYO0ZMkS/x0sAACoshz9HqKbSVm/x+BWwPcQAQBuFTfF9xABAABUBQQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADBeuYLo/vvvV15eXqnt+fn5uv/++290JgAAAL8qVxBt2bJFRUVFpbb/8MMP2rp16w0PBQAA4E+BP2fxF198Yf/54MGDys7Otu8XFxcrJSVFv/jFLypuOgAAAD/4WUF07733yuVyyeVyXfWtsZCQEL388ssVNhwAAIA//KwgOnHihCzLUtOmTbV7927Vr1/f3hcUFKTw8HAFBARU+JAAAACV6WcFUePGjSVJJSUllTIMAACAE35WEP3/jh49qi1btignJ6dUID311FM3PBgAAIC/lCuIli9frt///veqV6+eIiMj5XK57H0ul4sgAgAAN5VyBdEf//hHPffcc5o9e3ZFzwMAAOB35foeotzcXA0dOrSiZwEAAHBEuYJo6NCh2rhxY0XPAgAA4IhyvWV255136sknn9TOnTvVqlUrVatWzWf/lClTKmQ4AAAAf3BZlmX93AdFR0f/9BO6XDp+/PgNDVUV5efny+12y+v1KiwszOlxKlWTOR9fd83J5/v7YRIAAG5MWX9+l+sM0YkTJ8o9GAAAQFVTrmuIAAAAbiXlOkM0ZsyYa+5/4403yjUMAACAE8oVRLm5uT73L168qP379ysvL++qv/QVAACgKitXEK1bt67UtpKSEk2YMEFNmza94aEAAAD8qcKuIbrttts0bdo0LV68uKKeEgAAwC8q9KLqY8eO6dKlSxX5lAAAAJWuXG+ZTZ8+3ee+ZVnKysrSxx9/rFGjRlXIYAAAAP5SriDau3evz/3bbrtN9evX18KFC6/7CTQAAICqplxBtHnz5oqeAwAAwDHlCqLLzp49qyNHjsjlcumuu+5S/fr1K2ouAAAAvynXRdUXLlzQmDFj1KBBA/3617/Wr371K3k8Ho0dO1bff/99Rc8IAABQqcoVRNOnT1daWpo++ugj5eXlKS8vTx988IHS0tI0Y8aMip4RAACgUpXrLbP33ntPf/nLX9StWzd7W79+/RQSEqJhw4YpOTm5ouYDAACodOU6Q/T9998rIiKi1Pbw8HDeMgMAADedcgVRXFycnn76af3www/2toKCAj3zzDOKi4ursOEAAAD8oVxvmb344ovq27evGjZsqNatW8vlcikjI0PBwcHauHFjRc8IAABQqcoVRK1atdKXX36pt99+W4cPH5ZlWXrwwQc1cuRIhYSEVPSMAAAAlapcQZSUlKSIiAiNGzfOZ/sbb7yhs2fPavbs2RUyHAAAgD+U6xqipUuXqnnz5qW233333XrttddueCgAAAB/KlcQZWdnq0GDBqW2169fX1lZWTc8FAAAgD+VK4iioqK0ffv2Utu3b98uj8dzw0MBAAD4U7muIXrkkUeUkJCgixcv6v7775ck/e1vf9OsWbP4pmoAAHDTKVcQzZo1S999950mTJigoqIiSVL16tU1e/ZszZ07t0IHBAAAqGzlCiKXy6X58+frySef1KFDhxQSEqKYmBgFBwdX9HwAAACVrlxBdFmtWrXUoUOHipoFAADAEeW6qLoyJCUlyeVyKSEhwd5mWZbmzZsnj8ejkJAQdevWTQcOHPB5XGFhoSZPnqx69eqpZs2aGjRokL755hufNbm5uYqPj5fb7Zbb7VZ8fLzy8vL8cFQAAOBmUCWCaM+ePVq2bJnuuecen+0LFizQokWLtGTJEu3Zs0eRkZHq2bOnzp07Z69JSEjQunXrtGbNGm3btk3nz5/XgAEDVFxcbK8ZMWKEMjIylJKSopSUFGVkZCg+Pt5vxwcAAKo2x4Po/PnzGjlypJYvX67atWvb2y3L0osvvqjHH39cDzzwgGJjY7Vq1Sp9//33Wr16tSTJ6/Xq9ddf18KFC9WjRw+1adNGb7/9tvbt26dNmzZJkg4dOqSUlBT96U9/UlxcnOLi4rR8+XL99a9/1ZEjR35yrsLCQuXn5/vcAADArcnxIJo4caL69++vHj16+Gw/ceKEsrOz1atXL3tbcHCwunbtqh07dkiS0tPTdfHiRZ81Ho9HsbGx9ppPP/1UbrdbHTt2tNd06tRJbrfbXnM1SUlJ9ltsbrdbUVFRFXK8AACg6nE0iNasWaPPP/9cSUlJpfZlZ2dLkiIiIny2R0RE2Puys7MVFBTkc2bpamvCw8NLPX94eLi95mrmzp0rr9dr3zIzM3/ewQEAgJvGDX3K7EZkZmZq6tSp2rhxo6pXr/6T61wul899y7JKbbvSlWuutv56zxMcHMzXCAAAYAjHzhClp6crJydH7dq1U2BgoAIDA5WWlqaXXnpJgYGB9pmhK8/i5OTk2PsiIyNVVFSk3Nzca645c+ZMqdc/e/ZsqbNPAADATI4FUffu3bVv3z5lZGTYt/bt22vkyJHKyMhQ06ZNFRkZqdTUVPsxRUVFSktLU+fOnSVJ7dq1U7Vq1XzWZGVlaf/+/faauLg4eb1e7d69216za9cueb1eew0AADCbY2+ZhYaGKjY21mdbzZo1VbduXXt7QkKCEhMTFRMTo5iYGCUmJqpGjRoaMWKEJMntdmvs2LGaMWOG6tatqzp16mjmzJlq1aqVfZF2ixYt1KdPH40bN05Lly6VJD366KMaMGCAmjVr5scjBgAAVZVjQVQWs2bNUkFBgSZMmKDc3Fx17NhRGzduVGhoqL1m8eLFCgwM1LBhw1RQUKDu3btr5cqVCggIsNe88847mjJliv1ptEGDBmnJkiV+Px4AAFA1uSzLspwe4maQn58vt9str9ersLAwp8epVE3mfHzdNSef7++HSQAAuDFl/fnt+PcQAQAAOI0gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGA8gggAABiPIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGA8gggAABiPIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGA8gggAABiPIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGA8gggAABiPIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGA8gggAABiPIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGA8R4MoKSlJHTp0UGhoqMLDwzVkyBAdOXLEZ41lWZo3b548Ho9CQkLUrVs3HThwwGdNYWGhJk+erHr16qlmzZoaNGiQvvnmG581ubm5io+Pl9vtltvtVnx8vPLy8ir7EAEAwE3A0SBKS0vTxIkTtXPnTqWmpurSpUvq1auXLly4YK9ZsGCBFi1apCVLlmjPnj2KjIxUz549de7cOXtNQkKC1q1bpzVr1mjbtm06f/68BgwYoOLiYnvNiBEjlJGRoZSUFKWkpCgjI0Px8fF+PV4AAFA1uSzLspwe4rKzZ88qPDxcaWlp+vWvfy3LsuTxeJSQkKDZs2dL+vFsUEREhObPn6/HHntMXq9X9evX11tvvaXhw4dLkk6fPq2oqCitX79evXv31qFDh9SyZUvt3LlTHTt2lCTt3LlTcXFxOnz4sJo1a3bd2fLz8+V2u+X1ehUWFlZ5fwlVQJM5H193zcnn+/thEgAAbkxZf35XqWuIvF6vJKlOnTqSpBMnTig7O1u9evWy1wQHB6tr167asWOHJCk9PV0XL170WePxeBQbG2uv+fTTT+V2u+0YkqROnTrJ7Xbba65UWFio/Px8nxsAALg1VZkgsixL06dPV5cuXRQbGytJys7OliRFRET4rI2IiLD3ZWdnKygoSLVr177mmvDw8FKvGR4ebq+5UlJSkn29kdvtVlRU1I0dIAAAqLKqTBBNmjRJX3zxhf785z+X2udyuXzuW5ZVatuVrlxztfXXep65c+fK6/Xat8zMzLIcBgAAuAlViSCaPHmyPvzwQ23evFkNGza0t0dGRkpSqbM4OTk59lmjyMhIFRUVKTc395przpw5U+p1z549W+rs02XBwcEKCwvzuQEAgFuTo0FkWZYmTZqktWvX6u9//7uio6N99kdHRysyMlKpqan2tqKiIqWlpalz586SpHbt2qlatWo+a7KysrR//357TVxcnLxer3bv3m2v2bVrl7xer70GAACYK9DJF584caJWr16tDz74QKGhofaZILfbrZCQELlcLiUkJCgxMVExMTGKiYlRYmKiatSooREjRthrx44dqxkzZqhu3bqqU6eOZs6cqVatWqlHjx6SpBYtWqhPnz4aN26cli5dKkl69NFHNWDAgDJ9wgwAANzaHA2i5ORkSVK3bt18tq9YsUKjR4+WJM2aNUsFBQWaMGGCcnNz1bFjR23cuFGhoaH2+sWLFyswMFDDhg1TQUGBunfvrpUrVyogIMBe884772jKlCn2p9EGDRqkJUuWVO4BAgCAm0KV+h6iqozvIfLF9xABAG4GN+X3EAEAADiBIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxiOIAACA8QgiAABgPIIIAAAYjyACAADGI4gAAIDxCCIAAGA8gggAABiPIAIAAMYjiAAAgPEIIgAAYDyCCAAAGI8gAgAAxgt0egDcnJrM+fi6a04+398PkwAAcOM4QwQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4xFEAADAeAQRAAAwHkEEAACMRxABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAQAA4wU6PQD8q8mcj50eAQCAKoczRAAAwHgEEQAAMB5vmaHSlOXtuZPP9/fDJAAAXBtniAAAgPEIIgAAYDyCCAAAGI8gAgAAxjMqiF599VVFR0erevXqateunbZu3er0SAAAoAow5lNm7777rhISEvTqq6/qvvvu09KlS9W3b18dPHhQjRo1cno8Y/n7iyL5VBsA4GpclmVZTg/hDx07dlTbtm2VnJxsb2vRooWGDBmipKSk6z4+Pz9fbrdbXq9XYWFhlTnqVfEN01ULYQUAN4ey/vw24gxRUVGR0tPTNWfOHJ/tvXr10o4dO676mMLCQhUWFtr3vV6vpB//Yita7NMbKvw5UbkaTfs/To9ghP3P9L7umrL891OW5wFwa7r8c/t653+MCKJvv/1WxcXFioiI8NkeERGh7Ozsqz4mKSlJzzzzTKntUVFRlTIjgNLcL1at5wFw8zp37pzcbvdP7jciiC5zuVw+9y3LKrXtsrlz52r69On2/ZKSEn333XeqVq2aGjVqpMzMTEfeOnNSfn6+oqKijDx2yezj59jNPHbJ7OM3+dilW+f4LcvSuXPn5PF4rrnOiCCqV6+eAgICSp0NysnJKXXW6LLg4GAFBwf7bLv99tvtU29hYWE39b8gN8LkY5fMPn6O3cxjl8w+fpOPXbo1jv9aZ4YuM+Jj90FBQWrXrp1SU1N9tqempqpz584OTQUAAKoKI84QSdL06dMVHx+v9u3bKy4uTsuWLdOpU6c0fvx4p0cDAAAOMyaIhg8frn/961969tlnlZWVpdjYWK1fv16NGzf+Wc8THBysp59+utTbaSYw+dgls4+fYzfz2CWzj9/kY5fMO35jvocIAADgpxhxDREAAMC1EEQAAMB4BBEAADAeQQQAAIxHEP0Mr776qqKjo1W9enW1a9dOW7dudXokv/jf//1fDRw4UB6PRy6XS++//77TI/lNUlKSOnTooNDQUIWHh2vIkCE6cuSI02P5TXJysu655x77i9ni4uL0ySefOD2WI5KSkuRyuZSQkOD0KJVu3rx5crlcPrfIyEinx/Krf/7zn3r44YdVt25d1ahRQ/fee6/S09OdHqvSNWnSpNQ/e5fLpYkTJzo9WqUjiMro3XffVUJCgh5//HHt3btXv/rVr9S3b1+dOnXK6dEq3YULF9S6dWstWbLE6VH8Li0tTRMnTtTOnTuVmpqqS5cuqVevXrpw4YLTo/lFw4YN9fzzz+uzzz7TZ599pvvvv1+DBw/WgQMHnB7Nr/bs2aNly5bpnnvucXoUv7n77ruVlZVl3/bt2+f0SH6Tm5ur++67T9WqVdMnn3yigwcPauHChbr99tudHq3S7dmzx+ef++UvNB46dKjDk/mBhTL55S9/aY0fP95nW/Pmza05c+Y4NJEzJFnr1q1zegzH5OTkWJKstLQ0p0dxTO3ata0//elPTo/hN+fOnbNiYmKs1NRUq2vXrtbUqVOdHqnSPf3001br1q2dHsMxs2fPtrp06eL0GFXC1KlTrTvuuMMqKSlxepRKxxmiMigqKlJ6erp69erls71Xr17asWOHQ1PBCV6vV5JUp04dhyfxv+LiYq1Zs0YXLlxQXFyc0+P4zcSJE9W/f3/16NHD6VH86ssvv5TH41F0dLQefPBBHT9+3OmR/ObDDz9U+/btNXToUIWHh6tNmzZavny502P5XVFRkd5++22NGTPmJ38R+q2EICqDb7/9VsXFxaV+EWxERESpXxiLW5dlWZo+fbq6dOmi2NhYp8fxm3379qlWrVoKDg7W+PHjtW7dOrVs2dLpsfxizZo1+vzzz5WUlOT0KH7VsWNHvfnmm9qwYYOWL1+u7Oxsde7cWf/617+cHs0vjh8/ruTkZMXExGjDhg0aP368pkyZojfffNPp0fzq/fffV15enkaPHu30KH5hzK/uqAhXFrJlWUZUM340adIkffHFF9q2bZvTo/hVs2bNlJGRoby8PL333nsaNWqU0tLSbvkoyszM1NSpU7Vx40ZVr17d6XH8qm/fvvafW7Vqpbi4ON1xxx1atWqVpk+f7uBk/lFSUqL27dsrMTFRktSmTRsdOHBAycnJ+t3vfufwdP7z+uuvq2/fvvJ4PE6P4hecISqDevXqKSAgoNTZoJycnFJnjXBrmjx5sj788ENt3rxZDRs2dHocvwoKCtKdd96p9u3bKykpSa1bt9b//M//OD1WpUtPT1dOTo7atWunwMBABQYGKi0tTS+99JICAwNVXFzs9Ih+U7NmTbVq1Upffvml06P4RYMGDUoFf4sWLYz4EM1lX3/9tTZt2qRHHnnE6VH8hiAqg6CgILVr186+2v6y1NRUde7c2aGp4A+WZWnSpElau3at/v73vys6OtrpkRxnWZYKCwudHqPSde/eXfv27VNGRoZ9a9++vUaOHKmMjAwFBAQ4PaLfFBYW6tChQ2rQoIHTo/jFfffdV+rrNY4ePfqzfxn4zWzFihUKDw9X//79nR7Fb3jLrIymT5+u+Ph4tW/fXnFxcVq2bJlOnTql8ePHOz1apTt//ry++uor+/6JEyeUkZGhOnXqqFGjRg5OVvkmTpyo1atX64MPPlBoaKh9ltDtdiskJMTh6Srff/7nf6pv376KiorSuXPntGbNGm3ZskUpKSlOj1bpQkNDS10rVrNmTdWtW/eWv4Zs5syZGjhwoBo1aqScnBz98Y9/VH5+vkaNGuX0aH4xbdo0de7cWYmJiRo2bJh2796tZcuWadmyZU6P5hclJSVasWKFRo0apcBAgzLB2Q+53VxeeeUVq3HjxlZQUJDVtm1bYz56vXnzZktSqduoUaOcHq3SXe24JVkrVqxwejS/GDNmjP3vfP369a3u3btbGzdudHosx5jysfvhw4dbDRo0sKpVq2Z5PB7rgQcesA4cOOD0WH710UcfWbGxsVZwcLDVvHlza9myZU6P5DcbNmywJFlHjhxxehS/clmWZTmTYgAAAFUD1xABAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjEcQAajStmzZIpfLpby8PEnSypUrdfvttzs6k9NGjx6tIUOGOD0GcEshiACU2+jRo+Vyua76O/0mTJggl8ul0aNHV+hrDh8+XEePHq3Q57yaqhAdJ0+elMvlUkZGhqNzACYgiADckKioKK1Zs0YFBQX2th9++EF//vOfK+WX/4aEhCg8PLzCnxeA2QgiADekbdu2atSokdauXWtvW7t2raKiotSmTRuftZZlacGCBWratKlCQkLUunVr/eUvf/FZs379et11110KCQnRv/3bv+nkyZM++698y+zYsWMaPHiwIiIiVKtWLXXo0EGbNm3yeUyTJk2UmJioMWPGKDQ0VI0aNbrh31x+8OBB9evXT7Vq1VJERITi4+P17bff2vu7deumKVOmaNasWapTp44iIyM1b948n+c4fPiwunTpourVq6tly5batGmTXC6X3n//fUlSdHS0JKlNmzZyuVzq1q2bz+NfeOEFNWjQQHXr1tXEiRN18eLFGzomwGQEEYAb9h//8R9asWKFff+NN97QmDFjSq174okntGLFCiUnJ+vAgQOaNm2aHn74YaWlpUmSMjMz9cADD6hfv37KyMjQI488ojlz5lzztc+fP69+/fpp06ZN2rt3r3r37q2BAwfq1KlTPusWLlyo9u3ba+/evZowYYJ+//vf6/Dhw+U63qysLHXt2lX33nuvPvvsM6WkpOjMmTMaNmyYz7pVq1apZs2a2rVrlxYsWKBnn31WqampkqSSkhINGTJENWrU0K5du7Rs2TI9/vjjPo/fvXu3JGnTpk3Kysryic7Nmzfr2LFj2rx5s1atWqWVK1dq5cqV5ToeAJIsACinUaNGWYMHD7bOnj1rBQcHWydOnLBOnjxpVa9e3Tp79qw1ePBga9SoUZZlWdb58+et6tWrWzt27PB5jrFjx1oPPfSQZVmWNXfuXKtFixZWSUmJvX/27NmWJCs3N9eyLMtasWKF5Xa7rzlXy5YtrZdfftm+37hxY+vhhx+275eUlFjh4eFWcnLydY/tap588kmrV69ePtsyMzMtSdaRI0csy7Ksrl27Wl26dPFZ06FDB2v27NmWZVnWJ598YgUGBlpZWVn2/tTUVEuStW7dOsuyLOvEiROWJGvv3r2lZmvcuLF16dIle9vQoUOt4cOH/+TxALi2QGdzDMCtoF69eurfv79WrVoly7LUv39/1atXz2fNwYMH9cMPP6hnz54+24uKiuy31g4dOqROnTrJ5XLZ++Pi4q752hcuXNAzzzyjv/71rzp9+rQuXbqkgoKCUmeI7rnnHvvPLpdLkZGRysnJKdfxpqena/PmzapVq1apfceOHdNdd91V6jUlqUGDBvZrHjlyRFFRUYqMjLT3//KXvyzzDHfffbcCAgJ8nnvfvn0/6zgA/D8EEYAKMWbMGE2aNEmS9Morr5TaX1JSIkn6+OOP9Ytf/MJnX3BwsKQfrzH6uf7whz9ow4YNeuGFF3TnnXcqJCRE//7v/66ioiKfddWqVfO573K57Jl+rpKSEg0cOFDz588vta9BgwZlek3LsnzC7+eqyOMBQBABqCB9+vSxI6R3796l9rds2VLBwcE6deqUunbtetXnaNmypX1B8WU7d+685utu3bpVo0eP1m9+8xtJP15TdOWF2BWtbdu2eu+999SkSRMFBpbvf6PNmzfXqVOndObMGUVEREiS9uzZ47MmKChIklRcXHxjAwO4Li6qBlAhAgICdOjQIR06dMjnrZzLQkNDNXPmTE2bNk2rVq3SsWPHtHfvXr3yyitatWqVJGn8+PE6duyYpk+friNHjmj16tXXvVD4zjvv1Nq1a5WRkaF//OMfGjFiRIWdKfF6vcrIyPC5nTp1ShMnTtR3332nhx56SLt379bx48e1ceNGjRkzpszx0rNnT91xxx0aNWqUvvjiC23fvt2+qPrymaPw8HCFhITYF217vd4KOS4ApRFEACpMWFiYwsLCfnL/f/3Xf+mpp55SUlKSWrRood69e+ujjz6yP17eqFEjvffee/roo4/UunVrvfbaa0pMTLzmay5evFi1a9dW586dNXDgQPXu3Vtt27atkOPZsmWL2rRp43N76qmn5PF4tH37dhUXF6t3796KjY3V1KlT5Xa7ddttZfvfakBAgN5//32dP39eHTp00COPPKInnnhCklS9enVJUmBgoF566SUtXbpUHo9HgwcPrpDjAlCayyrPm/YAgAq3fft2denSRV999ZXuuOMOp8cBjEIQAYBD1q1bp1q1aikmJkZfffWVpk6dqtq1a2vbtm1OjwYYh4uqAcAh586d06xZs5SZmal69eqpR48eWrhwodNjAUbiDBEAADAeF1UDAADjEUQAAMB4BBEAADAeQQQAAIxHEAEAAOMRRAAAwHgEEQAAMB5BBAAAjPd/ARnLDUKZCKtuAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "norm_lengths = []\n",
    "for alignment, (med,norm_lengths_og) in clusters_len_norm.items():\n",
    "    for norm_length in norm_lengths_og:\n",
    "        norm_lengths.append(norm_length)\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.hist(norm_lengths, bins=50)\n",
    "ax.set_xlabel('Median Length')\n",
    "ax.set_ylabel('count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 479 nonzero clusters\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'N clusters')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAGwCAYAAABcnuQpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAA9hAAAPYQGoP6dpAAArsElEQVR4nO3df3RU5Z3H8c9AYEggifycIRohqamAgFKiaBQSVhOlHKQGf1R+qtXFgkqkLj+ku46sJJCtNFUUhboxHJfSY0VrxUKiSJCmlgiN0uAKaICoyWZVTMIPJ0Ke/cPDXcdAzECSmSe+X+fcc7jPfebOd/IA+ZznPneuyxhjBAAAYKlOoS4AAADgbBBmAACA1QgzAADAaoQZAABgNcIMAACwGmEGAABYjTADAACsFhHqAtpaY2OjPvnkE0VHR8vlcoW6HAAA0ALGGNXX1ysuLk6dOjU/99Lhw8wnn3yi+Pj4UJcBAADOQGVlpc4777xm+3T4MBMdHS3p6x9GTExMiKsBAAAtUVdXp/j4eOf3eHM6fJg5eWkpJiaGMAMAgGVaskSEBcAAAMBqIQ0zAwcOlMvlarLNnj1b0teLf3w+n+Li4hQZGam0tDSVl5eHsmQAABBmQhpmSktLVVVV5WxFRUWSpJtuukmSlJubq+XLl2vFihUqLS2V1+tVenq66uvrQ1k2AAAIIy5jjAl1ESdlZWXplVde0d69eyVJcXFxysrK0vz58yVJfr9fHo9Hy5Yt08yZM095Dr/fL7/f7+yfXEBUW1vLmhkAACxRV1en2NjYFv3+Dps1Mw0NDXruued0xx13yOVyqaKiQtXV1crIyHD6uN1upaamqqSk5LTnycnJUWxsrLNxWzYAAB1b2ISZl156SV988YVuu+02SVJ1dbUkyePxBPTzeDzOsVNZuHChamtrna2ysrLNagYAAKEXNrdmP/PMMxo3bpzi4uIC2r99S5YxptnbtNxut9xud5vUCAAAwk9YzMwcOHBAr732mu68806nzev1SlKTWZiampomszUAAOD7KyzCTH5+vvr166fx48c7bQkJCfJ6vc4dTtLX62qKi4uVkpISijIBAEAYCvllpsbGRuXn52vGjBmKiPj/clwul7KyspSdna2kpCQlJSUpOztbUVFRmjx5cggrBgAA4STkYea1117TwYMHdccddzQ5Nm/ePB07dkyzZs3SoUOHNGrUKBUWFrboOQ0AAOD7Iay+Z6YtBHOfOgAACA9Wfs8MAADAmSDMAAAAqxFmAACA1UK+ABiwycAFGwL29y8df5qeAID2wswMAACwGmEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFYjzAAAAKsRZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBphBgAAWI0wAwAArEaYAQAAViPMAAAAqxFmAACA1QgzAADAaoQZAABgNcIMAACwGmEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFYjzAAAAKuFPMx8/PHHmjp1qnr37q2oqChdcskl2rFjh3PcGCOfz6e4uDhFRkYqLS1N5eXlIawYAACEk5CGmUOHDunKK69Uly5d9Oc//1m7d+/Wo48+qnPOOcfpk5ubq+XLl2vFihUqLS2V1+tVenq66uvrQ1c4AAAIGxGhfPNly5YpPj5e+fn5TtvAgQOdPxtjlJeXp0WLFikzM1OSVFBQII/Ho7Vr12rmzJlNzun3++X3+539urq6tvsAAAAg5EI6M/Pyyy8rOTlZN910k/r166cRI0Zo9erVzvGKigpVV1crIyPDaXO73UpNTVVJSckpz5mTk6PY2Fhni4+Pb/PPAQAAQiekYebDDz/UypUrlZSUpE2bNunuu+/WfffdpzVr1kiSqqurJUkejyfgdR6Pxzn2bQsXLlRtba2zVVZWtu2HAAAAIRXSy0yNjY1KTk5Wdna2JGnEiBEqLy/XypUrNX36dKefy+UKeJ0xpknbSW63W263u+2KBgAAYSWkMzP9+/fXkCFDAtoGDx6sgwcPSpK8Xq8kNZmFqampaTJbAwAAvp9CGmauvPJKvf/++wFte/bs0YABAyRJCQkJ8nq9Kioqco43NDSouLhYKSkp7VorAAAITyG9zHT//fcrJSVF2dnZuvnmm7V9+3atWrVKq1atkvT15aWsrCxlZ2crKSlJSUlJys7OVlRUlCZPnhzK0gEAQJgIaZi59NJL9eKLL2rhwoVavHixEhISlJeXpylTpjh95s2bp2PHjmnWrFk6dOiQRo0apcLCQkVHR4ewcgAAEC5cxhgT6iLaUl1dnWJjY1VbW6uYmJhQlwPLDVywIWB//9LxIaoEADq2YH5/h/xxBgAAAGeDMAMAAKwW0jUzQEfApScACC1mZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBphBgAAWI0wAwAArEaYAQAAViPMAAAAqxFmAACA1QgzAADAaoQZAABgNcIMAACwGmEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFYjzAAAAKsRZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBphBgAAWI0wAwAArEaYAQAAViPMAAAAqxFmAACA1QgzAADAaoQZAABgNcIMAACwGmEGAABYLaRhxufzyeVyBWxer9c5boyRz+dTXFycIiMjlZaWpvLy8hBWDAAAwk3IZ2YuuugiVVVVOduuXbucY7m5uVq+fLlWrFih0tJSeb1epaenq76+PoQVAwCAcBLyMBMRESGv1+tsffv2lfT1rExeXp4WLVqkzMxMDR06VAUFBTp69KjWrl0b4qoBAEC4CHmY2bt3r+Li4pSQkKCf/vSn+vDDDyVJFRUVqq6uVkZGhtPX7XYrNTVVJSUlpz2f3+9XXV1dwAYAADqukIaZUaNGac2aNdq0aZNWr16t6upqpaSk6LPPPlN1dbUkyePxBLzG4/E4x04lJydHsbGxzhYfH9+mnwEAAIRWSMPMuHHjNGnSJA0bNkzXXHONNmzYIEkqKChw+rhcroDXGGOatH3TwoULVVtb62yVlZVtUzwAAAgLIb/M9E3du3fXsGHDtHfvXueupm/PwtTU1DSZrfkmt9utmJiYgA0AAHRcYRVm/H6/3nvvPfXv318JCQnyer0qKipyjjc0NKi4uFgpKSkhrBIAAISTiFC++QMPPKAJEybo/PPPV01NjR555BHV1dVpxowZcrlcysrKUnZ2tpKSkpSUlKTs7GxFRUVp8uTJoSwbAACEkZCGmY8++ki33nqrPv30U/Xt21eXX3653nrrLQ0YMECSNG/ePB07dkyzZs3SoUOHNGrUKBUWFio6OjqUZQMAgDAS0jCzbt26Zo+7XC75fD75fL72KQgAAFgnrNbMAAAABIswAwAArEaYAQAAViPMAAAAqxFmAACA1QgzAADAaoQZAABgNcIMAACwGmEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFaLCHUBQDgYuGBDwP7+peNDVAkAIFjMzAAAAKsRZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBphBgAAWI0wAwAArEaYAQAAViPMAAAAqxFmAACA1QgzAADAaoQZAABgNcIMAACwGmEGAABYLeJsT1BXV6fNmzfrwgsv1ODBg1ujJiAsDFywIWB//9LxIaoEANCcoGdmbr75Zq1YsUKSdOzYMSUnJ+vmm2/W8OHD9cILL7R6gQAAAM0JOsxs3bpVo0ePliS9+OKLMsboiy++0GOPPaZHHnmk1QsEAABoTtBhpra2Vr169ZIkbdy4UZMmTVJUVJTGjx+vvXv3tnqBAAAAzQk6zMTHx+uvf/2rjhw5oo0bNyojI0OSdOjQIXXr1q3VCwQAAGhO0AuAs7KyNGXKFPXo0UMDBgxQWlqapK8vPw0bNqy16wMAAGhW0GFm1qxZGjVqlA4ePKj09HR16vT15E5iYiJrZgAAQLsL6jLTV199pcTEREVGRuqGG25Qjx49nGPjx4/XlVdeecaF5OTkyOVyKSsry2kzxsjn8ykuLk6RkZFKS0tTeXn5Gb8HAADoeIIKM126dJHf75fL5WrVIkpLS7Vq1SoNHz48oD03N1fLly/XihUrVFpaKq/Xq/T0dNXX17fq+wMAAHsFvQD43nvv1bJly3T8+PFWKeDw4cOaMmWKVq9erZ49ezrtxhjl5eVp0aJFyszM1NChQ1VQUKCjR49q7dq1rfLeAADAfkGvmfnb3/6m119/XYWFhRo2bJi6d+8ecHz9+vVBnW/27NkaP368rrnmmoA1NxUVFaqurnbulpIkt9ut1NRUlZSUaObMmac8n9/vl9/vd/br6uqCqgcAANgl6DBzzjnnaNKkSa3y5uvWrdPOnTtVWlra5Fh1dbUkyePxBLR7PB4dOHDgtOfMycnRww8/3Cr1AQCA8Bd0mMnPz2+VN66srNScOXNUWFjY7PfTfHt9jjGm2TU7Cxcu1Ny5c539uro6xcfHn33BAAAgLJ3RU7OPHz+u1157TU8//bSzGPeTTz7R4cOHW3yOHTt2qKamRiNHjlRERIQiIiJUXFysxx57TBEREc6MzMkZmpNqamqazNZ8k9vtVkxMTMAGAAA6rqBnZg4cOKDrrrtOBw8elN/vV3p6uqKjo5Wbm6svv/xSTz31VIvOc/XVV2vXrl0BbbfffrsGDRqk+fPnKzExUV6vV0VFRRoxYoQkqaGhQcXFxVq2bFmwZQMAgA4q6DAzZ84cJScn65133lHv3r2d9htuuEF33nlni88THR2toUOHBrR1795dvXv3dtqzsrKUnZ2tpKQkJSUlKTs7W1FRUZo8eXKwZQMAgA4q6DCzbds2/eUvf1HXrl0D2gcMGKCPP/641QqTpHnz5unYsWOaNWuWDh06pFGjRqmwsFDR0dGt+j4AAMBeQYeZxsZGnThxokn7Rx99dNYhY8uWLQH7LpdLPp9PPp/vrM4LAAA6rqAXAKenpysvL8/Zd7lcOnz4sB566CH9+Mc/bs3aAAAAvlPQMzO//vWvNXbsWA0ZMkRffvmlJk+erL1796pPnz763e9+1xY1AgAAnFbQYSYuLk5lZWVat26dduzYocbGRv3sZz/TlClTFBkZ2RY1AgAAnFbQYWbr1q1KSUnR7bffrttvv91pP378uLZu3aoxY8a0aoEAAADNCXrNzNixY/X55583aa+trdXYsWNbpSgAAICWCjrMnO5xAp999lmTh04CAAC0tRZfZsrMzJT09d1Lt912m9xut3PsxIkTevfdd5WSktL6FQIAADSjxWEmNjZW0tczM9HR0QGLfbt27arLL79cd911V+tXCAAA0IwWh5mTT8seOHCgHnjgAS4pAQCAsBD0mpl58+YFrJk5cOCA8vLyVFhY2KqFAQAAtETQYWbixIlas2aNJOmLL77QZZddpkcffVQTJ07UypUrW71AAACA5gQdZnbu3KnRo0dLkv7whz/I6/XqwIEDWrNmjR577LFWLxAAAKA5QYeZo0ePOg+ULCwsVGZmpjp16qTLL79cBw4caPUCAQAAmhN0mLngggv00ksvqbKyUps2bVJGRoYkqaamRjExMa1eIAAAQHOCDjP/9m//pgceeEADBw7UqFGjdMUVV0j6epZmxIgRrV4gAABAc4J+NtONN96oq666SlVVVbr44oud9quvvlo33HBDqxYHAADwXYIOM5Lk9Xrl9XoD2i677LJWKQgAACAYQYeZsWPHnvLZTCdt3rz5rAoCAAAIRtBh5pJLLgnY/+qrr1RWVqZ//OMfmjFjRmvVBQAA0CJBh5lf//rXp2z3+Xw6fPjwWRcEAAAQjKDvZjqdqVOn6j//8z9b63QAAAAt0mph5q9//au6devWWqcDAABokaAvM2VmZgbsG2NUVVWlt99+W//6r//aaoUBAAC0RNBhJjY2NmC/U6dOuvDCC7V48WLn24ABAADaS9BhJj8/vy3qAAAAOCOttmYGAAAgFFo0M9OzZ89mvyjvmz7//POzKggAACAYLQozeXl5bVwGAADAmWlRmOGbfQEAQLgKes3Mq6++qk2bNjVpLyws1J///OdWKQoAAKClgg4zCxYs0IkTJ5q0NzY2asGCBa1SFAAAQEsFHWb27t2rIUOGNGkfNGiQ9u3b1ypFAQAAtFTQYSY2NlYffvhhk/Z9+/ape/furVIUAABASwUdZq6//nplZWXpgw8+cNr27dunX/ziF7r++utbtTgAAIDvEnSY+Y//+A91795dgwYNUkJCghISEjR48GD17t1bv/rVr9qiRgAAgNM6o2czlZSUqKioSO+8844iIyM1fPhwjRkzpi3qAwAAaFbQYUaSXC6XMjIyeLAkAAAIOZ7NBAAArEaYAQAAViPMAAAAqxFmAACA1UIaZlauXKnhw4crJiZGMTExuuKKKwKe72SMkc/nU1xcnCIjI5WWlqby8vIQVgwAAMJNi8NMp06d1Llz52a3iIjgbo4677zztHTpUr399tt6++239U//9E+aOHGiE1hyc3O1fPlyrVixQqWlpfJ6vUpPT1d9fX1wnxIAAHRYLU4fL7744mmPlZSU6PHHH5cxJqg3nzBhQsD+kiVLtHLlSr311lsaMmSI8vLytGjRImVmZkqSCgoK5PF4tHbtWs2cOfOU5/T7/fL7/c5+XV1dUDUBAAC7tDjMTJw4sUnbf//3f2vhwoX605/+pClTpujf//3fz7iQEydO6Pnnn9eRI0d0xRVXqKKiQtXV1QHfZeN2u5WamqqSkpLThpmcnBw9/PDDZ1wHOr6BCzYE7O9fOj5ElQAAWsMZrZn55JNPdNddd2n48OE6fvy4ysrKVFBQoPPPPz/oc+3atUs9evSQ2+3W3XffrRdffFFDhgxRdXW1JMnj8QT093g8zrFTWbhwoWpra52tsrIy6JoAAIA9glrkUltbq+zsbD3++OO65JJL9Prrr2v06NFnVcCFF16osrIyffHFF3rhhRc0Y8YMFRcXO8ddLldAf2NMk7ZvcrvdcrvdZ1UTAACwR4vDTG5urpYtWyav16vf/e53p7zsdCa6du2qCy64QJKUnJys0tJS/eY3v9H8+fMlSdXV1erfv7/Tv6ampslsDQAA+P5qcZhZsGCBIiMjdcEFF6igoEAFBQWn7Ld+/fqzKsgYI7/fr4SEBHm9XhUVFWnEiBGSpIaGBhUXF2vZsmVn9R4AAKDjaHGYmT59erOXd87Egw8+qHHjxik+Pl719fVat26dtmzZoo0bN8rlcikrK0vZ2dlKSkpSUlKSsrOzFRUVpcmTJ7dqHQAAwF4tDjPPPvtsq7/5//zP/2jatGmqqqpSbGyshg8fro0bNyo9PV2SNG/ePB07dkyzZs3SoUOHNGrUKBUWFio6OrrVawFaE3dMAUD7Ce5b7lrZM8880+xxl8sln88nn8/XPgUBAADr8GwmAABgNcIMAACwGmEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqIf3SPKCt8U28ANDxMTMDAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFYjzAAAAKsRZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBphBgAAWI0wAwAArEaYAQAAViPMAAAAqxFmAACA1QgzAADAaoQZAABgNcIMAACwGmEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFgtpGEmJydHl156qaKjo9WvXz/95Cc/0fvvvx/Qxxgjn8+nuLg4RUZGKi0tTeXl5SGqGAAAhJuQhpni4mLNnj1bb731loqKinT8+HFlZGToyJEjTp/c3FwtX75cK1asUGlpqbxer9LT01VfXx/CygEAQLiICOWbb9y4MWA/Pz9f/fr1044dOzRmzBgZY5SXl6dFixYpMzNTklRQUCCPx6O1a9dq5syZTc7p9/vl9/ud/bq6urb9EAAAIKTCas1MbW2tJKlXr16SpIqKClVXVysjI8Pp43a7lZqaqpKSklOeIycnR7Gxsc4WHx/f9oUDAICQCZswY4zR3LlzddVVV2no0KGSpOrqakmSx+MJ6OvxeJxj37Zw4ULV1tY6W2VlZdsWDgAAQiqkl5m+6Z577tG7776rbdu2NTnmcrkC9o0xTdpOcrvdcrvdbVIjAAAIP2ExM3Pvvffq5Zdf1htvvKHzzjvPafd6vZLUZBampqamyWwNAAD4fgppmDHG6J577tH69eu1efNmJSQkBBxPSEiQ1+tVUVGR09bQ0KDi4mKlpKS0d7kAACAMhfQy0+zZs7V27Vr98Y9/VHR0tDMDExsbq8jISLlcLmVlZSk7O1tJSUlKSkpSdna2oqKiNHny5FCW3qyBCzYE7O9fOj5ElQAA0PGFNMysXLlSkpSWlhbQnp+fr9tuu02SNG/ePB07dkyzZs3SoUOHNGrUKBUWFio6OrqdqwUAAOEopGHGGPOdfVwul3w+n3w+X9sXBAAArBMWC4ABAADOFGEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGC1sHlqNtDR8ZgLAGgbzMwAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFYjzAAAAKsRZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBphBgAAWC0i1AXYbuCCDQH7+5eOD1El3y/83AEAJzEzAwAArEaYAQAAViPMAAAAqxFmAACA1VgAjA6DRcEA8P3EzAwAALAaYQYAAFiNMAMAAKxGmAEAAFYjzAAAAKsRZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBphBgAAWI0wAwAArBbSMLN161ZNmDBBcXFxcrlceumllwKOG2Pk8/kUFxenyMhIpaWlqby8PDTFAgCAsBTSMHPkyBFdfPHFWrFixSmP5+bmavny5VqxYoVKS0vl9XqVnp6u+vr6dq4UAACEq5A+NXvcuHEaN27cKY8ZY5SXl6dFixYpMzNTklRQUCCPx6O1a9dq5syZ7VkqAAAIU2G7ZqaiokLV1dXKyMhw2txut1JTU1VSUnLa1/n9ftXV1QVsAACg4wrpzExzqqurJUkejyeg3ePx6MCBA6d9XU5Ojh5++OE2rQ1oLQMXbAjY3790fIgqAQB7he3MzEkulytg3xjTpO2bFi5cqNraWmerrKxs6xIBAEAIhe3MjNfrlfT1DE3//v2d9pqamiazNd/kdrvldrvbvD4AABAewnZmJiEhQV6vV0VFRU5bQ0ODiouLlZKSEsLKAABAOAnpzMzhw4e1b98+Z7+iokJlZWXq1auXzj//fGVlZSk7O1tJSUlKSkpSdna2oqKiNHny5BBWDQAAwklIw8zbb7+tsWPHOvtz586VJM2YMUPPPvus5s2bp2PHjmnWrFk6dOiQRo0apcLCQkVHR4eqZAAAEGZCGmbS0tJkjDntcZfLJZ/PJ5/P135FAQAAq4TtmhkAAICWIMwAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKwWtg+axPfTwAUbAvb3Lx0fokpCh58BAASHmRkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFbjbqZ28s07VLg75WstvWuHu3sAAM1hZgYAAFiNMAMAAKxGmAEAAFYjzAAAAKuxANhCLIgFAOD/MTMDAACsRpgBAABWI8wAAACrEWYAAIDVWACMU2rtRcYsWgYAtBVmZgAAgNUIMwAAwGqEGQAAYDXCDAAAsBoLgEOkpQtibVw4a2PNHUFH/jsFAM1hZgYAAFiNMAMAAKxGmAEAAFYjzAAAAKuxADiMtMfCzLN5j1O9lsWk7eObP+ezHTMb+7VUR/n72FE+B9BemJkBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA17mYCd05YKJzGrD3uXArV5+UREU19nz4rTu9M77BsK1bMzDz55JNKSEhQt27dNHLkSL355puhLgkAAISJsA8zv//975WVlaVFixbp73//u0aPHq1x48bp4MGDoS4NAACEgbAPM8uXL9fPfvYz3XnnnRo8eLDy8vIUHx+vlStXhro0AAAQBsJ6zUxDQ4N27NihBQsWBLRnZGSopKTklK/x+/3y+/3Ofm1trSSprq6uTWps9B8N2K+rq/vOtpb0CbatLeulLbza2us9TyWc3vds+rVUuL9vKIRzbWg/3/4/qS2cPK8x5rs7mzD28ccfG0nmL3/5S0D7kiVLzA9/+MNTvuahhx4yktjY2NjY2Ng6wFZZWfmdeSGsZ2ZOcrlcAfvGmCZtJy1cuFBz58519hsbG/X555+rd+/ep31NMOrq6hQfH6/KykrFxMSc9flw5hiL8MFYhBfGI3wwFmfOGKP6+nrFxcV9Z9+wDjN9+vRR586dVV1dHdBeU1Mjj8dzyte43W653e6AtnPOOafVa4uJieEvZphgLMIHYxFeGI/wwVicmdjY2Bb1C+sFwF27dtXIkSNVVFQU0F5UVKSUlJQQVQUAAMJJWM/MSNLcuXM1bdo0JScn64orrtCqVat08OBB3X333aEuDQAAhIGwDzO33HKLPvvsMy1evFhVVVUaOnSoXn31VQ0YMCAk9bjdbj300ENNLmWh/TEW4YOxCC+MR/hgLNqHy5iW3PMEAAAQnsJ6zQwAAMB3IcwAAACrEWYAAIDVCDMAAMBqhJkgPPnkk0pISFC3bt00cuRIvfnmm6EuqcPLycnRpZdequjoaPXr108/+clP9P777wf0McbI5/MpLi5OkZGRSktLU3l5eYgq/v7IycmRy+VSVlaW08ZYtK+PP/5YU6dOVe/evRUVFaVLLrlEO3bscI4zHu3j+PHj+uUvf6mEhARFRkYqMTFRixcvVmNjo9OHsWhjZ/XwpO+RdevWmS5dupjVq1eb3bt3mzlz5pju3bubAwcOhLq0Du3aa681+fn55h//+IcpKysz48ePN+eff745fPiw02fp0qUmOjravPDCC2bXrl3mlltuMf379zd1dXUhrLxj2759uxk4cKAZPny4mTNnjtPOWLSfzz//3AwYMMDcdttt5m9/+5upqKgwr732mtm3b5/Th/FoH4888ojp3bu3eeWVV0xFRYV5/vnnTY8ePUxeXp7Th7FoW4SZFrrsssvM3XffHdA2aNAgs2DBghBV9P1UU1NjJJni4mJjjDGNjY3G6/WapUuXOn2+/PJLExsba5566qlQldmh1dfXm6SkJFNUVGRSU1OdMMNYtK/58+ebq6666rTHGY/2M378eHPHHXcEtGVmZpqpU6caYxiL9sBlphZoaGjQjh07lJGREdCekZGhkpKSEFX1/VRbWytJ6tWrlySpoqJC1dXVAWPjdruVmprK2LSR2bNna/z48brmmmsC2hmL9vXyyy8rOTlZN910k/r166cRI0Zo9erVznHGo/1cddVVev3117Vnzx5J0jvvvKNt27bpxz/+sSTGoj2E/TcAh4NPP/1UJ06caPJwS4/H0+QhmGg7xhjNnTtXV111lYYOHSpJzs//VGNz4MCBdq+xo1u3bp127typ0tLSJscYi/b14YcfauXKlZo7d64efPBBbd++Xffdd5/cbremT5/OeLSj+fPnq7a2VoMGDVLnzp114sQJLVmyRLfeeqsk/m20B8JMEFwuV8C+MaZJG9rOPffco3fffVfbtm1rcoyxaXuVlZWaM2eOCgsL1a1bt9P2YyzaR2Njo5KTk5WdnS1JGjFihMrLy7Vy5UpNnz7d6cd4tL3f//73eu6557R27VpddNFFKisrU1ZWluLi4jRjxgynH2PRdrjM1AJ9+vRR586dm8zC1NTUNEnaaBv33nuvXn75Zb3xxhs677zznHav1ytJjE072LFjh2pqajRy5EhFREQoIiJCxcXFeuyxxxQREeH8vBmL9tG/f38NGTIkoG3w4ME6ePCgJP5ttKd/+Zd/0YIFC/TTn/5Uw4YN07Rp03T//fcrJydHEmPRHggzLdC1a1eNHDlSRUVFAe1FRUVKSUkJUVXfD8YY3XPPPVq/fr02b96shISEgOMJCQnyer0BY9PQ0KDi4mLGppVdffXV2rVrl8rKypwtOTlZU6ZMUVlZmRITExmLdnTllVc2+ZqCPXv2OA/h5d9G+zl69Kg6dQr8ddq5c2fn1mzGoh2EcPGxVU7emv3MM8+Y3bt3m6ysLNO9e3ezf//+UJfWof385z83sbGxZsuWLaaqqsrZjh496vRZunSpiY2NNevXrze7du0yt956K7c8tpNv3s1kDGPRnrZv324iIiLMkiVLzN69e81//dd/maioKPPcc885fRiP9jFjxgxz7rnnOrdmr1+/3vTp08fMmzfP6cNYtC3CTBCeeOIJM2DAANO1a1fzox/9yLk9GG1H0im3/Px8p09jY6N56KGHjNfrNW6324wZM8bs2rUrdEV/j3w7zDAW7etPf/qTGTp0qHG73WbQoEFm1apVAccZj/ZRV1dn5syZY84//3zTrVs3k5iYaBYtWmT8fr/Th7FoWy5jjAnlzBAAAMDZYM0MAACwGmEGAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgyAVuFyufTSSy+FuoxW8+yzz+qcc84JdRkAWoAwA+A7VVdX695771ViYqLcbrfi4+M1YcIEvf76623yflu2bJHL5dIXX3zRJueXpDfeeENjx45Vr169FBUVpaSkJM2YMUPHjx+XJN1yyy3as2dPm70/gNZDmAHQrP3792vkyJHavHmzcnNztWvXLm3cuFFjx47V7NmzQ11es4wxTjj5pvLyco0bN06XXnqptm7dql27dunxxx9Xly5dnCcdR0ZGql+/fu1dMoAzQJgB0KxZs2bJ5XJp+/btuvHGG/XDH/5QF110kebOnau33nrrlK851cxKWVmZXC6X9u/fL0k6cOCAJkyYoJ49e6p79+666KKL9Oqrr2r//v0aO3asJKlnz55yuVy67bbbJH0dTnJzc5WYmKjIyEhdfPHF+sMf/tDkfTdt2qTk5GS53W69+eabTeorKipS//79lZubq6FDh+oHP/iBrrvuOv32t79V165dJTW9zDRw4EC5XK4m20kff/yxbrnlFvXs2VO9e/fWxIkTnc8KoG1FhLoAAOHr888/18aNG7VkyRJ17969yfGzWVMye/ZsNTQ0aOvWrerevbt2796tHj16KD4+Xi+88IImTZqk999/XzExMYqMjJQk/fKXv9T69eu1cuVKJSUlaevWrZo6dar69u2r1NRU59zz5s3Tr371KyUmJp6yRq/Xq6qqKm3dulVjxoxpUb2lpaU6ceKEJOnEiRO68cYb1aVLF0nS0aNHNXbsWI0ePVpbt25VRESEHnnkEV133XV69913nYAEoG0QZgCc1r59+2SM0aBBg1r93AcPHtSkSZM0bNgwSVJiYqJzrFevXpKkfv36OWHkyJEjWr58uTZv3qwrrrjCec22bdv09NNPB4SZxYsXKz09/bTvfdNNN2nTpk1KTU2V1+vV5ZdfrquvvlrTp09XTEzMKV/Tt29f589z5sxRVVWVSktLJUnr1q1Tp06d9Nvf/taZrcnPz9c555yjLVu2KCMjI9gfD4AgcJkJwGkZYyQp4HJKa7nvvvv0yCOP6Morr9RDDz2kd999t9n+u3fv1pdffqn09HT16NHD2dasWaMPPvggoG9ycnKz5+rcubPy8/P10UcfKTc3V3FxcVqyZIkuuugiVVVVNfvaVatW6ZlnntEf//hHJ+Ds2LFD+/btU3R0tFNXr1699OWXXzapDUDrI8wAOK2kpCS5XC699957Qb2uU6ev/2s5GYYk6auvvgroc+edd+rDDz/UtGnTtGvXLiUnJ+vxxx8/7TlPLszdsGGDysrKnG337t0B62YknfKS2Kmce+65mjZtmp544gknLD311FOn7b9lyxbde++9WrNmjS6++OKA2kaOHBlQV1lZmfbs2aPJkye3qBYAZ44wA+C0evXqpWuvvVZPPPGEjhw50uT46W6dPjlj8c1ZjrKysib94uPjdffdd2v9+vX6xS9+odWrV0uSs8bk5BoVSRoyZIjcbrcOHjyoCy64IGCLj48/04/o6Nmzp/r373/Kzyl9fclt0qRJevDBB5WZmRlw7Ec/+pH27t2rfv36NaktNjb2rGsD0DzCDIBmPfnkkzpx4oQuu+wyvfDCC9q7d6/ee+89PfbYY87alW87GTB8Pp/27NmjDRs26NFHHw3ok5WVpU2bNqmiokI7d+7U5s2bNXjwYEnSgAED5HK59Morr+h///d/dfjwYUVHR+uBBx7Q/fffr4KCAn3wwQf6+9//rieeeEIFBQVBfaann35aP//5z1VYWKgPPvhA5eXlmj9/vsrLyzVhwoQm/Y8dO6YJEybokksu0T//8z+rurra2SRpypQp6tOnjyZOnKg333xTFRUVKi4u1pw5c/TRRx8FVRuAM2AA4Dt88sknZvbs2WbAgAGma9eu5txzzzXXX3+9eeONN5w+ksyLL77o7G/bts0MGzbMdOvWzYwePdo8//zzRpKpqKgwxhhzzz33mB/84AfG7Xabvn37mmnTpplPP/3Uef3ixYuN1+s1LpfLzJgxwxhjTGNjo/nNb35jLrzwQtOlSxfTt29fc+2115ri4mJjjDFvvPGGkWQOHTrU7OfZuXOnmTp1qklISDBut9v07t3bjBkzxrz88stOn/z8fBMbG2uMMaaiosJIOuV2UlVVlZk+fbrp06ePcbvdJjEx0dx1112mtrY2+B84gKC4jPnGRW0AAADLcJkJAABYjTADAACsRpgBAABWI8wAAACrEWYAAIDVCDMAAMBqhBkAAGA1wgwAALAaYQYAAFiNMAMAAKxGmAEAAFb7P1oud7W0ZFmiAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clust_lengths = []\n",
    "\n",
    "min_clust_size = 4\n",
    "\n",
    "#File to list small clusters (less than min_clust_size proteins) that are filtered out\n",
    "small_cluster_filt_fname = aln_dir + os.sep + os.path.normpath('clustering/small_clusters.txt')\n",
    "\n",
    "#File to list clusters with no SC proteins\n",
    "sc_cluster_filt_fname = aln_dir + os.sep + os.path.normpath('clustering/non_sc_clusters.txt')\n",
    "\n",
    "\n",
    "with(open(small_cluster_filt_fname, 'w')) as f_small:\n",
    "    with(open(sc_cluster_filt_fname, 'w')) as f_no_sc:\n",
    "        f_small.write('OG_cluster\\tcluster_genes\\n')\n",
    "        f_no_sc.write('OG_cluster\\tcluster_size\\n')\n",
    "        for og_clust, cluster in clusters_len_filt.items():\n",
    "            no_sc_clust = True\n",
    "            clust_len = len(cluster)\n",
    "            clust_lengths.append(clust_len)\n",
    "            if clust_len>0: \n",
    "\n",
    "                if clust_len<min_clust_size: \n",
    "                    f_small.write(og_clust+'\\t[' + \",\".join(cluster) + ']\\n' )\n",
    "\n",
    "                for seq in cluster: \n",
    "                    if seq.split('_')[0]=='REF':  #This means there is an S.cer protein present\n",
    "                        no_sc_clust = False\n",
    "\n",
    "                if no_sc_clust: \n",
    "                    f_no_sc.write(og_clust + '\\t' + str(clust_len) + '\\n')\n",
    "                \n",
    "            \n",
    "clust_count = Counter(clust_lengths)\n",
    "\n",
    "print('There are ' + str(len(clusters) - clust_count[0]) + ' nonzero clusters')\n",
    "\n",
    "sizes = []\n",
    "nclusts = []\n",
    "for size, nclust in clust_count.items(): \n",
    "    if size>0:\n",
    "        sizes.append(size)\n",
    "        nclusts.append(nclust)\n",
    "\n",
    "#Plot Histogram of all nonzero clusters\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.bar(sizes, nclusts)\n",
    "ax.set_xlabel('Cluster Size')\n",
    "ax.set_ylabel('N clusters')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#List of clusters that don't have an SC protein but have more than a few proteins present.  \n",
    "no_sc_clusters = pd.read_table(sc_cluster_filt_fname)\n",
    "no_sc_clusters_filt = no_sc_clusters[no_sc_clusters['cluster_size']>2].sort_values(by='cluster_size', ascending=False)\n",
    "\n",
    "#'OG5006_C1' is strange - 5 different proteins from K.mar.  The S. cer orthoglog is COB, a mitochondrial encoded Cytochrome B Oxidase, as well as an overlapping transcript, BI3 which is a Maturase which mediates splicing of an intron.  \n",
    "# OG4755_C1 also has many K.mar genes and is COX2\n",
    "# OG5438 is ortholog of BIO3 (biotin synthesis)\n",
    "# OG5722 is ADH4 and has duplicates in a. hylecoeti"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make a file to list all the clusters.  Also assigns first reference structure in a cluster as the representative (so we don't have redundant clusters)\n",
    "#If there is no S. cer protein present it picks the tm_align assigned reference structure (which had a *) for the name - see clusters_ref\n",
    "cluster_name_map = {}\n",
    "with open(aln_dir + os.sep + ('clustering/cluster_mapping.csv'), 'w') as f_out: \n",
    "    f_out.write('primary_id,og_clust,cluster_rep,no_sc\\n')\n",
    "    for og_clust, cluster in clusters_len_filt.items():\n",
    "        if len(cluster)>0:\n",
    "            og, clust = og_clust.split('_')\n",
    "            if og_clust in set(no_sc_clusters_filt['OG_cluster']):  #For clusters with no S. cer reference ##Use line with star as a reference\n",
    "                seq = clusters_ref[og_clust]\n",
    "                cluster_name_map[seq]=og_clust\n",
    "                ref = seq.split('.')[0]\n",
    "                f_out.write(og+ '_' + ref + ',' + og_clust + ',1,TRUE\\n')\n",
    "            else: #For clusters that have an S.cer reference\n",
    "                cluster_rep = 1\n",
    "                for seq in cluster: #lists same cluster multiple times if there are multiple S.cer refs present\n",
    "                    if seq.split('_')[0]=='REF':\n",
    "                        cluster_name_map[seq] = og_clust \n",
    "                        ref = seq.split('.')[0]\n",
    "                        f_out.write(og+ '_' + ref + ',' + og_clust + ',' + str(cluster_rep) + ',FALSE\\n')\n",
    "                        cluster_rep = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "small_cluster_filt_fname = aln_dir + os.sep + os.path.normpath('clustering/small_clusters.txt')\n",
    "small_clusters = pd.read_table(small_cluster_filt_fname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'OG1004_C0',\n",
       " 'OG1060_C2',\n",
       " 'OG1070_C2',\n",
       " 'OG1273_C2',\n",
       " 'OG1287_C2',\n",
       " 'OG1310_C0',\n",
       " 'OG1354_C1',\n",
       " 'OG1380_C0',\n",
       " 'OG1394_C1',\n",
       " 'OG1468_C1',\n",
       " 'OG1597_C1',\n",
       " 'OG1603_C2',\n",
       " 'OG1640_C1',\n",
       " 'OG1714_C0',\n",
       " 'OG1714_C1',\n",
       " 'OG1819_C2',\n",
       " 'OG1897_C0',\n",
       " 'OG2048_C2',\n",
       " 'OG2071_C1',\n",
       " 'OG2129_C2',\n",
       " 'OG2147_C0',\n",
       " 'OG2227_C2',\n",
       " 'OG2290_C1',\n",
       " 'OG2646_C0',\n",
       " 'OG2901_C0',\n",
       " 'OG3000_C0',\n",
       " 'OG3030_C1',\n",
       " 'OG3074_C0',\n",
       " 'OG3104_C1',\n",
       " 'OG3151_C1',\n",
       " 'OG3313_C0',\n",
       " 'OG3581_C2',\n",
       " 'OG4312_C0',\n",
       " 'OG5006_C0',\n",
       " 'OG5006_C1',\n",
       " 'OG5314_C0',\n",
       " 'OG5314_C1',\n",
       " 'OG5735_C1',\n",
       " 'OG7818_C1'}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(small_clusters['OG_cluster'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['eremothecium_gossypii__OG7818__40_277.pdb',\n",
       " 'kluyveromyces_lactis__OG7818__50_4420.pdb',\n",
       " 'lachancea_thermotolerans__OG7818__64_3371.pdb']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_len_filt[og_clust]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rename the TM-align structural alignments to match our orignal filenames. \n",
    "#Filter out\n",
    "#  OGs that had too few sequences\n",
    "#  peptide sequences that were too short\n",
    "\n",
    "#pep_seq_dir = base_dir + os.sep + os.path.normpath('og_sequences/proteome')\n",
    "\n",
    "clust_align_dir = aln_dir + os.sep + 'fasta' + os.sep\n",
    "clust_align_rename_dir = aln_dir + os.sep + 'fasta_renamed' + os.sep\n",
    "#clust_align_rename_all_dir = tm_align_dir + os.sep + 'fasta_renamed_all' + os.sep\n",
    "\n",
    "cluster_mapping = pd.read_csv(aln_dir + os.sep + os.path.normpath('clustering/cluster_mapping.csv'))\n",
    "cluster_mapping_dict = dict(zip(cluster_mapping['primary_id'], cluster_mapping['og_clust']))\n",
    "\n",
    "\n",
    "#Option 1:  File only for representative clusters: \n",
    "for (ind, (og_ref,og_clust, cluster_rep, no_sc)) in cluster_mapping.iterrows():\n",
    "    if cluster_rep == 1: \n",
    "        if not(og_clust in set(small_clusters['OG_cluster'])):\n",
    "            og, clust = og_clust.split('_')\n",
    "            \n",
    "            full_cluster_in_fname =  clust_align_dir + og+clust + '.fa'\n",
    "            \n",
    "            full_cluster_in = SeqIO.parse(full_cluster_in_fname, 'fasta')\n",
    "            \n",
    "            clust_list = clusters_len_filt[og_clust]\n",
    "            \n",
    "            with open(clust_align_rename_dir+og_ref + '.tm.fasta', 'w') as f_out: \n",
    "                for record in full_cluster_in: \n",
    "                    if record.id in clust_list: \n",
    "                        f_out.write('>' + record.description + '\\n')\n",
    "                        f_out.write(str(record.seq) + '\\n')\n",
    "                        \n",
    "# #Option 2: File for every cluster that is present in the main data set (and some are redundant)\n",
    "# for prot_fname in os.listdir(pep_seq_dir):\n",
    "#     og_ref = prot_fname.split('.')[0]\n",
    "#     clust = cluster_mapping_dict[og_ref]\n",
    "#     og, clust = clust.split('_')\n",
    "#     shutil.copyfile(clust_align_dir + og+clust + '.fa', clust_align_rename_all_dir + og_ref + '.tm.fasta')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Build a file with each proteome sequence and coding sequence for each item in the MSA\n",
    "\n",
    "#Make big dictionary for selected coding sequences \n",
    "cds_dir = base_dir + os.sep + os.path.normpath('selected_proteins/cds') + os.sep\n",
    "\n",
    "selected_cds_seqs = {}\n",
    "\n",
    "for cds_file in os.listdir(cds_dir):\n",
    "    cds_fasta = SeqIO.parse(cds_dir + cds_file, 'fasta')\n",
    "    for record in cds_fasta: \n",
    "        selected_cds_seqs[record.id] = str(record.seq)\n",
    "\n",
    "#Iterate through each msa \n",
    "clust_align_rename_dir = aln_dir + os.sep + 'fasta_renamed' + os.sep\n",
    "\n",
    "seq_dir = base_dir + os.sep + 'selected_proteins' + os.sep +  'og_sequences'\n",
    "\n",
    "for msa_file in os.listdir(clust_align_rename_dir):\n",
    "    cds_file_out = seq_dir + os.sep + 'cds_tm_5' + os.sep + msa_file.split('.')[0] + '.cds.fasta'\n",
    "    pep_file_out = seq_dir + os.sep + 'proteome_tm_5' + os.sep + msa_file.split('.')[0] + '.pep.fasta'\n",
    "    \n",
    "    msa_file_fasta = SeqIO.parse(clust_align_rename_dir + msa_file, 'fasta')\n",
    "    \n",
    "    with open(cds_file_out,'w') as f_out_cds: \n",
    "        with open(pep_file_out,'w') as f_out_pep:\n",
    "            for record in msa_file_fasta: \n",
    "                 #Squeeze out dashes and record protein sequences\n",
    "                pep_seq = ''.join([res for res in record.seq if res !='-'])\n",
    "                f_out_pep.write('>' + record.id + '\\n')\n",
    "                f_out_pep.write(pep_seq + '\\n') \n",
    "                \n",
    "                #Write CDS sequences\n",
    "                try:\n",
    "                    f_out_cds.write('>' + record.id + '\\n')\n",
    "                    f_out_cds.write(selected_cds_seqs[record.id.split('.')[0]] + '\\n') #Remove .pdb from name\n",
    "                except KeyError:\n",
    "                    print(record.id +  ' cds not found for ' + msa_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Thread alignments inside inside singularity container\n",
    "#call_code_singularity.sh\n",
    "#which calls /home/heineike_wsl2/github_s/diverse_yeast/20221206_struct_align_dnds_msa.sh\n",
    "#which calls /home/heineike_wsl2/github_s/diverse_yeast/20221206_struct_align_dnds_msa.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Sequence MSAs - also describe structural MSAs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Original code in case we want to go back and do it for Oliver's alignments\n",
    "\n",
    "# #Build a big dictionary for selected coding sequences and make a file for each og_ref name with corresponding cds\n",
    "\n",
    "# #Make a dictionary of coding sequences\n",
    "# cds_dir = base_dir + os.sep + os.path.normpath('selected_proteins/cds') + os.sep\n",
    "\n",
    "# selected_cds_seqs = {}\n",
    "\n",
    "# for cds_file in os.listdir(cds_dir):\n",
    "#     cds_fasta = SeqIO.parse(cds_dir + cds_file, 'fasta')\n",
    "#     for record in cds_fasta: \n",
    "#         selected_cds_seqs[record.id] = str(record.seq)\n",
    "\n",
    "# #Iterate through each msa and make a codon sequence fasta (unaligned)\n",
    "\n",
    "# seq_dir = base_dir + os.sep + 'og_sequences'\n",
    "\n",
    "# for protein_file in os.listdir(seq_dir + os.sep + 'proteome' ):\n",
    "#     cds_file_out = seq_dir + os.sep + 'cds' + os.sep + protein_file.split('.')[0] + '.cds.fasta'\n",
    "    \n",
    "#     protein_file_full = seq_dir + os.sep+ 'proteome' + os.sep + protein_file    \n",
    "#     protein_file_fasta = SeqIO.parse(protein_file_full, 'fasta')\n",
    "    \n",
    "#     with open(cds_file_out,'w') as f_out_cds: \n",
    "#         for record in protein_file_fasta: \n",
    "#             try:\n",
    "#                 f_out_cds.write('>' + record.id + '\\n')\n",
    "#                 f_out_cds.write(selected_cds_seqs[record.id] + '\\n') \n",
    "#             except KeyError:\n",
    "#                 print(record.id +  'cds not found for ' + protein_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "##This is improved above\n",
    "# #Make corresponding codon alignments. \n",
    "\n",
    "# #First make cds lists that contain the same sequences as the clusters\n",
    "# tm_align_dir = base_dir + os.sep +  os.path.normpath('msas/structural/tm_align')\n",
    "# clust_align_rename_dir = tm_align_dir + os.sep + 'fasta_renamed' + os.sep\n",
    "# cds_seq_dir = base_dir + os.sep + os.path.normpath('og_sequences/cds') + os.sep\n",
    "# cds_seq_dir_out = tm_align_dir + os.sep + 'cds' + os.sep\n",
    "\n",
    "\n",
    "\n",
    "# for tm_align_fname in os.listdir(clust_align_rename_dir): \n",
    "#     tm_align = SeqIO.parse(clust_align_rename_dir + tm_align_fname, 'fasta')\n",
    "#     og_ref_base = tm_align_fname.split('.')[0] \n",
    "#     cds_seq_fname = cds_seq_dir + og_ref_base + '.cds.fasta'\n",
    "#     cds_seq = SeqIO.parse(cds_seq_fname, 'fasta')\n",
    "#     cds_seq_lookup = {}\n",
    "#     for record_cds in cds_seq: \n",
    "#         cds_seq_lookup[record_cds.id] = str(record_cds.seq)\n",
    "    \n",
    "#     cds_seq_fname_out = cds_seq_dir_out + og_ref_base + '.tm_present.cds.fasta'\n",
    "#     with open(cds_seq_fname_out, 'w') as f_out:\n",
    "#         for record in tm_align: \n",
    "#             record_id = record.id.split('.')[0]\n",
    "#             try: \n",
    "#                 f_out.write('>' + record_id + '\\n')\n",
    "#                 f_out.write(cds_seq_lookup[record_id] + '\\n')\n",
    "#             except KeyError:\n",
    "#                 print('No cds file for ' + record_id + ' in ' + tm_align_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize phylogenetic trees of orthogroups that are broken apart by clustering with cluster colored in\n",
    "\n",
    "species_table = pd.read_csv(base_dir + os.sep + 'species_selection.csv')\n",
    "species_table_short = species_table.loc[species_table['Load']=='Y', ['original_genome_id', 'jacobs_name']]\n",
    "spec_time_tree_2_orig = dict(zip(species_table_short['jacobs_name'], species_table_short['original_genome_id']))\n",
    "\n",
    "out_image_dir = base_dir + os.sep + os.path.normpath('msas/structural/tm_align/clustering/cluster_trees')+ os.sep\n",
    "\n",
    "\n",
    "ogs_w_subclusters = []\n",
    "for og_clust in list(set(cluster_mapping['og_clust'])): \n",
    "    og, clust = og_clust.split('_')\n",
    "    if clust=='C2': \n",
    "        ogs_w_subclusters.append(og)\n",
    "\n",
    "for og in ogs_w_subclusters: #og = 'OG1004'\n",
    "\n",
    "    cluster_names_og = {}\n",
    "\n",
    "    for cluster in ['C1','C2']: \n",
    "        cluster_names_og[cluster] = [clust_name.split('.')[0] for clust_name in clusters[og+'_' + cluster]]\n",
    "\n",
    "    tree_dir = base_dir + os.sep + 'msas' + os.sep + 'ogs_pep_trees'\n",
    "    fname_tree = tree_dir + os.sep + og + '.mfaa.mafft.clipkit.treefile'\n",
    "    t = Tree(fname_tree, format=0)  \n",
    "    ts = TreeStyle()\n",
    "    ts.show_leaf_name = False\n",
    "    ts.show_branch_length = True\n",
    "\n",
    "    red_node = NodeStyle()\n",
    "    red_node['bgcolor'] = 'red'\n",
    "\n",
    "    blue_node = NodeStyle()\n",
    "    blue_node['bgcolor']= 'LightSteelBlue'\n",
    "\n",
    "\n",
    "    for node in t.traverse(): \n",
    "        if node.is_leaf():\n",
    "            (spec_time_tree, y1000_id) = node.name.split('|')\n",
    "            spec_orig = spec_time_tree_2_orig[spec_time_tree]\n",
    "            if spec_orig in ['saccharomyces_cerevisiae', 'candida_albicans', 'schizosaccaromyces_pombe']:  #probably none in Pombe\n",
    "                print(node.name)\n",
    "            else:\n",
    "                new_name = spec_orig + '__' + og + '__' + y1000_id\n",
    "\n",
    "                if new_name in cluster_names_og['C1']:\n",
    "                    node.set_style(blue_node)\n",
    "                elif new_name in cluster_names_og['C2']: \n",
    "                    node.set_style(red_node)\n",
    "\n",
    "                node.name = new_name\n",
    "                name_face = AttrFace(\"name\",fsize=20)\n",
    "                node.add_face(name_face, column=0, position=\"branch-right\") \n",
    "                #convert name to primary id\n",
    "\n",
    "    t.ladderize()\n",
    "    #t.render('%%inline', tree_style=ts)\n",
    "    t.render(out_image_dir + og + '.pdf', tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "No C2: \n",
    "1603\n",
    "1060 \n",
    "\n",
    "1168 - a lot missing\n",
    "\n",
    "C1 clusters not together: \n",
    "1354, 1394, 1468, 1597, 2071, 2290, 2357 (interesting), 3030, 3104, 3151\n",
    "\n",
    "5735 - how did this pass our filter - not very many sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nir spotted some orthogroups in which Pombe genes are shared in multiple orthogroups\n",
    "\n",
    "OG1\tOG2\n",
    "Spom_AF-Q10142-F1-model_v2.pdb\tOG1242\tOG4624\n",
    "Spom_AF-P55306-F1-model_v2.pdb\t        OG1310\tOG5404\n",
    "Spom_AF-O14192-F1-model_v2.pdb\tOG1377\tOG3639\n",
    "Spom_AF-Q09755-F1-model_v2.pdb\tOG2032\tOG4665\n",
    "Spom_AF-Q09741-F1-model_v2.pdb\tOG2508\tOG3801\n",
    "Spom_AF-O94413-F1-model_v2.pdb\tOG2588\tOG2968\n",
    "Spom_AF-P33075-F1-model_v2.pdb\t        OG2740\tOG5118\n",
    "Spom_AF-Q7LKX0-F1-model_v2.pdb\tOG3021\tOG3909\n",
    "Spom_AF-O13917-F1-model_v2.pdb\tOG4317\tOG4398\n",
    "\n",
    "Oliver sees it is in the og_metadata\n",
    "\n",
    "OGs sharing same pathway (e.g. for the first three)\n",
    "Spom_AF-Q10142-F1-model_v2.pdb\tOG1242\tOG4624\t\t\"['SPHINGOLIPID-SYN-PWY-1']\"\n",
    "Spom_AF-P55306-F1-model_v2.pdb\t       OG1310\tOG5404\t\t\"['DETOX1-PWY']\"\n",
    "Spom_AF-O14192-F1-model_v2.pdb\tOG1377\tOG3639 \t\"['ALL-CHORISMATE-PWY-1', 'COMPLETE-ARO-PWY-1', 'PWY3O-351', 'PWY3O-4108', 'PWY3O-4153']\"\n",
    "\n",
    "\n",
    "S.pombe Q10142 is https://www.uniprot.org/uniprotkb/Q10142/entry\n",
    "AUR1 Ortholog of AUR1 in S.cer\n",
    "\n",
    "OG124 is AUR1 in S.cer\n",
    "\n",
    "OG4624 is IPT1 in S.cer which also looks like a distant ortholog to AUR1\n",
    "\n",
    "http://eggnog5.embl.de/#/app/results?seqid=Q10142&target_nogs=ENOG502QPQM#ENOG5028HC7_datamenu\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Model species info\n",
    "\n",
    "#C. albicans\n",
    "#http://www.candidagenome.org/download/sequence/C_albicans_SC5314/Assembly22/current/  \n",
    "# C_albicans_SC5314_A22_current_default_coding.fasta\n",
    "# C_albicans_SC5314_A22_current_default_protein.fasta\n",
    "#downloaded on 20221012 and stored in\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\Candida_albicans\n",
    "\n",
    "\n",
    "#S. cerevisiae\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\Saccharomyces_cerevisiae\\S288C_reference_genome_R64-2-1_20150113\n",
    "#orf_coding_all_R64-2-1_20150113.fasta\n",
    "#orf_trans_all_R64-2-1_20150113.fasta\n",
    "\n",
    "#S. pombe\n",
    "#From https://www.pombase.org/data/genome_sequence_and_features/feature_sequences/ on 20221012\n",
    "#cds.fa \n",
    "#peptide.fa  \n",
    "#both last modified 2022-10-12 03:26\n",
    "#Stored at\n",
    "#G:\\My Drive\\Crick_LMS\\external_data\\genomes\\Schizosaccharomyces_pombe"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
