{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load oliver's structural alignment\n",
    "import pickle\n",
    "import os\n",
    "import pandas as pd\n",
    "#from Bio import SeqIO\n",
    "from Bio import AlignIO, Align\n",
    "#from Bio.Align import MultipleSeqAlignment\n",
    "\n",
    "#y1000plus_tools_flag = input(\"Use y1000plus_tools functions? (Y/n)\")\n",
    "\n",
    "# if y1000plus_tools_flag =='Y':\n",
    "#     y1000plus_tools_path = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools') + os.sep\n",
    "#     #io_library_path_core = io_library_path + 'core' + os.sep\n",
    "#     if not(y1000plus_tools_path in sys.path):\n",
    "#         sys.path.append(y1000plus_tools_path)\n",
    "#         print(\"Added \" + y1000plus_tools_path + \" to path\" )\n",
    "\n",
    "#     import y1000plus_tools\n",
    "#     y1000plus_tools.base_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools') + os.sep\n",
    "#     print(\"y1000plus_tools dir is \" + y1000plus_tools.base_dir )                      \n",
    "#     y1000plus_tools.y1000plus_dir = y1000plus_tools.base_dir + 'data' + os.sep \n",
    "#     print(\"y1000plus data dir is \" + y1000plus_tools.y1000plus_dir)\n",
    "\n",
    "\n",
    "y1000plus_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools/data') + os.sep\n",
    "\n",
    "#Load y1000 species table\n",
    "y1000_species = pd.read_csv(y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/y1000_species_table.csv\"), index_col=0)\n",
    "\n",
    "#fig2_orig_genome_lookup = dict(zip(y1000_species['species_names_fig2'], y1000_species['original_genome_id']))\n",
    "old_orig_genome_lookup = dict(zip(y1000_species['old_species_names'], y1000_species['original_genome_id']))\n",
    "\n",
    "\n",
    "goi= 'HEM14'\n",
    "\n",
    "\n",
    "base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts')\n",
    "divyeast_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/diverse_yeast')\n",
    "ex_dict = {'CDC19': \n",
    "               { 'struct_pkl': os.path.normpath('alphafold\\msas\\REF_Scer_AF-P00549-F1-model_v2.pkl'),\n",
    "                'og': 'OG1299' \n",
    "                #'seq_aln_orig': os.path.normpath('alphafold\\msas\\FILES_ogs_pep_aligned\\OG1299.mfaa.mafft')\n",
    "                 \n",
    "                },\n",
    "            'HEM14': \n",
    "               { 'struct_pkl': os.path.normpath('alphafold\\msas\\REF_Scer_AF-P40012-F1-model_v2.pkl'), \n",
    "                 'og':'OG1273'\n",
    "                 #'seq_aln_orig': os.path.normpath('alphafold\\msas\\FILES_ogs_pep_aligned\\OG1299.mfaa.mafft'), \n",
    "               }\n",
    "          }\n",
    "\n",
    "\n",
    "og = ex_dict[goi]['og']\n",
    "\n",
    "struct_data_fname = base_dir + os.sep + (ex_dict[goi]['struct_pkl'])\n",
    "\n",
    "og_data = pickle.load(open(struct_data_fname,\"rb\"))\n",
    "# A[“Molecules”] contains all molecular coordinates and information\n",
    "# A[“Molecules”][“REF_Scer_AF-P00549-F1-model_v2\"][“Ligand Indices”] contains the indices for binding site residues for the reference structure\n",
    "# NOTE: The indexing comes from the *.pdb, thus starts at 1 (and not at 0)\n",
    "# From visual inspection: Binding site 1: indices <= 298 (last); Binding site 2: indices >= 401 (first)\n",
    "# A[“Mapping”] contains all mapped information\n",
    "# NOTE: Where no mapping could be found either “XYZ” (Amino Acids) or “NoMatch” (else) was used\n",
    "# A[“Mapping”][“Mapping”] contains the indices for mapping\n",
    "# NOTE: Since this is Python generated, indexing starts at 0, -1 denotes no mapping found\n",
    "# A[“Mapping Colors”] is only for my plots and not useful for you at the moment (IMO)\n",
    "#If there are any questions, give me a note"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "fname_seq_aln_orig = base_dir + os.sep +  os.path.normpath(\"alphafold\\msas\\FILES_ogs_pep_aligned\\\"\" + og + '.mfaa.mafft')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['REF_Scer_AF-P40012-F1-model_v2',\n",
       " 'Calb_AF-A0A1D8PEF0-F1-model_v2',\n",
       " 'Spom_AF-Q10062-F1-model_v2',\n",
       " 'alloascoidea_hylecoeti__OG1273__0_850',\n",
       " 'ascoidea_rubescens__OG1273__6_3529',\n",
       " 'candida_apicola__OG1273__13_90',\n",
       " 'cyberlindnera_jadinii__OG1273__35_845',\n",
       " 'eremothecium_gossypii__OG1273__40_1859',\n",
       " 'geotrichum_candidum__OG1273__43_2414',\n",
       " 'kazachstania_naganishii__OG1273__49_4322',\n",
       " 'lipomyces_starkeyi__OG1273__66_4579',\n",
       " 'ogataea_parapolymorpha__OG1273__104_1602',\n",
       " 'pachysolen_tannophilus__OG1273__106_2860',\n",
       " 'torulaspora_delbrueckii__OG1273__135_637',\n",
       " 'vanderwaltozyma_polyspora__OG1273__136_3359',\n",
       " 'yHMPu5000034604_sporopachydermia_lactativora_160519__OG1273__170_486',\n",
       " 'yarrowia_lipolytica__OG1273__144_4673']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_data['Mapping']['Keys']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Edit cdc19 alignment file to include only sequences contained in the structural alignment\n",
    "#Convert names to match Structural alignment names\n",
    "\n",
    "#\n",
    "\n",
    "\n",
    "seqs_to_remove = {'OG1299': \n",
    "                    ['Alloascoidea_hylecoeti|0_2015',\n",
    "                    'Vanderwaltozyma_polyspora|136_2033',\n",
    "                    'Zygosaccharomyces_rouxii|342_99',\n",
    "                    'Kluyveromyces_lactis|50_393',\n",
    "                    'Kluyveromyces_marxianus|51_4988',\n",
    "                    'Lachancea_thermotolerans|64_2570',\n",
    "                    'yHMPu5000034957_Hanseniaspora_osmophila|247_3065',\n",
    "                    'Wickerhamomyces_anomalus|139_1240',\n",
    "                    'Komagataella_pastoris|52_2808',\n",
    "                    'Debaryomyces_hansenii|36_3767',\n",
    "                    'Candida_tropicalis|30_3271'\n",
    "                    ], \n",
    "                  'OG1273': \n",
    "                  ['Alloascoidea_hylecoeti|0_850', # looks bad - known bug\n",
    "                   'Candida_albicans|12_1927', #C1_10180C_B - ECM21 Ortholog #The next three look very strange - not sure how they got into the list C. albicans, C. tropicalis, D. hansenii.  How do they get included in the Orthogroup?\n",
    "                   'Candida_tropicalis|30_5966',\n",
    "                   'Debaryomyces_hansenii|36_1610',\n",
    "                   'Wickerhamomyces_anomalus|139_2647',\n",
    "                   'yHMPu5000034957_Hanseniaspora_osmophila|247_2887',  #These were not present on the Structural alignment list\n",
    "                   'Candida_tropicalis|30_2375',\n",
    "                   'Debaryomyces_hansenii|36_1877',\n",
    "                   'Kluyveromyces_lactis|50_4648',\n",
    "                   'Kluyveromyces_marxianus|51_1602',\n",
    "                   'Komagataella_pastoris|52_2088',\n",
    "                   'Lachancea_thermotolerans|64_3902'\n",
    "                  ]\n",
    "\n",
    "                 }\n",
    "\n",
    "                   \n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "seqnames_to_conv = {'OG1299': \n",
    "                    {'Saccharomyces_cerevisiae|110_4286' : 'REF_Scer_AF-P00549-F1-model_v2',  #CDC19  \n",
    "                    'Saccharomyces_cerevisiae|110_1977' : 'REF_Scer_AF-P52489-F1-model_v2', #PYK2\n",
    "                    'Candida_albicans|12_2204': 'Calb_AF-P46614-F1-model_v2'\n",
    "                    },\n",
    "                    'OG1273': \n",
    "                    {'Saccharomyces_cerevisiae|110_1590': 'REF_Scer_AF-P40012-F1-model_v2',\n",
    "                     'Candida_albicans|12_3920':  'Calb_AF-A0A1D8PEF0-F1-model_v2'  #Candida_albicans|12_3920  #C1_08880W_A   # HEM14  https://www.uniprot.org/uniprotkb/A0A1D8PEF0/entry    \n",
    "                    }\n",
    "                   }\n",
    "\n",
    "\n",
    "\n",
    "fname_seq_aln_orig = base_dir + os.sep + os.path.normpath('alphafold\\msas\\FILES_ogs_pep_aligned\\\\' + og + '.mfaa.mafft')\n",
    "fname_seq_aln_out_filt = base_dir + os.sep + os.path.normpath('alphafold\\msas\\\\' + og + '_seq_filt.fasta')\n",
    "fname_seq_aln_out = base_dir + os.sep + os.path.normpath('alphafold\\msas\\\\' + og + '_seq.fasta')\n",
    "\n",
    "# convert fname to structure alignment names\n",
    "# remove missing sequences from alignment\n",
    "##remove columns with no AAs?\n",
    "\n",
    "seq_align = AlignIO.read(fname_seq_aln_orig, 'fasta')\n",
    "with open(fname_seq_aln_out_filt, 'w') as f_out: \n",
    "    for record in seq_align: \n",
    "        name_orig = record.id\n",
    "        if not(name_orig in seqs_to_remove[og]):\n",
    "            if name_orig in seqnames_to_conv[og].keys():\n",
    "                #(spec, y1000_id) = name_orig.split('|')\n",
    "                #orig_genome_id = old_orig_genome_lookup[spec]\n",
    "                pdb_name = seqnames_to_conv[og][name_orig]\n",
    "            else: \n",
    "                (spec, y1000_id) = name_orig.split('|')\n",
    "                orig_genome_id = old_orig_genome_lookup[spec]\n",
    "                pdb_name = orig_genome_id + '__' + og + '__' + y1000_id\n",
    "\n",
    "            f_out.write('>' + pdb_name + '\\n')\n",
    "            f_out.write(str(record.seq) + '\\n')\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_align_filt = AlignIO.read(fname_seq_aln_out_filt, 'fasta')\n",
    "\n",
    "with open(fname_seq_aln_out, 'w') as f_out: \n",
    "    \n",
    "    #Obtain indices of sequences that are not all gaps\n",
    "    aln_length = seq_align_filt.get_alignment_length()\n",
    "    N_seqs = len(seq_align_filt)\n",
    "    all_gaps = '-'*N_seqs\n",
    "\n",
    "    degap_inds = []\n",
    "    for jj in range(0,aln_length):\n",
    "        if not(seq_align_filt[:,jj] == all_gaps):\n",
    "            degap_inds.append(jj)\n",
    "\n",
    "\n",
    "    for record in seq_align_filt:\n",
    "        f_out.write('>' + record.id + '\\n')\n",
    "        degapped_seq = ''.join([record.seq[jj] for jj in degap_inds])\n",
    "        f_out.write(degapped_seq + '\\n')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make fasta alignment file out of amino acid mapping for CDC19\n",
    "\n",
    "\n",
    "\n",
    "fname_struct_aln = base_dir + os.sep + os.path.normpath('alphafold\\msas\\\\' + og + '_struct.fasta')\n",
    "\n",
    "seqs_to_remove_struct = {'OG1299': ['alloascoidea_hylecoeti__OG1299__0_2015', #Need to remove this because of the a. hylecoeti bug\n",
    "                                    'Spom_AF-Q10208-F1-model_v2'    # The sequence Alignment doesn't contain the S.pombe file.  \n",
    "                                   ],\n",
    "                         'OG1273': [ 'alloascoidea_hylecoeti__OG1273__0_850', \n",
    "                                     'Spom_AF-Q10062-F1-model_v2'    #https://www.uniprot.org/uniprotkb/Q10062/entry SPAC1F5.07c Pombe:  https://www.pombase.org/gene/SPAC1F5.07c\n",
    "                                   ]\n",
    "                        }\n",
    "                             \n",
    "                         \n",
    "#Load aa_table\n",
    "aa_table = pd.read_table(divyeast_dir + os.sep + 'aa_table.txt')\n",
    "aa_dict = dict(zip(aa_table['Three_letter_caps'], aa_table['Single_letter']))\n",
    "\n",
    "struct_aln_lines = {}\n",
    "for jj, name in enumerate(og_data['Molecules'].keys()):\n",
    "    if not(name in seqs_to_remove_struct[og]):\n",
    "        msa_line = ''\n",
    "        for aa in og_data['Mapping']['Amino Acid'][jj]: \n",
    "            aa_sl = aa_dict[aa]\n",
    "            msa_line = msa_line + aa_sl\n",
    "        struct_aln_lines[name] = msa_line\n",
    "            \n",
    "seq_align_out = AlignIO.read(fname_seq_aln_out, 'fasta')      \n",
    "names_seq = [record.id for record in seq_align_out]\n",
    "\n",
    "#checks to see if same names in both alignments\n",
    "assert (len(set(names_seq)-set(struct_aln_lines.keys()))==0), 'More names in sequence based alignment than structure based alignment ' \n",
    "assert (len(set(struct_aln_lines.keys()-set(names_seq)))==0), 'More names in structure based alignment than sequence based alignment ' \n",
    "\n",
    "        \n",
    "with open(fname_struct_aln, 'w') as f_out:\n",
    "    seq_align_out = AlignIO.read(fname_seq_aln_out, 'fasta')\n",
    "    for record in seq_align_out:\n",
    "        name = record.id        \n",
    "        f_out.write('>' + name + '\\n')\n",
    "        f_out.write(struct_aln_lines[name] + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Translate Names for Fastas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
