{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8b6ecbea-9bc2-4b28-9211-a389d72238aa",
   "metadata": {},
   "source": [
    "Align mitochondrial genes based on mitochondrial sequences from Wolters et al 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eedd0f84-1087-4d68-a386-c7ee4d4b9d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "\n",
    "from ete3 import Tree\n",
    "\n",
    "import shutil\n",
    "\n",
    "import diverse_yeast_tools as dyt  \n",
    "\n",
    "\n",
    "genomes_dir = dyt.genomes_dir\n",
    "mito_aln_dir = genomes_dir + os.sep + os.path.normpath(\"yeast_mitochondrial/Finalized_Assemblies/gene_alignments/ind_genes\")\n",
    "y1000plus_dir = dyt.y1000plus_dir\n",
    "\n",
    "#Load y1000 species table\n",
    "y1000_species = pd.read_csv(y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/y1000_species_table.csv\"), index_col=0)\n",
    "\n",
    "base_dir = dyt.base_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b80b0f79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #mito_gois is the corse set of genes shared by all species (per Wolters et al)\n",
    "# mito_gois = ['cox1','cox2','cox3','cob','atp6','atp8','atp9']\n",
    "\n",
    "# name_errata = {goi: {'substitutions': {}, 'missing_species':set()} for goi in mito_gois}\n",
    "\n",
    "# for goi in mito_gois: \n",
    "#     name_errata[goi]['substitutions'] = {'Pachysolen__tannophilus_' + goi: 'Pachysolen tannophilus' , \n",
    "#                                         'Candida__apicola_' + goi: 'Starmerella apicola'\n",
    "#                                         }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85add6d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #old version\n",
    "# #Load selected species: \n",
    "# species_set = pd.read_csv(os.path.normpath(base_dir + '/selected_proteins/species_selection/species_selection.csv'), index_col = 0)\n",
    "# species_set_filt = species_set.loc[species_set['Load']=='Y']\n",
    "# species_names = set(species_set_filt.index)\n",
    "\n",
    "# mito_encoded_genes = os.listdir(mito_aln_dir)\n",
    "\n",
    "# og_list = {}\n",
    "\n",
    "# for goi in mito_gois: #os.listdir(mito_aln_dir):\n",
    "#   print(goi)\n",
    "#   #goi = 'atp6'\n",
    "\n",
    "\n",
    "#   name_errata_goi = name_errata[goi]\n",
    "\n",
    "#   og_list_goi = {}\n",
    "\n",
    "#   species_present = []\n",
    "\n",
    "#   #load fasta of all mitochondrially aligned genes and cycle through to capture sequence information of selected genes.  Also capture missing species\n",
    "#   for record in SeqIO.parse(mito_aln_dir + os.sep + goi + os.sep + goi + '.combined.translated.fasta', 'fasta'):\n",
    "#       spec_name = ''\n",
    "#       if len(record.id.split('.'))==3:\n",
    "#         sep = '.'\n",
    "#         spec_name = ' '.join(record.id.split(sep)[0:2])\n",
    "#         if spec_name == 'Komagataella phaffii':  # species as in shen et al 2018 in table 1 from supplement is changed.  \n",
    "#             spec_name = 'Komagataella pastoris'\n",
    "#       elif len(record.id.split('__'))>=3:\n",
    "#         sep = '__'\n",
    "#         spec_name = ' '.join(record.id.split(sep)[0:2])\n",
    "#       elif record.id in name_errata_goi['substitutions'].keys():\n",
    "#         spec_name = name_errata_goi['substitutions'][record.id]         \n",
    "      \n",
    "#       if spec_name in species_names:\n",
    "#           if spec_name == 'Candida albicans':\n",
    "#             gene_id = record.description.split('.')[2]\n",
    "#           elif len(record.description.split(' '))==1:\n",
    "#             gene_id = record.description\n",
    "#             print(record.description + 'does not have a name separated by a space.')\n",
    "#           else: \n",
    "#             gene_id = record.description.split(' ')[2]\n",
    "#           species_present.append(spec_name)\n",
    "#           og_list_goi[record.id] = {'species': spec_name,\n",
    "#                                     'gene_id': gene_id,\n",
    "#                                   'seq':str(record.seq)}\n",
    "\n",
    "#   missing_specs = species_names - set(species_present)\n",
    "#   print(missing_specs)\n",
    "#   name_errata_goi['missing_species']= missing_specs\n",
    "  \n",
    "#   og_list[goi] = og_list_goi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d092ebae-4fe4-48d3-848d-48cb0cdbc899",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'cox1': {'Ascoidea rubescens', 'Schizosaccharomyces pombe', 'Sporopachydermia lactativora', 'Alloascoidea hylecoeti'}, 'cox2': {'Pachysolen tannophilus', 'Ascoidea rubescens', 'Schizosaccharomyces pombe', 'Alloascoidea hylecoeti'}, 'cox3': {'Ascoidea rubescens', 'Schizosaccharomyces pombe', 'Alloascoidea hylecoeti'}, 'cob': {'Ascoidea rubescens', 'Schizosaccharomyces pombe', 'Alloascoidea hylecoeti'}, 'atp6': {'Ascoidea rubescens', 'Schizosaccharomyces pombe', 'Alloascoidea hylecoeti'}, 'atp8': {'Eremothecium gossypii', 'Ascoidea rubescens', 'Schizosaccharomyces pombe', 'Alloascoidea hylecoeti'}, 'atp9': {'Ascoidea rubescens', 'Schizosaccharomyces pombe', 'Kluyveromyces marxianus', 'Alloascoidea hylecoeti'}}\n"
     ]
    }
   ],
   "source": [
    "#rewritten version\n",
    "\n",
    "#Load selected species: \n",
    "species_set = pd.read_csv(os.path.normpath(base_dir + '/selected_proteins/species_selection/species_selection.csv'), index_col = 0)\n",
    "species_set_filt = species_set.loc[species_set['Load']=='Y']\n",
    "\n",
    "#mito_encoded_genes = os.listdir(mito_aln_dir)\n",
    "\n",
    "mito_genes_core_list = ['cox1', 'cox2', 'cox3', 'cob', 'atp6','atp8','atp9']\n",
    "\n",
    "#Pachysolen Tahhophilus has only a single underline after its name\n",
    "#Starmerella apicola is called Candida Apicola in wolters et al. \n",
    "\n",
    "name_errata = {goi:{'Pachysolen__tannophilus_'+goi:'Pachysolen tannophilus', 'Candida__apicola_'+goi:'Starmerella apicola'} for goi in mito_genes_core_list}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "missing_mito_genes = {}\n",
    "\n",
    "mito_genes_wolters = {}\n",
    "\n",
    "for goi in mito_genes_core_list: \n",
    "  #goi = 'atp6'\n",
    "\n",
    "  #load fasta of all mitochondrially aligned genes\n",
    "  species_names = set(species_set_filt.index)\n",
    "\n",
    "  mito_genes_wolters_goi = {}\n",
    "\n",
    "  species_present = []\n",
    "\n",
    "  for record in SeqIO.parse(mito_aln_dir + os.sep + goi + os.sep + goi + '.combined.translated.fasta', 'fasta'):\n",
    "      spec_name = ''\n",
    "      if len(record.id.split('.'))>=3:\n",
    "        sep = '.'\n",
    "        spec_name = ' '.join(record.id.split(sep)[0:2])\n",
    "      elif len(record.id.split('__'))>=3:\n",
    "        sep = '__'\n",
    "        spec_name = ' '.join(record.id.split(sep)[0:2])\n",
    "      elif record.id in name_errata[goi].keys():\n",
    "        spec_name = name_errata[goi][record.id]\n",
    "\n",
    "      # In Wolters et al, K. pastoris is called K. phaffii\n",
    "      if spec_name == 'Komagataella phaffii':\n",
    "        spec_name = 'Komagataella pastoris'\n",
    "      \n",
    "      if spec_name in species_names:        \n",
    "          species_present.append(spec_name)\n",
    "\n",
    "          if ((goi=='cox3')&(spec_name == 'Candida albicans')):\n",
    "            gene_id = record.description.split('.')[2]\n",
    "          elif ((goi =='atp9')&(spec_name=='Wickerhamomyces anomalus')):\n",
    "            gene_id = record.description.split('__')[2]\n",
    "          else: \n",
    "            gene_id = record.description.split(' ')[2]\n",
    "\n",
    "          mito_genes_wolters_goi[record.id] = {'species': spec_name,\n",
    "                                    'gene_id': gene_id,\n",
    "                                  'seq':str(record.seq)}\n",
    "\n",
    "  missing_specs = species_names - set(species_present)\n",
    "  missing_mito_genes[goi] = missing_specs\n",
    "  mito_genes_wolters[goi] = mito_genes_wolters_goi\n",
    "  \n",
    "  #print(missing_specs)\n",
    "print(missing_mito_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eacf2ff",
   "metadata": {},
   "source": [
    "Missing Species: \n",
    "1. S. pombe - extracted from Pombase\n",
    "\n",
    "\n",
    "2) Alloascoidea hylecoeti and Ascoidea rubescens\n",
    "\n",
    "From Antonis: \n",
    "\n",
    " we left these two species out because we could only find a few genes and only get very fragmented assemblies. For Alloascoidea, we could find 11 mt genes but the longest contig contained only 2 of them. For Ascoidea, we could only find 4 genes and the longest contig contained a single gene. Thus, it was hard to trust that we knew what was going on (in contrast to most other species, where we could retrieve a single contig that contained the entire genome).\n",
    "\n",
    "\n",
    "Search of ref genomes with tblastn for missing genes: \n",
    "\n",
    "These genes need to be translated using the ncbi translation table 4 which is the one used for mitochondria for the other three ascoidea species, and also for Sporopahydermia which are closest in the tree from that paper.   \n",
    "\n",
    "Searched, then saved hit as [spec_abbrev]_[gene_id]_[search original filename].xml\n",
    "\n",
    "Then got sequence for hit and saved as [spec_abbrev]_[gene_id]_cds.fasta\n",
    "\n",
    "Then imported and translated with Biopython.Seq.Translate\n",
    "\n",
    "A. hyl: https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_001600815.1/\n",
    "\n",
    "['cox1', 'cox2', 'cox3', 'cob', 'atp6','atp8','atp9']\n",
    "\n",
    "Cox1: NP_009305.1, Hit was spread across several scaffolds.   Left out\n",
    "Cox2: NP_009326.1, Single hit \n",
    "Cox3: NP_009328.1, Single hit\n",
    "Cob: NP_009315.1, Three hits somewhat overalapping across 3 scaffolds, left out\n",
    "Atp6: NP_009313.1, Single hit, only covered about 6 residues, left out\n",
    "Atp8: NP_009312.1, No hits\n",
    "Atp9: NP_009319.1, No hits\n",
    "\n",
    "\n",
    "A. rub: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_001661345.1/\n",
    "\n",
    "Cox1: NP_009305.1, Two nonoverlapping hits across two scaffolds, but big gaps. Left out.\n",
    "Cox2: NP_009326.1, No hits\n",
    "Cox3: NP_009328.1, No hits\n",
    "Cob: NP_009315.1, No hits\n",
    "Atp6: NP_009313.1, No Hits\n",
    "Atp8: NP_009312.1, No Hits\n",
    "Atp9: NP_009319.1, No Hits\n",
    "\n",
    "3) Other searches: \n",
    "\n",
    "Sporopahcydermia lactitavora for cox1 gene.  Antonis found Mitochondrial genome and found a good hit there by doing tblastn with  \n",
    "\n",
    "Query: \n",
    "Cox1: NP_009305.1 \n",
    "\n",
    "Subject: S. lactitavora mitochondrial genome\n",
    "gi|2618218503|gb|OR704033.1|\n",
    "https://www.ncbi.nlm.nih.gov/nuccore/OR704033.1?report=GenBank \n",
    "\n",
    "Translated with codon table 4\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Missing Genes found: \n",
    "\n",
    "These genes were placed into a fasta file called: \n",
    "\n",
    "examples/etc/mitochondrial_sequences/missing_sequences.fasta\n",
    "\n",
    "\n",
    ">Eremothecium__gossypii__atp8 atp8 uniprot__Q75G40\n",
    "https://www.uniprot.org/uniprotkb/Q75G40/entry\n",
    ">Pachysolen__tannophilus__cox2 cox2 uniprot__Q9XM29\n",
    "https://www.uniprot.org/uniprotkb/Q9XM29/entry\n",
    ">Pachysolen__tannophilus__cox2 cox2 uniprot__A0A1E4TMZ3 - this one is shorter, probably incorrect\n",
    "https://www.uniprot.org/uniprotkb/A0A1E4TMZ3/entry\n",
    ">Kluyveromyces marxianus__atp9 atp9 ncbi__AP012221.1\n",
    "\n",
    "\n",
    "Saved search: \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3e817ed4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "1\n",
      "-2\n",
      "Warning:  Overlap between Slac_cox1_4_cds.fasta and Slac_cox1_5_cds.fasta\n",
      "FFGHPEVYILIIPGFGVVSQIVSTYAKKPIFGRIGMIYAMASIGLLGFLVWSHHMYVVGLDADTRAYFTSATMIIAIPTGIKIWA\n",
      "LFSWIASLYGGSIVFTTPMLYALAFLFLFTIGGLSGVALSNSSLDIAFHDKIFIV\n",
      "first sequence overlap is IWA, second sequence overlap is LFS, using second sequence overlap\n",
      "-4\n",
      "Warning:  Overlap between Slac_cox1_5_cds.fasta and Slac_cox1_1_cds.fasta\n",
      "LFSWIASLYGGSIVFTTPMLYALAFLFLFTIGGLSGVALSNSSLDIAFHDKIFIV\n",
      "TYYVIAHFHYVLSMGAVFSLFAGYYYWSPKILGLYYNERLAQIQFWLLFIGANLTFFVMHFLGLQGMPRRIPDYPDAYYGWNYIASFGSLISVISVIVFIYVLYDQLVYGLDNKYNGKAVTILTTPDVSESNLVYLNNNSHRGTSLEWVLNSPPAFHSFNTPAIQ\n",
      "first sequence overlap is KIFIV, second sequence overlap is TYYVI, using second sequence overlap\n"
     ]
    }
   ],
   "source": [
    "#Import hits from A. hylecoeti and Sporopachydermia and translate using translation table 4\n",
    "\n",
    "retranslated_genes_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/retranslated.fasta')\n",
    "\n",
    "genes_to_translate = {'Alloascoidea__hylecoeti': {'coding_table': 4, \n",
    "                                                 'genes': \n",
    "                                                    {'cox2':  {1: {'hit_frame':-2, 'hit_start':86, 'hit_stop': 151}},\n",
    "                                                    'cox3': {1: {'hit_frame':-2, 'hit_start':6, 'hit_stop': 153}}\n",
    "                                                    }\n",
    "                                                },\n",
    "                      'Sporopahcydermia__lactitavora': {'coding_table': 4, \n",
    "                                                       'genes': \n",
    "                                                           {'cox1': \n",
    "                                                            {1: {'hit_frame':-1, 'hit_start':370, 'hit_stop': 533},\n",
    "                                                             2: {'hit_frame':-2, 'hit_start':4, 'hit_stop': 129},\n",
    "                                                             3: {'hit_frame':-2, 'hit_start':130, 'hit_stop': 237},\n",
    "                                                             4: {'hit_frame':-2, 'hit_start':238, 'hit_stop': 322}, \n",
    "                                                             5: {'hit_frame':-3, 'hit_start':320, 'hit_stop': 374}\n",
    "                                                             }\n",
    "                                                           }\n",
    "                                                        }\n",
    "                      }\n",
    "\n",
    "#Not sure what hit frame means in the context of the blast hits.  Using Expasy, only the 3'5' frame1 seemed to match the search. \n",
    "\n",
    "spec_abbrevs = {'Alloascoidea__hylecoeti': 'Ahyl', \n",
    "                'Sporopahcydermia__lactitavora': 'Slac'}\n",
    "\n",
    "spec_abbrev_rlookup = dict(zip(spec_abbrevs.values(),spec_abbrevs.keys()))\n",
    "\n",
    "cds_dir = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/cds')\n",
    "cds_fnames = os.listdir(cds_dir)\n",
    "\n",
    "cds_fnames_dict = {}\n",
    "for cds_fname in cds_fnames: \n",
    "    (spec_abbrev, goi, hit_number, *rest) = cds_fname.split('_')\n",
    "    species = spec_abbrev_rlookup[spec_abbrev]\n",
    "    goi_hit_data = genes_to_translate[species]['genes'][goi][int(hit_number)]\n",
    "    start = goi_hit_data['hit_start']\n",
    "    stop = goi_hit_data['hit_stop']\n",
    "    frame = goi_hit_data['hit_frame']\n",
    "\n",
    "    cds_fnames_dict[cds_fname] = (species,spec_abbrev, goi, start, stop,frame,int(hit_number))\n",
    "\n",
    "cds_fnames_df = pd.DataFrame.from_dict(cds_fnames_dict, orient='index', columns = ['species','spec_abbrev','goi','start','stop','frame','hit_number'])\n",
    "     \n",
    "with open(retranslated_genes_fname,'w') as f_out:\n",
    "    for species, genes_to_translate_data in genes_to_translate.items():\n",
    "\n",
    "        #species = 'Sporopahcydermia lactitavora' #'Alloascoidea hylecoeti'#\n",
    "        #genes_to_translate_data = genes_to_translate[species]\n",
    "\n",
    "        spec_abbrev = spec_abbrevs[species]\n",
    "        coding_table = genes_to_translate_data['coding_table']\n",
    "        gois = genes_to_translate_data['genes']\n",
    "\n",
    "        #Subset files based on spec_abbrev and goi\n",
    "\n",
    "        for goi in gois.keys():\n",
    "            cds_fnames_subset = cds_fnames_df[(cds_fnames_df['spec_abbrev']==spec_abbrev) & (cds_fnames_df['goi']==goi)]\n",
    "\n",
    "            if len(cds_fnames_subset)==1: #No need to merge different exons\n",
    "                cds_fname = cds_fnames_subset.index[0]\n",
    "\n",
    "                #For some reason I had to do the reverse complement here - I suspect it is somethinga bout the reading frame. \n",
    "                cds_full_seq_rc = str(SeqIO.read(cds_dir + os.sep + cds_fname,'fasta').seq)\n",
    "                cds_full_seq = Seq.reverse_complement(cds_full_seq_rc)\n",
    "\n",
    "                protein_seq = Seq.translate(cds_full_seq, table=4)\n",
    "\n",
    "                start=cds_fnames_subset['start'].iloc[0]\n",
    "                stop= cds_fnames_subset['stop'].iloc[0]\n",
    "                \n",
    "            else: \n",
    "                protein_seq_list = []\n",
    "                cds_fnames_subset_sorted = cds_fnames_subset.sort_values(by='start')\n",
    "                start = cds_fnames_subset_sorted['start'].iloc[0]\n",
    "                current_stop = cds_fnames_subset_sorted['stop'].iloc[0]\n",
    "                \n",
    "                current_fname = cds_fnames_subset_sorted.index[0]\n",
    "                current_cds_rc = str(SeqIO.read(cds_dir + os.sep + current_fname,'fasta').seq)\n",
    "                #For some reason I had to do the reverse complement here - I suspect it is something about the reading frame - they are all negative, but the value didn't seem to matter. \n",
    "                current_cds = Seq.reverse_complement(current_cds_rc)\n",
    "                current_protein_seq = Seq.translate(current_cds, table=coding_table)\n",
    "                protein_seq_list.append(current_protein_seq)\n",
    "\n",
    "                # #check current protein_seq is correct length\n",
    "                # assert current_stop-start==len(current_protein_seq), 'Protein Sequence not correct length fname = {}'.format(current_fname) \n",
    "\n",
    "                for jj in range(1,len(cds_fnames_subset_sorted)):\n",
    "                    next_fname = cds_fnames_subset_sorted.index[jj]\n",
    "                    next_start = cds_fnames_subset_sorted['start'].iloc[jj]\n",
    "                    next_stop = cds_fnames_subset_sorted['stop'].iloc[jj]\n",
    "\n",
    "                    next_cds_rc = str(SeqIO.read(cds_dir + os.sep + next_fname,'fasta').seq)\n",
    "                    next_cds = Seq.reverse_complement(next_cds_rc)\n",
    "                    next_protein_seq = Seq.translate(next_cds, table=coding_table)\n",
    "                    # assert next_stop-next_start==len(next_protein_seq), 'Protein Sequence not correct length fname = {}'.format(next_fname) \n",
    "\n",
    "                    gap = next_start-current_stop     \n",
    "                    print(gap)   \n",
    "                    if gap == 1: \n",
    "                        protein_seq_list.append(next_protein_seq)\n",
    "                    elif gap>1: \n",
    "                        print('Warning:  Gap between {} and {}, merging sequence ignoring gap'.format(current_fname, next_fname) )\n",
    "                        protein_seq_list.append(next_protein_seq)\n",
    "                    elif gap<1:\n",
    "                        print('Warning:  Overlap between {} and {}'.format(current_fname, next_fname) )\n",
    "                        print(current_protein_seq)\n",
    "                        overlap_current = current_protein_seq[gap-1:]\n",
    "                        print(next_protein_seq)\n",
    "                        overlap_next = next_protein_seq[0:-gap+1]\n",
    "                        print('first sequence overlap is {}, second sequence overlap is {}, using second sequence overlap'.format(overlap_current, overlap_next))\n",
    "                        protein_seq_list[jj-1]=protein_seq_list[jj-1][0:gap-1]\n",
    "                        protein_seq_list.append(next_protein_seq)\n",
    "\n",
    "                    current_start = next_start\n",
    "                    current_stop = next_stop\n",
    "                    current_fname = next_fname\n",
    "                    current_protein_seq = next_protein_seq\n",
    "                \n",
    "                protein_seq = ''.join(protein_seq_list)\n",
    "\n",
    "                stop = current_stop\n",
    "            \n",
    "            f_out.write('>' + species + '__' + goi + ' ' + goi + ' retranslated__start_' + str(start) + '__stop_' + str(stop) + '\\n')\n",
    "            f_out.write(protein_seq + '*\\n')\n",
    "\n",
    "\n",
    "\n",
    "# # coding_dna = \n",
    "# # \n",
    "# # print(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "438e4057",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract peptide sequences from wolters et al\n",
    "\n",
    "mito_genes_wolters_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/from_wolters.fasta')\n",
    "\n",
    "with open(mito_genes_wolters_fname, 'w', newline='\\n') as f_out:\n",
    "\n",
    "    for goi, og_list_goi in og_list.items():\n",
    "        \n",
    "        for wolters_id, gene_info in og_list_goi.items():\n",
    "            spec = gene_info['species']\n",
    "            f_out.write('>' + spec.split(' ')[0] + '__' + spec.split(' ')[1] + '__' +goi + ' ' + goi + ' wolters__' + gene_info['gene_id'] + '\\n')\n",
    "            f_out.write(gene_info['seq'] + '\\n')\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a24ce7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract proteins from Pombe genome\n",
    "pombe_genename_map_rev = {mito_goi : mito_goi for mito_goi in mito_gois}\n",
    "pombe_genename_map_rev['cob'] = 'cob1'\n",
    "pombe_genename_map = {genename_pombe:genename for genename, genename_pombe in pombe_genename_map_rev.items()}\n",
    "\n",
    "pombe_proteome_fname = genomes_dir + os.path.normpath('schizosaccharomyces_pombe/peptide_20240918.fa')\n",
    "\n",
    "pombe_mito_genes_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/pombe_core_genes.fasta')\n",
    "\n",
    "with open(pombe_mito_genes_fname, 'w', newline='\\n') as f_out: \n",
    "    pombe_prots = SeqIO.parse(pombe_proteome_fname, 'fasta')\n",
    "    for record in pombe_prots:\n",
    "        pombe_genename = record.description.split(' ')[1].split('|')[0]\n",
    "        if pombe_genename in set(pombe_genename_map.keys()):\n",
    "            genename = pombe_genename_map[pombe_genename] \n",
    "            pombase_id = record.id.split(':')[0]\n",
    "            f_out.write('>Schizosaccharomyces__pombe__' +genename + ' ' + genename + ' pombase__' + pombe_genename + '_' + pombase_id + '\\n')\n",
    "            f_out.write(str(record.seq) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "77987a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mito_genes_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/mito_core_genes.fasta')\n",
    "\n",
    "pombe_mito_genes_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/pombe_core_genes.fasta')\n",
    "mito_genes_wolters_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/from_wolters.fasta')\n",
    "missing_genes_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/missing_sequences.fasta')\n",
    "retranslated_genes_fname = base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/retranslated.fasta')\n",
    "\n",
    "\n",
    "with open(mito_genes_fname,'w') as f_out:\n",
    "    for fname in [pombe_mito_genes_fname, mito_genes_wolters_fname, missing_genes_fname,retranslated_genes_fname]:\n",
    "        with open(fname,'r') as f_in:\n",
    "           shutil.copyfileobj(f_in, f_out)\n",
    "        #f_out.write('\\n')\n",
    "\n",
    "\n",
    "# with open(newfile,'wb') as destination:\n",
    "#     with open(file_1,'rb') as source:\n",
    "#         shutil.copyfileobj(source, destination)\n",
    "#     destination.write(os.linesep*2)\n",
    "#     with open(file_2,'rb') as source:\n",
    "#         shutil.copyfileobj(source, destination)\n",
    "#     destination.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dd1c38d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;31mSignature:\u001b[0m \u001b[0mshutil\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcopyfileobj\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfsrc\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfdst\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlength\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mDocstring:\u001b[0m copy data from file-like object fsrc to file-like object fdst\n",
      "\u001b[1;31mFile:\u001b[0m      c:\\users\\bheineike\\anaconda3\\envs\\bmh_bioinformatics\\lib\\shutil.py\n",
      "\u001b[1;31mType:\u001b[0m      function"
     ]
    }
   ],
   "source": [
    "shutil.copyfileobj?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5e606c1",
   "metadata": {},
   "source": [
    "## Move mitochondrial PDB files into appropriate folders, then build list of filenames to run USalign\n",
    "usalign: \n",
    "Note: Sandra didn't make this sequence >Alloascoidea__hylecoeti__cox2 cox2 retranslated__start_86__stop_151\n",
    "\n",
    "#Kluyveromyces__marxianus__atp9 is far too long and doesn't align well.  Remove from list. \n",
    "Lipomyces__starkeyi__cox2.pdb looks bad but it is from the Wolters data Lipomyces__starkeyi__cox2 cox2 wolters__gene-94-mRNA-1 will leave it in\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "65817083",
   "metadata": {},
   "outputs": [],
   "source": [
    "mito_genes_core_list = ['cox1', 'cox2', 'cox3', 'cob', 'atp6','atp8','atp9']\n",
    "\n",
    "pdb_base_dir =  base_dir + os.sep + os.path.normpath('examples/etc/mitochondrial_sequences/pdbs')\n",
    "\n",
    "seq_filter = {'atp9': ['Kluyveromyces__marxianus__atp9'], 'cox2': ['Lipomyces__starkeyi__cox2']}\n",
    "\n",
    "mito_pdbs = os.listdir(pdb_base_dir + os.sep + 'pdbs')\n",
    "\n",
    "mito_pdb_lists = {mito_gene:[] for mito_gene in mito_genes_core_list }\n",
    "\n",
    "for mito_pdb in mito_pdbs:\n",
    "    mito_pdb_base = mito_pdb.split('.')[0]\n",
    "    mito_gene = mito_pdb_base.split('__')[2]\n",
    "    mito_pdb_lists[mito_gene].append(mito_pdb_base)\n",
    "\n",
    "for mito_gene, drop_list in seq_filter.items(): \n",
    "    old_list = mito_pdb_lists[mito_gene]\n",
    "    filtered_list = [item for item in old_list if not(item in drop_list)]\n",
    "    mito_pdb_lists[mito_gene] = filtered_list\n",
    "\n",
    "\n",
    "\n",
    "for mito_gene in mito_genes_core_list: \n",
    "    pdb_list_fname = pdb_base_dir + os.sep + 'pdb_list_' + mito_gene + '.txt'\n",
    "\n",
    "    with open(pdb_list_fname,'w') as f_out: \n",
    "        for pdb_fname in mito_pdb_lists[mito_gene]: \n",
    "            f_out.write(pdb_fname + '\\n') \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f0a5823",
   "metadata": {},
   "source": [
    "use 20241023_usalign_mito_gene.sh which runs \n",
    "\n",
    "<code>\n",
    "/home/heineike/usalign/USalign -dir $PDB_DIR $PDB_LIST -suffix .pdb -mm 4 > ${MITO_PDB_BASE}/us_align_${MITO_CORE_GENE}.fasta\n",
    "</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "42ab44e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove headers from usalign files\n",
    "for mito_gene in mito_genes_core_list: \n",
    "    us_align_fname = pdb_base_dir + os.sep + 'us_align_' + mito_gene + '.fasta'\n",
    "    us_align_clean_fname = pdb_base_dir + os.sep + 'fasta_renamed' + os.sep + 'us_align_clean_' + mito_gene + '.fasta'\n",
    "    with open(us_align_fname,'r') as f_in: \n",
    "        with open(us_align_clean_fname,'w') as f_out: \n",
    "            for line in f_in: \n",
    "                if line[0]=='>':\n",
    "                    f_out.write(line)\n",
    "                    seq_line = next(f_in)\n",
    "                    f_out.write(seq_line)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "670022de",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Trees of protein sequences\n",
    "\n",
    "#Make files for Oliver to order sequence names into trees and visualize them\n",
    "tree_dir = base_dir + os.sep + os.path.normpath('/examples/etc/mitochondrial_sequences/pdbs/trees/mito_trees')\n",
    "\n",
    "for tree_fname in os.listdir(tree_dir):\n",
    "    if (tree_fname.split('.')[-1])=='treefile':\n",
    "        mito_gene = tree_fname.split('.')[0]\n",
    "\n",
    "        #tree_fname = base_dir + os.sep + os.path.normpath('msas/structural/tm_align/trees/OG1022_REF_Scer_AF-P38715-F1-model_v2.tm.fasta.clipkit.treefile')\n",
    "        t = Tree(tree_dir + os.sep + tree_fname, format=1)\n",
    "\n",
    "        t.ladderize\n",
    "\n",
    "        fname_out = tree_dir  + os.sep + mito_gene + '_tree_list.txt'\n",
    "\n",
    "        with open(fname_out,'w') as f_out:\n",
    "            for leaf in t.iter_leaf_names():\n",
    "                leaf_tidy = leaf.split('.pdb_A')[0] + '.pdb'\n",
    "                f_out.write(leaf_tidy + '\\n')\n",
    "\n",
    "        image_out_fname = tree_dir  + os.sep + mito_gene + '_tree.png'\n",
    "        t.render(file_name=image_out_fname)\n",
    "\n",
    "    #t.render('%%inline')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bmh_bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
