# diverse_yeast
 scripts for BMH's contributions to the Ralser Lab Yeast Structural Evolution Project.  Includes codes for 
   - Assembling orthogroup sequences
   - Making sequence based alignments
   - Refining structural alignment to prepare for DN/DS calculations
   - DN/DS Calculations

Core functions are contained in diverse yeast tools.py

Also uses a few functions from https://github.com/heineike02/y1000plus_tools
  make_og_genes_lookup
  extract_protein_seqs

Requires following directories: 
base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts/alphafold')

y1000plus_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools/data') + os.sep

genomes_dir:  Includes genomic information from model organisms. 


# Key Scripts: 

## Ipython Scripts; 
All these scripts run on bmh_bioinformatics.yaml anaconda environment in Windows 11, except for Sce_Flux_C13_bmh_edits.ipynb which runs on the cobra.yaml environment
### 20211025_alphafold_selection.ipynb
select orthogroups and orthologs based on selected species and annotations from yeast pathway database.  Builds list of proteins of interest based on orthogroups from orthomcl clusters

Visualize species trees for various figures. 

Assemble orthogroups and annotate metadata for each orthogroup in base_dir/selected_proteins/og_metadata.json.  Build a single fasta for all peptides in base_dir/selected_proteins/all_proteins.fasta. 

Quantify proteins that were calculated. 

### 20220719_alphafold_PLDDT_scores.ipynb
Assesses PLDDT scores of calculated structure and analyses structures that could not be calculated. 

### 20221012_assemble_peptide_cds_fastas.ipynb
Makes selected_protein_ids.json to list proteins we consider for each species. 
Builds list of coding sequences for all proteins we consider for each species. 
Loads tm_align clustering data and assembles and assesses subclusters. 
Filters sequences for dN/dS analysis based on cluster size and sequence length.  

### 20220921_map_proteomes_to_shen.ipynb
Uses Pairwise sequence similarity scores to map uniprot ids to shen ids in two groups - first proteomes that were part of the proteomics experiment, then other proteomes with uniprot ids.  

### 20240220_alignment_example.ipynb
visualizes phylogenetic tree for alignment example of OG3416_P40099.

### 20230314_struct_seq_alignment_comparisons.ipynb
Compares Tree and Alignment properties between sequence-based and structural alignments. 

### 20230201_ipath_plots.ipynb
prepares data for visualization in ipath

### 20230724_phenotype_selection.ipynb
Filters phenotype data from lu et al for further analysis.  Also displays phenotypes used during our analysis. 

### 20221206_selection_calculations.ipynb
Extracts DN/DS data from raw codeml output and analyses data. 

Identifies branches to investigate with Branch/Site Models, and extracts residues under positive selection from Branch/Site calculations.  

Analysis of ETC, POX1, HSU1, TDH3 and TPI examples. 

### 20230913_selection_calculations_breakdown_by_features.ipynb
subsets alignments by features (binding_site, uniprot_binding_site, core, surface) and analyzed m0 data from those. 

Also uses features to parse site model data.

### 20240312_thiamine_pathway.ipynb
Analysis of thiamine pathway example

### 20240801_mitochondrial_alignments.ipynb
Prepares alignments and trees for missing mitochondrially encoded genes. 

### 20241120_domain_mapping.ipynb
Analysis of TED domains for our proteins of interest.  Also analysis of HSU1 example. 

### 20241210_xylose_pathway_analysis.ipynb
Analysis of xylose pathway

### 20250122_etc_evcoupling.ipynb
Initial efforts to run evolutionary coupling analysis.

### 20250206_ec_regression.ipynb
linear regression to check whether specific groups of enzymes are changing more than expected based on phenotype separation. 

### Sce_Flux_C13_bmh_edits.ipynb
Feiran Li's analysis of C13 flux measurements in comparison with model predictions.  Uses cobra environment. 

## Shell and Python Scripts
These were run within a singularity shell environment created using codeml_singularity.cfg.

The environment, codeml.sif is saved at
codeml.sif library://heineike02/codeml/codeml:0.6

### call_code_singularity.sh
Simple shell script to call other shell scripts using the singularity shell.  Maps key directories prior to execution. 

### 20221206_struct_align_dnds_msas.sh/.py
Make Strictly trimmed CDS for DN/DS calculation based on tm_align msas

### 20221206_struct_align_dnds_trees.sh/.py
Make trees compatible with paml from structural MSAs generated by TMalign

### Alignment and tree preparation for special cases

#### Mitochondria
20241023_struct_align_dnds_msas_mito.sh/.py
20241023_struct_align_dnds_trees_mito.sh/.py
Mitochondrially encoded proteins had curated orthogroup after initial run. 

#### Structural features: 
20230913_struct_align_dnds_msas_binding_site.sh/.py
20240109_struct_align_dnds_msas_uniprot_binding_site.sh/.py
20240916_struct_align_dnds_msas_core_surface.sh/.py

#### Trimming sequences
20230301_struct_align_dnds_msas_yn00_filter.sh/.py
20230307_struct_align_dnds_msas_m0_filter.sh/.py
Trimmed short sequences to troubleshoot yn00 and m0.  For m0 needed to trim tree as well. 

#### Sequence Based alignments
20230324_seq_align_dnds_trees.sh/.py
20230326_calculate_tree_quantities.sh/.py

#### Tree for example orthogroup 
20240215_full_og_tree.sh/.py
Make trimmed alignments in order to display a phylogenetic tree for an orthogroup before it was split into subgroups.  

### 20230110_seq_align.sh/.py
Makes MAFFT and Clustalo alignments corresponding to structural alignments. 

### 20230326_calculate_tree_quantities.sh/.py
Uses phykit to calculate tree based quantities from trees. 

### US align scripts
Most US-align calls were done by Nir, but these were done separately
20240215_usalign_example_og.sh
20241023_usalign_mito_genes.sh
20250108_usalign_hsu1_str2.sh

### DN/DS

#### YN00
20221207_run_yn00.sh
20230301_run_yn00_test_cleandata.sh
20230301_run_yn00_troubleshoot.sh

#### m0
20221214_run_m0.sh
20230914_run_m0_subset.sh
20221220_test_m0.sh
20230926_run_m0_v2.sh
20230309_run_m0_troubleshoot.sh

#### Site Model
20231010_run_site_model.sh
20241025_site_model_check_run.py
20230926_call_hyphy_busted.sh : alternative to codeML site model

#### Branch Site
20230119_run_m1.sh.  Model m1 to use as filter to identify branches to run Branch Site Model on
20221214_run_BS.sh.  Branch Site Model

### 20250122_evcoupling.sh/.py
Run evolutionary coupling analysis for select proteins. 

# Proteome Annotation

Downloaded 'uniprot-proteome_UP000002311.tab' on 20220120 from uniprot after adding additional name fields to include the systematic name (Gene names  (ordered locus ))
