{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "#from Bio import SeqIO\n",
    "from Bio import AlignIO   #, Align\n",
    "\n",
    "base_dir = os.path.normpath('G:/My Drive/Crick_LMS/projects/diverse_yeasts/alphafold')\n",
    "divyeast_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/diverse_yeast')\n",
    "y1000plus_dir = os.path.normpath('C:/Users/heineib/Documents/GitHub/y1000plus_tools/data') + os.sep\n",
    "\n",
    "\n",
    "#C:\\Users\\heineib\\Documents\\GitHub\\y1000plus_tools\\data\\y1000plus_tools_data\\y1000plus\\id_lookups\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Goal:  Go from filename from Oliver's script to what would be a standard filename for any generated structure.  That will then be converted to Jacob's.  \n",
    "# While doing so remove from both the sequence and the protein alignment proteins that have been flagged by oliver.  \n",
    "\n",
    "#Structural alignment fasta headers\n",
    "# <orig_species_name>__<og>__<y1000_id>\n",
    "\n",
    "# Except for C.alb, S.cer and S. pombe which have\n",
    "# <species_abbrev>_AF-<swissprot_id>-F1-model_V2\n",
    "#\n",
    "# Unless it is the reference protein in which case it has: \n",
    "\n",
    "#REF_<species_abbrev>_AF-<swissprot_id>-F1-model_V2\n",
    "\n",
    "#Sequence alignment fasta headers\n",
    "\n",
    "#<old_species_name>|<y1000_id>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#species_lookup - y1000 species table\n",
    "y1000_species = pd.read_csv(y1000plus_dir + os.path.normpath(\"y1000plus_tools_data/y1000plus/y1000_species_table.csv\"), index_col=0)\n",
    "\n",
    "#fig2_orig_genome_lookup = dict(zip(y1000_species['species_names_fig2'], y1000_species['original_genome_id']))\n",
    "#old_orig_genome_lookup = dict(zip(y1000_species['old_species_names'], y1000_species['original_genome_id']))\n",
    "orig_old_genome_lookup = dict(zip(y1000_species['original_genome_id'], y1000_species['old_species_names']))\n",
    "\n",
    "\n",
    "spec_abbrev_dict = {'Scer': 'Saccharomyces_cerevisiae', \n",
    "                    'Calb': 'Candida_albicans', \n",
    "                    'Spom': 'Schizosaccharomyces_pombe'}\n",
    "\n",
    "\n",
    "#Lookup from gene_id to y1000_id\n",
    "\n",
    "gene_id_2_y1000_id = {}\n",
    "\n",
    "#Load S.cer lookup table: \n",
    "scer_lookup_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/saccharomyces_cerevisiae.csv')\n",
    "scer_lookup = pd.read_csv(scer_lookup_fname, index_col=0)\n",
    "gene_id_2_y1000_id['Scer'] = dict(zip(scer_lookup.index,scer_lookup['y1000_id']))\n",
    "\n",
    "#Load C.alb lookup table\n",
    "calb_lookup_fname = y1000plus_dir + os.path.normpath('y1000plus_tools_data/y1000plus/id_lookups/candida_albicans.csv')\n",
    "calb_lookup = pd.read_csv(calb_lookup_fname, index_col=0)\n",
    "gene_id_2_y1000_id['Calb'] = dict(zip(calb_lookup.index,calb_lookup['y1000_id']))\n",
    "\n",
    "#Lookup from swissprot_id to gene_id\n",
    "swissprot_id_2_gene_id= {}\n",
    "scer_swissprot_id_2_gene_id_df = pd.read_table(base_dir + os.sep + os.path.normpath('msas/structural/Scer_protein_names.tsv'))\n",
    "swissprot_id_2_gene_id['Scer'] = dict(zip(scer_swissprot_id_2_gene_id_df['Swiss-Prot'],scer_swissprot_id_2_gene_id_df['OLN']))\n",
    "\n",
    "calb_swissprot_id_2_gene_id = pickle.load(open(base_dir + os.sep + os.path.normpath('msas/structural/Mapping_calb.pkl'),\"rb\"))\n",
    "swissprot_id_2_gene_id['Calb'] = dict(zip(calb_swissprot_id_2_gene_id.values(),calb_swissprot_id_2_gene_id.keys()))\n",
    "\n",
    "\n",
    "#Load analysis file which identifies bad structures\n",
    "struct_analysis = pickle.load(open(base_dir + os.sep + os.path.normpath('Output/data/Analysis_new_02.pkl'), 'rb'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in OG: OG1054 , REF: Scer_AF-P16861-F1-model_v2, REF_Scer_AF-P16861-F1-model_v2 converts to Saccharomyces_cerevisiae|110_2957 and is not in the sequence ids\n",
      "in OG: OG1054 , REF: Scer_AF-P16862-F1-model_v2, REF_Scer_AF-P16861-F1-model_v2 converts to Saccharomyces_cerevisiae|110_2957 and is not in the sequence ids\n",
      "in OG: OG1254 , REF: Scer_AF-P40395-F1-model_v2, Calb_AF-A0A1D8PCB4-F1-model_v2 converts to Candida_albicans|12_4978 and is not in the sequence ids\n",
      "in OG: OG1254 , REF: Scer_AF-Q01519-F1-model_v2, Calb_AF-A0A1D8PCB4-F1-model_v2 converts to Candida_albicans|12_4978 and is not in the sequence ids\n",
      "in OG: OG1355 , REF: Scer_AF-P10614-F1-model_v2, REF_Scer_AF-P10614-F1-model_v2 converts to Saccharomyces_cerevisiae|110_3481 and is not in the sequence ids\n",
      "in OG: OG1468 , REF: Scer_AF-P17898-F1-model_v2, REF_Scer_AF-P17898-F1-model_v2 converts to Saccharomyces_cerevisiae|110_204 and is not in the sequence ids\n",
      "in OG: OG1468 , REF: Scer_AF-P22140-F1-model_v2, REF_Scer_AF-P17898-F1-model_v2 converts to Saccharomyces_cerevisiae|110_204 and is not in the sequence ids\n",
      "in OG: OG1498 , REF: Scer_AF-Q06702-F1-model_v2, komagataella_pastoris__OG1498__52_4581 converts to Komagataella_pastoris|52_4581 and is not in the sequence ids\n",
      "in OG: OG1498 , REF: Scer_AF-Q06703-F1-model_v2, komagataella_pastoris__OG1498__52_4581 converts to Komagataella_pastoris|52_4581 and is not in the sequence ids\n",
      "in OG: OG1565 , REF: Scer_AF-P38604-F1-model_v2, REF_Scer_AF-P38604-F1-model_v2 converts to Saccharomyces_cerevisiae|110_4016 and is not in the sequence ids\n",
      "in OG: OG1630 , REF: Scer_AF-P50264-F1-model_v2, Calb_AF-Q5AMQ8-F1-model_v2 converts to Candida_albicans|12_3302 and is not in the sequence ids\n",
      "in OG: OG1696 , REF: Scer_AF-P53598-F1-model_v2, REF_Scer_AF-P53598-F1-model_v2 converts to Saccharomyces_cerevisiae|110_2875 and is not in the sequence ids\n",
      "in OG: OG1746 , REF: Scer_AF-P32642-F1-model_v2, REF_Scer_AF-Q03835-F1-model_v2 converts to Saccharomyces_cerevisiae|110_5840 and is not in the sequence ids\n",
      "in OG: OG1746 , REF: Scer_AF-Q03835-F1-model_v2, REF_Scer_AF-Q03835-F1-model_v2 converts to Saccharomyces_cerevisiae|110_5840 and is not in the sequence ids\n",
      "in OG: OG2038 , REF: Scer_AF-P12709-F1-model_v2, REF_Scer_AF-P12709-F1-model_v2 converts to Saccharomyces_cerevisiae|110_5695 and is not in the sequence ids\n",
      "in OG: OG2912 , REF: Scer_AF-P21182-F1-model_v2, REF_Scer_AF-P21182-F1-model_v2 converts to Saccharomyces_cerevisiae|110_4155 and is not in the sequence ids\n",
      "in OG: OG3006 , REF: Scer_AF-P08465-F1-model_v2, REF_Scer_AF-P08465-F1-model_v2 converts to Saccharomyces_cerevisiae|110_5544 and is not in the sequence ids\n",
      "in OG: OG3063 , REF: Scer_AF-P36010-F1-model_v2, lipomyces_starkeyi__OG3063__66_2744 converts to Lipomyces_starkeyi|66_2744 and is not in the sequence ids\n",
      "in OG: OG3909 , REF: Scer_AF-P40319-F1-model_v2, Calb_AF-A0A1D8PGU2-F1-model_v2 converts to Candida_albicans|12_2342 and is not in the sequence ids\n",
      "in OG: OG4185 , REF: Scer_AF-P0CX77-F1-model_v2, REF_Scer_AF-P0CX77-F1-model_v2 converts to Saccharomyces_cerevisiae|110_4578 and is not in the sequence ids\n",
      "in OG: OG4185 , REF: Scer_AF-P0CX78-F1-model_v2, REF_Scer_AF-P0CX77-F1-model_v2 converts to Saccharomyces_cerevisiae|110_4578 and is not in the sequence ids\n",
      "in OG: OG4185 , REF: Scer_AF-P0CX79-F1-model_v2, REF_Scer_AF-P0CX77-F1-model_v2 converts to Saccharomyces_cerevisiae|110_4578 and is not in the sequence ids\n",
      "in OG: OG4185 , REF: Scer_AF-P0CZ17-F1-model_v2, REF_Scer_AF-P0CX77-F1-model_v2 converts to Saccharomyces_cerevisiae|110_4578 and is not in the sequence ids\n",
      "in OG: OG4185 , REF: Scer_AF-P38986-F1-model_v2, REF_Scer_AF-P0CX77-F1-model_v2 converts to Saccharomyces_cerevisiae|110_4578 and is not in the sequence ids\n",
      "in OG: OG4751 , REF: Scer_AF-P04039-F1-model_v2, Calb_AF-A0A1D8PHJ3-F1-model_v2 converts to Candida_albicans|12_6080 and is not in the sequence ids\n"
     ]
    }
   ],
   "source": [
    "#convert_qc = {}\n",
    "\n",
    "keep_in_filt_seq = True #flag to skip filtered seqs that don't show up in orthogroups\n",
    "\n",
    "for fasta_fname in os.listdir(base_dir + os.sep +  os.path.normpath('msas\\structural\\FASTA\\\\') ):\n",
    "#fasta_fname = 'OG1004_REF_Scer_AF-P15938-F1-model_v2.FASTA'\n",
    "\n",
    "    convert_qc_og_ref = {}\n",
    "    og = fasta_fname.split('_')[0]\n",
    "    ref = fasta_fname.split('_')[2] + '_' + fasta_fname.split('_')[3] + '_' + fasta_fname.split('_')[4].split('.')[0]\n",
    "\n",
    "    seqs_to_remove = struct_analysis[fasta_fname.split('.')[0]]['Files to be excluded']\n",
    "    \n",
    "    fname_struct_aln_orig = base_dir + os.sep +  os.path.normpath('msas\\structural\\FASTA\\\\' + fasta_fname)   # os.path.normpath('msas\\FILES_ogs_pep_aligned\\\\' + og + '.mfaa.mafft')\n",
    "    fname_seq_aln_orig = base_dir + os.sep + os.path.normpath('msas\\FILES_ogs_pep_aligned\\\\' + og + '.mfaa.mafft')\n",
    "\n",
    "    fname_struct_aln_renamed_out = base_dir + os.sep + os.path.normpath('msas\\structural\\\\fasta_renamed\\\\' + og + '__' + ref + '.struct.fasta')\n",
    "    fname_struct_aln_filt_out = base_dir + os.sep + os.path.normpath('msas\\structural\\\\fasta_renamed_filt\\\\' + og + '__' + ref + '.struct_filt.fasta')\n",
    "    fname_seq_aln_filt_out =  base_dir + os.sep + os.path.normpath('msas\\structural\\\\fasta_renamed_seq_filt\\\\' + og + '__' + ref + '.seq_filt.fasta')\n",
    "\n",
    "    #Read in Sequence File, extract all names make dict of seq_alignment sequences\n",
    "    seq_align = AlignIO.read(fname_seq_aln_orig, 'fasta')\n",
    "    seq_align_names = [record.id for record in seq_align]\n",
    "    seq_align_seqs = {record.id: record.seq for record in seq_align}\n",
    "    \n",
    "    #Read in dictionary of ref name map to sequence name\n",
    "    struct_align = AlignIO.read(fname_struct_aln_orig, 'fasta')\n",
    "    struct_names_converted = []\n",
    "    with open(fname_struct_aln_renamed_out, 'w') as f_out: \n",
    "        with open(fname_struct_aln_filt_out , 'w') as f_out_filt:\n",
    "            with open(fname_seq_aln_filt_out, 'w') as f_out_seq_filt: \n",
    "                for record in struct_align: \n",
    "                    skip = False    #Right now only skipping Spom sequences since those are not in sequence alignment\n",
    "                    name_orig = record.id    \n",
    "\n",
    "                    #Check if name in ref sequence (S.cer, C.alb, and S. pom) + convert\n",
    "\n",
    "                    if name_orig.split('_')[0] in set(['Scer', 'REF', 'Calb', 'Spom']):\n",
    "                        if name_orig.split('_')[0] == 'REF': \n",
    "                            spec_abbrev = name_orig.split('_')[1]\n",
    "                            swissprot_id = name_orig.split('_')[2].split('-')[1]\n",
    "                        else: \n",
    "                            spec_abbrev = name_orig.split('_')[0]\n",
    "                            swissprot_id = name_orig.split('_')[1].split('-')[1]\n",
    "\n",
    "                        if spec_abbrev == 'Spom':\n",
    "                            #S pombe not currently included in sequence alignments\n",
    "                            skip = True\n",
    "                        else: \n",
    "                            spec_old = spec_abbrev_dict[spec_abbrev]\n",
    "\n",
    "                            y1000_id = gene_id_2_y1000_id[spec_abbrev][swissprot_id_2_gene_id[spec_abbrev][swissprot_id]]\n",
    "\n",
    "\n",
    "                    else: \n",
    "                        #Convert name            \n",
    "                        (spec_orig, og, y1000_id) = name_orig.split('__')\n",
    "                        spec_old = orig_old_genome_lookup[spec_orig]\n",
    "\n",
    "                    if not(skip):\n",
    "\n",
    "                        seq_aln_id = spec_old + '|' + y1000_id \n",
    "\n",
    "                        if not(seq_aln_id in seq_align_names): \n",
    "                            keep_in_filt_seq = False\n",
    "                            print('in OG: ' + og + ' , REF: ' + ref + ', ' + name_orig + ' converts to ' + seq_aln_id + ' and is not in the sequence ids')\n",
    "\n",
    "                        #Print out fasta without filter\n",
    "                        struct_names_converted.append(seq_aln_id)\n",
    "                        f_out.write('>' + seq_aln_id + '\\n')\n",
    "                        f_out.write(str(record.seq) + '\\n')\n",
    "\n",
    "                        #Filter out seqs with bad alignments\n",
    "                                                \n",
    "                        if not(name_orig in seqs_to_remove): \n",
    "                            f_out_filt.write('>' + seq_aln_id + '\\n')\n",
    "                            f_out_filt.write(str(record.seq) + '\\n')\n",
    "                            if keep_in_filt_seq: \n",
    "                                f_out_seq_filt.write('>' + seq_aln_id + '\\n')\n",
    "                                f_out_seq_filt.write(str(seq_align_seqs[seq_aln_id]) + '\\n')\n",
    "                            \n",
    "                        keep_in_filt_seq = True\n",
    "                        \n",
    "\n",
    "\n",
    "\n",
    "    #Double check lists of names\n",
    "\n",
    "    #seq_not_struct = list(set(seq_align_names)-set(struct_names_converted))\n",
    "    #convert_qc_og_ref['seq_not_struct'] = seq_not_struct\n",
    "\n",
    "    #convert_qc[og_ref] = convert_qc_og_ref\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open(base_dir + os.sep + 'og_metadata.json', 'r') as f:\n",
    "    og_metadata = json.load(f) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'N_genes': 50,\n",
       " 'pct_spec': 1.0,\n",
       " 'og_genes': ['0_3372',\n",
       "  '0_7632',\n",
       "  '0_842',\n",
       "  '104_1425',\n",
       "  '104_4147',\n",
       "  '106_3026',\n",
       "  '106_4589',\n",
       "  '110_2957',\n",
       "  '110_4171',\n",
       "  '12_1254',\n",
       "  '12_5593',\n",
       "  '134_532',\n",
       "  '135_1844',\n",
       "  '135_604',\n",
       "  '136_4557',\n",
       "  '136_4679',\n",
       "  '136_5128',\n",
       "  '139_1297',\n",
       "  '139_1946',\n",
       "  '13_3097',\n",
       "  '144_4509',\n",
       "  '170_2397',\n",
       "  '170_5931',\n",
       "  '247_3196',\n",
       "  '247_4481',\n",
       "  '30_2127',\n",
       "  '30_4309',\n",
       "  '342_2471',\n",
       "  '342_3054',\n",
       "  '35_3053',\n",
       "  '35_5019',\n",
       "  '36_2507',\n",
       "  '36_5240',\n",
       "  '40_3514',\n",
       "  '40_367',\n",
       "  '43_4452',\n",
       "  '43_767',\n",
       "  '49_1933',\n",
       "  '49_2285',\n",
       "  '49_4805',\n",
       "  '50_1049',\n",
       "  '50_1134',\n",
       "  '51_494',\n",
       "  '51_705',\n",
       "  '52_1129',\n",
       "  '52_3574',\n",
       "  '64_1246',\n",
       "  '64_645',\n",
       "  '66_3500',\n",
       "  '6_6292'],\n",
       " 'sc_genes': ['YGR240C', 'YMR205C'],\n",
       " 'sc_names': ['PFK1', 'PFK2'],\n",
       " 'pathways_id': [\"['GLUCFERMEN-PWY', 'GLYCOLYSIS']\",\n",
       "  \"['GLUCFERMEN-PWY', 'GLYCOLYSIS']\"],\n",
       " 'pathways_name': [\"['glycolysis I (from glucose 6-phosphate)', 'superpathway of glucose fermentation']\",\n",
       "  \"['glycolysis I (from glucose 6-phosphate)', 'superpathway of glucose fermentation']\"],\n",
       " 'EC_number': [\"['2.7.1.11']\", \"['2.7.1.11']\"],\n",
       " 'PDB_ID': [\"['3O8O']\", \"['3O8O']\"],\n",
       " 'spom_genes': ['SPBC16H5.02', 'SPBC16H5.02'],\n",
       " 'calb_genes': ['C5_04810W_A', 'C7_01800C_A'],\n",
       " 'pass_species_filter': True}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_metadata['OG1054']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#How do we choose the structural alignment??\n",
    "## Oliver:  scer genes have 'ref'\n",
    "\n",
    "I see a possible bug. Right now in the file 'OG1004_REF_Scer_AF-P15938-F1-model_v2.FASTA'\n",
    "\n",
    "All the S. cer genes have REF in front of them - not just Scer_AF-P15938-F1-model_v2\n",
    "\n",
    "REF_Scer_AF-P15938-F1-model_v2\n",
    "REF_Scer_AF-P20095-F1-model_v2\n",
    "REF_Scer_AF-P24384-F1-model_v2\n",
    "REF_Scer_AF-P40459-F1-model_v2\n",
    "REF_Scer_AF-P53131-F1-model_v2\n",
    "\n",
    "shouldn't it be:\n",
    "\n",
    "REF_Scer_AF-P15938-F1-model_v2\n",
    "Scer_AF-P20095-F1-model_v2\n",
    "Scer_AF-P24384-F1-model_v2\n",
    "Scer_AF-P40459-F1-model_v2\n",
    "Scer_AF-P53131-F1-model_v2\n",
    "\n",
    "Oliver says that the word 'REF' means it is the Ref species not the ref structure.  The ref structure is just identified in the filename. \n",
    "\n",
    "\n",
    "## Jacob - do we need the identifier_proteome_id (if so use record.description.split(' ')[0])\n",
    "\n",
    "#Doesn't Wrap\n",
    "\n",
    "#S Pombe not included in sequence alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc_gene_id_ref_lookup = pickle.load(open(base_dir + os.sep + os.path.normpath('msas/structural/Genes_identifier_dict.pkl'),\"rb\"))\n",
    "#sc_ref_gene_id_lookup = dict(zip(sc_gene_id_ref_lookup.values(), sc_gene_id_ref_lookup.keys()))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../Data/Scer_protein_names.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-4-de3e3fbda2db>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[0mpath_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m\"/Users/Oliver/OneDrive - Charité - Universitätsmedizin Berlin/Alphafold/Alphafold_ranked_0_PDB/Output/data\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mdata\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"../Data/Scer_protein_names.tsv\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0msep\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"\\t\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex_col\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"OLN\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\sc_seq_py\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36mread_csv\u001b[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[0;32m    684\u001b[0m     )\n\u001b[0;32m    685\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 686\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    687\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    688\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\sc_seq_py\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    450\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    451\u001b[0m     \u001b[1;31m# Create the parser.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 452\u001b[1;33m     \u001b[0mparser\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    453\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    454\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\sc_seq_py\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[0;32m    944\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"has_index_names\"\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"has_index_names\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    945\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 946\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    947\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    948\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\sc_seq_py\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[1;34m(self, engine)\u001b[0m\n\u001b[0;32m   1176\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"c\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1177\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"c\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1178\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1179\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1180\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"python\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\envs\\sc_seq_py\\lib\\site-packages\\pandas\\io\\parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, src, **kwds)\u001b[0m\n\u001b[0;32m   2006\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"usecols\"\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2007\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2008\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2009\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2010\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../Data/Scer_protein_names.tsv'"
     ]
    }
   ],
   "source": [
    "path_file = \"/Users/Oliver/OneDrive - Charité - Universitätsmedizin Berlin/Alphafold/Alphafold_ranked_0_PDB/Output/data\"\n",
    "data = pd.read_csv(\"../Data/Scer_protein_names.tsv\",sep=\"\\t\", index_col=\"OLN\")\n",
    "\n",
    "\n",
    "\n",
    "# Conversion dictionaries\n",
    "# protein2og = pickle.load(open(os.path.join(path_file,\"Genes_scer_dict.pkl\"),\"rb\"))\n",
    "# protein2file = pickle.load(open(os.path.join(path_file,\"files_scer.pkl\"),\"rb\"))\n",
    "# protein2gene = {data.loc[index,\"Swiss-Prot\"]:index for index in data.index}\n",
    "\n",
    "# gene2identifier = {protein2gene[key]:protein2og[key]+\"_REF_Scer_\"+protein2file[key][:-4] for key in protein2file.keys()}\n",
    "\n",
    "\n",
    "#I want to go back from Oliver's protein name to the name from the sequence alignment.  \n",
    "\n",
    "#This is especially hard for S. cer, S.pom and C. alb which have special names from uniprot.  \n",
    "\n",
    "identifier2gene = \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
